

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Matrix module &mdash; FAN-C 0.8.17-beta documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Hic module" href="hic.html" />
    <link rel="prev" title="Pairs module" href="pairs.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> FAN-C
          

          
          </a>

          
            
            
              <div class="version">
                0.8.17
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting started with FAN-C</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fanc.html">fanc: command line tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fancplot.html">fancplot: plotting from the command line</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">Python API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../quickstart.html">Quickstart</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interfaces.html">Interfaces</a></li>
<li class="toctree-l2"><a class="reference internal" href="../generate.html">Generate Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analyse.html">Analyse Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../plot.html">Plotting (API)</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../modules.html">Reference</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="init.html">init</a></li>
<li class="toctree-l3"><a class="reference internal" href="map.html">map</a></li>
<li class="toctree-l3"><a class="reference internal" href="regions.html">regions</a></li>
<li class="toctree-l3"><a class="reference internal" href="pairs.html">pairs</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="hic.html">hic</a></li>
<li class="toctree-l3"><a class="reference internal" href="peaks.html">peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="compatibility.html">compatibility</a></li>
<li class="toctree-l3"><a class="reference internal" href="architecture.html">architecture</a></li>
<li class="toctree-l3"><a class="reference internal" href="plotting.html">plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="tools.html">tools</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">FAN-C</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../api.html">fanc Python module</a> &raquo;</li>
        
          <li><a href="../modules.html">FAN-C Python modules</a> &raquo;</li>
        
      <li>Matrix module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/api/modules/matrix.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-fanc.matrix">
<span id="matrix-module"></span><h1>Matrix module<a class="headerlink" href="#module-fanc.matrix" title="Permalink to this headline">¶</a></h1>
<dl class="class">
<dt id="fanc.matrix.Edge">
<em class="property">class </em><code class="sig-prename descclassname">fanc.matrix.</code><code class="sig-name descname">Edge</code><span class="sig-paren">(</span><em class="sig-param">source</em>, <em class="sig-param">sink</em>, <em class="sig-param">_weight_field='weight'</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.Edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>A contact / an Edge between two genomic regions.</p>
<dl class="attribute">
<dt id="fanc.matrix.Edge.source">
<code class="sig-name descname">source</code><a class="headerlink" href="#fanc.matrix.Edge.source" title="Permalink to this definition">¶</a></dt>
<dd><p>The index of the “source” genomic region. By convention,
source &lt;= sink.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.Edge.sink">
<code class="sig-name descname">sink</code><a class="headerlink" href="#fanc.matrix.Edge.sink" title="Permalink to this definition">¶</a></dt>
<dd><p>The index of the “sink” genomic region.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.Edge.bias">
<code class="sig-name descname">bias</code><a class="headerlink" href="#fanc.matrix.Edge.bias" title="Permalink to this definition">¶</a></dt>
<dd><p>Bias factor obtained via normalisation of the Hi-C matrix</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.Edge.source_node">
<code class="sig-name descname">source_node</code><a class="headerlink" href="#fanc.matrix.Edge.source_node" title="Permalink to this definition">¶</a></dt>
<dd><p>The first <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> in this contact</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.Edge.sink_node">
<code class="sig-name descname">sink_node</code><a class="headerlink" href="#fanc.matrix.Edge.sink_node" title="Permalink to this definition">¶</a></dt>
<dd><p>The second <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> in this contact</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.matrix.LazyEdge">
<em class="property">class </em><code class="sig-prename descclassname">fanc.matrix.</code><code class="sig-name descname">LazyEdge</code><span class="sig-paren">(</span><em class="sig-param">row</em>, <em class="sig-param">regions_table=None</em>, <em class="sig-param">_weight_field='weight'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.LazyEdge" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.matrix.Edge</span></code></a></p>
<p>An <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> equivalent supporting lazy loading.</p>
<dl class="attribute">
<dt id="fanc.matrix.LazyEdge.source">
<code class="sig-name descname">source</code><a class="headerlink" href="#fanc.matrix.LazyEdge.source" title="Permalink to this definition">¶</a></dt>
<dd><p>The index of the “source” genomic region. By convention,
source &lt;= sink.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.LazyEdge.sink">
<code class="sig-name descname">sink</code><a class="headerlink" href="#fanc.matrix.LazyEdge.sink" title="Permalink to this definition">¶</a></dt>
<dd><p>The index of the “sink” genomic region.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.LazyEdge.bias">
<code class="sig-name descname">bias</code><a class="headerlink" href="#fanc.matrix.LazyEdge.bias" title="Permalink to this definition">¶</a></dt>
<dd><p>Bias factor obtained via normalisation of the Hi-C matrix</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.LazyEdge.source_node">
<code class="sig-name descname">source_node</code><a class="headerlink" href="#fanc.matrix.LazyEdge.source_node" title="Permalink to this definition">¶</a></dt>
<dd><p>The first <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> in this contact</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.LazyEdge.sink_node">
<code class="sig-name descname">sink_node</code><a class="headerlink" href="#fanc.matrix.LazyEdge.sink_node" title="Permalink to this definition">¶</a></dt>
<dd><p>The second <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> in this contact</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.LazyEdge.update">
<code class="sig-name descname">update</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.LazyEdge.update" title="Permalink to this definition">¶</a></dt>
<dd><p>Write changes to PyTables row to file.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.matrix.RegionMatrix">
<em class="property">class </em><code class="sig-prename descclassname">fanc.matrix.</code><code class="sig-name descname">RegionMatrix</code><a class="headerlink" href="#fanc.matrix.RegionMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ma.core.MaskedArray</span></code></p>
<p>Subclass of <code class="xref py py-class docutils literal notranslate"><span class="pre">masked_array</span></code> with genomic region support.</p>
<p>Objects of this type are returned by <a class="reference internal" href="#fanc.matrix.RegionMatrixContainer.matrix" title="fanc.matrix.RegionMatrixContainer.matrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">matrix</span></code></a>.
<a class="reference internal" href="#fanc.matrix.RegionMatrix" title="fanc.matrix.RegionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrix</span></code></a> supports subsetting by <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
and region strings of the form <code class="code docutils literal notranslate"><span class="pre">&lt;chromosome&gt;[:&lt;start&gt;-&lt;end&gt;]</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>
<span class="n">hic</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;output/hic/binned/fanc_example_1mb.hic&quot;</span><span class="p">)</span>

<span class="n">m</span> <span class="o">=</span> <span class="n">hic</span><span class="o">.</span><span class="n">matrix</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr18&#39;</span><span class="p">))</span>
<span class="nb">type</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>  <span class="c1"># fanc.matrix.RegionMatrix</span>

<span class="n">m_sub</span> <span class="o">=</span> <span class="n">m</span><span class="p">[</span><span class="s1">&#39;chr18:1-5mb&#39;</span><span class="p">,</span> <span class="s1">&#39;chr18:1-10mb&#39;</span><span class="p">]</span>
<span class="nb">type</span><span class="p">(</span><span class="n">m_sub</span><span class="p">)</span>  <span class="c1"># fanc.matrix.RegionMatrix</span>
<span class="n">m</span><span class="o">.</span><span class="n">shape</span>  <span class="c1"># 5, 10</span>
<span class="n">m_sub</span><span class="o">.</span><span class="n">row_regions</span>  <span class="c1"># [chr18:1-1000000, chr18:1000001-2000000,</span>
                   <span class="c1">#  chr18:2000001-3000000, chr18:3000001-4000000,</span>
                   <span class="c1">#  chr18:4000001-5000000]</span>
</pre></div>
</div>
<p>If the associated row or col regions have a <code class="code docutils literal notranslate"><span class="pre">False</span></code> <code class="code docutils literal notranslate"><span class="pre">valid</span></code>
attribute, the rows/cols of the :<a class="reference internal" href="#fanc.matrix.RegionMatrix" title="fanc.matrix.RegionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrix</span></code></a> will be masked.</p>
<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.row_regions">
<code class="sig-name descname">row_regions</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.row_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of regions matching the first matrix dimension</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.col_regions">
<code class="sig-name descname">col_regions</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.col_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>A list of regions matching the second matrix dimension</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.all">
<code class="sig-name descname">all</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.all" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if all elements evaluate to True.</p>
<p>The output array is masked where all the values along the given axis
are masked: if the output would have been a scalar and that all the
values are masked, then the output is <cite>masked</cite>.</p>
<p>Refer to <cite>numpy.all</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.all()</span></code></dt><dd><p>corresponding function for ndarrays</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.all()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span><span class="o">.</span><span class="n">all</span><span class="p">()</span>
<span class="go">True</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="n">mask</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">all</span><span class="p">()</span> <span class="ow">is</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">masked</span><span class="p">)</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.anom">
<code class="sig-name descname">anom</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">dtype=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.anom" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the anomalies (deviations from the arithmetic mean)
along the given axis.</p>
<p>Returns an array of anomalies, with the same shape as the input and
where the arithmetic mean is computed along the given axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>int</em><em>, </em><em>optional</em>) – Axis over which the anomalies are taken.
The default is to use the mean of the flattened array as reference.</p></li>
<li><p><strong>dtype</strong> (<em>dtype</em><em>, </em><em>optional</em>) – <dl class="simple">
<dt>Type to use in computing the variance. For arrays of integer type</dt><dd><p>the default is float32; for arrays of float types it is the same as
the array type.</p>
</dd>
</dl>
</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#fanc.matrix.RegionMatrix.mean" title="fanc.matrix.RegionMatrix.mean"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mean()</span></code></a></dt><dd><p>Compute the mean of the array.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">anom</span><span class="p">()</span>
<span class="go">masked_array(data=[-1.,  0.,  1.],</span>
<span class="go">             mask=False,</span>
<span class="go">       fill_value=1e+20)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.any">
<code class="sig-name descname">any</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.any" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns True if any of the elements of <cite>a</cite> evaluate to True.</p>
<p>Masked values are considered as False during computation.</p>
<p>Refer to <cite>numpy.any</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.any()</span></code></dt><dd><p>corresponding function for ndarrays</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.any()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.argmax">
<code class="sig-name descname">argmax</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">fill_value=None</em>, <em class="sig-param">out=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.argmax" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns array of indices of the maximum values along the given axis.
Masked values are treated as if they had the value fill_value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>{None</em><em>, </em><em>integer}</em>) – If None, the index is into the flattened array, otherwise along
the specified axis</p></li>
<li><p><strong>fill_value</strong> (<em>{var}</em><em>, </em><em>optional</em>) – Value used to fill in the masked values.  If None, the output of
maximum_fill_value(self._data) is used instead.</p></li>
<li><p><strong>out</strong> (<em>{None</em><em>, </em><em>array}</em><em>, </em><em>optional</em>) – Array into which the result can be placed. Its type is preserved
and it must be of the right shape to hold the output.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>index_array</strong></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>{integer_array}</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">argmax</span><span class="p">()</span>
<span class="go">5</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="go">array([1, 1, 1])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="go">array([2, 2])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.argmin">
<code class="sig-name descname">argmin</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">fill_value=None</em>, <em class="sig-param">out=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.argmin" title="Permalink to this definition">¶</a></dt>
<dd><p>Return array of indices to the minimum values along the given axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>{None</em><em>, </em><em>integer}</em>) – If None, the index is into the flattened array, otherwise along
the specified axis</p></li>
<li><p><strong>fill_value</strong> (<em>{var}</em><em>, </em><em>optional</em>) – Value used to fill in the masked values.  If None, the output of
minimum_fill_value(self._data) is used instead.</p></li>
<li><p><strong>out</strong> (<em>{None</em><em>, </em><em>array}</em><em>, </em><em>optional</em>) – Array into which the result can be placed. Its type is preserved
and it must be of the right shape to hold the output.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>If multi-dimension input, returns a new ndarray of indices to the
minimum values along the given axis.  Otherwise, returns a scalar
of index to the minimum values along the given axis.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray or scalar</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">4</span><span class="p">),</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">shape</span> <span class="o">=</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[--, --],</span>
<span class="go">        [2, 3]],</span>
<span class="go">  mask=[[ True,  True],</span>
<span class="go">        [False, False]],</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=-</span><span class="mi">1</span><span class="p">)</span>
<span class="go">array([0, 0])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mi">9</span><span class="p">)</span>
<span class="go">array([1, 1])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.argpartition">
<code class="sig-name descname">argpartition</code><span class="sig-paren">(</span><em class="sig-param">kth</em>, <em class="sig-param">axis=-1</em>, <em class="sig-param">kind='introselect'</em>, <em class="sig-param">order=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.argpartition" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the indices that would partition this array.</p>
<p>Refer to <cite>numpy.argpartition</cite> for full documentation.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.8.0.</span></p>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.argpartition()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.argsort">
<code class="sig-name descname">argsort</code><span class="sig-paren">(</span><em class="sig-param">axis=&lt;no value&gt;</em>, <em class="sig-param">kind=None</em>, <em class="sig-param">order=None</em>, <em class="sig-param">endwith=True</em>, <em class="sig-param">fill_value=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.argsort" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an ndarray of indices that sort the array along the
specified axis.  Masked values are filled beforehand to
<cite>fill_value</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>int</em><em>, </em><em>optional</em>) – <p>Axis along which to sort. If None, the default, the flattened array
is used.</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.13.0: </span>Previously, the default was documented to be -1, but that was
in error. At some future date, the default will change to -1, as
originally intended.
Until then, the axis should be given explicitly when
<code class="docutils literal notranslate"><span class="pre">arr.ndim</span> <span class="pre">&gt;</span> <span class="pre">1</span></code>, to avoid a FutureWarning.</p>
</div>
</p></li>
<li><p><strong>kind</strong> (<em>{'quicksort'</em><em>, </em><em>'mergesort'</em><em>, </em><em>'heapsort'</em><em>, </em><em>'stable'}</em><em>, </em><em>optional</em>) – The sorting algorithm used.</p></li>
<li><p><strong>order</strong> (<em>list</em><em>, </em><em>optional</em>) – When <cite>a</cite> is an array with fields defined, this argument specifies
which fields to compare first, second, etc.  Not all fields need be
specified.</p></li>
<li><p><strong>endwith</strong> (<em>{True</em><em>, </em><em>False}</em><em>, </em><em>optional</em>) – Whether missing values (if any) should be treated as the largest values
(True) or the smallest values (False)
When the array contains unmasked values at the same extremes of the
datatype, the ordering of these values and the masked values is
undefined.</p></li>
<li><p><strong>fill_value</strong> (<em>{var}</em><em>, </em><em>optional</em>) – Value used internally for the masked values.
If <code class="docutils literal notranslate"><span class="pre">fill_value</span></code> is not None, it supersedes <code class="docutils literal notranslate"><span class="pre">endwith</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>index_array</strong> – Array of indices that sort <cite>a</cite> along the specified axis.
In other words, <code class="docutils literal notranslate"><span class="pre">a[index_array]</span></code> yields a sorted <cite>a</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray, int</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">MaskedArray.sort()</span></code></dt><dd><p>Describes sorting algorithms used.</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">lexsort()</span></code></dt><dd><p>Indirect stable sort with multiple keys.</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.sort()</span></code></dt><dd><p>Inplace sort.</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>See <cite>sort</cite> for notes on the different sorting algorithms.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">masked_array(data=[3, 2, --],</span>
<span class="go">             mask=[False, False,  True],</span>
<span class="go">       fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">argsort</span><span class="p">()</span>
<span class="go">array([1, 0, 2])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.astype">
<code class="sig-name descname">astype</code><span class="sig-paren">(</span><em class="sig-param">dtype</em>, <em class="sig-param">order='K'</em>, <em class="sig-param">casting='unsafe'</em>, <em class="sig-param">subok=True</em>, <em class="sig-param">copy=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.astype" title="Permalink to this definition">¶</a></dt>
<dd><p>Copy of the array, cast to a specified type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dtype</strong> (<em>str</em><em> or </em><em>dtype</em>) – Typecode or data-type to which the array is cast.</p></li>
<li><p><strong>order</strong> (<em>{'C'</em><em>, </em><em>'F'</em><em>, </em><em>'A'</em><em>, </em><em>'K'}</em><em>, </em><em>optional</em>) – Controls the memory layout order of the result.
‘C’ means C order, ‘F’ means Fortran order, ‘A’
means ‘F’ order if all the arrays are Fortran contiguous,
‘C’ order otherwise, and ‘K’ means as close to the
order the array elements appear in memory as possible.
Default is ‘K’.</p></li>
<li><p><strong>casting</strong> (<em>{'no'</em><em>, </em><em>'equiv'</em><em>, </em><em>'safe'</em><em>, </em><em>'same_kind'</em><em>, </em><em>'unsafe'}</em><em>, </em><em>optional</em>) – <p>Controls what kind of data casting may occur. Defaults to ‘unsafe’
for backwards compatibility.</p>
<blockquote>
<div><ul>
<li><p>’no’ means the data types should not be cast at all.</p></li>
<li><p>’equiv’ means only byte-order changes are allowed.</p></li>
<li><p>’safe’ means only casts which can preserve values are allowed.</p></li>
<li><p>’same_kind’ means only safe casts or casts within a kind,
like float64 to float32, are allowed.</p></li>
<li><p>’unsafe’ means any data conversions may be done.</p></li>
</ul>
</div></blockquote>
</p></li>
<li><p><strong>subok</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, then sub-classes will be passed-through (default), otherwise
the returned array will be forced to be a base-class array.</p></li>
<li><p><strong>copy</strong> (<em>bool</em><em>, </em><em>optional</em>) – By default, astype always returns a newly allocated array. If this
is set to false, and the <cite>dtype</cite>, <cite>order</cite>, and <cite>subok</cite>
requirements are satisfied, the input array is returned instead
of a copy.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>arr_t</strong> – Unless <cite>copy</cite> is False and the other conditions for returning the input
array are satisfied (see description for <cite>copy</cite> input parameter), <cite>arr_t</cite>
is a new array of the same shape as the input array, with dtype, order
given by <cite>dtype</cite>, <cite>order</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.17.0: </span>Casting between a simple data type and a structured one is possible only
for “unsafe” casting.  Casting to multiple fields is allowed, but
casting from multiple fields is not.</p>
</div>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.9.0: </span>Casting from numeric to string types in ‘safe’ casting mode requires
that the string dtype length is long enough to store the max
integer/float value converted.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>ComplexWarning</strong> – When casting from complex to float or int. To avoid this,
    one should use <code class="docutils literal notranslate"><span class="pre">a.real.astype(t)</span></code>.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mf">2.5</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([1. ,  2. ,  2.5])</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
<span class="go">array([1, 2, 2])</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.base">
<code class="sig-name descname">base</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.base" title="Permalink to this definition">¶</a></dt>
<dd><p>Base object if memory is from some other object.</p>
<p class="rubric">Examples</p>
<p>The base of an array that owns its memory is None:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">base</span> <span class="ow">is</span> <span class="kc">None</span>
<span class="go">True</span>
</pre></div>
</div>
<p>Slicing creates a view, whose memory is shared with x:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">x</span><span class="p">[</span><span class="mi">2</span><span class="p">:]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">base</span> <span class="ow">is</span> <span class="n">x</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.baseclass">
<em class="property">property </em><code class="sig-name descname">baseclass</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.baseclass" title="Permalink to this definition">¶</a></dt>
<dd><p>Class of the underlying data (read-only).</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.byteswap">
<code class="sig-name descname">byteswap</code><span class="sig-paren">(</span><em class="sig-param">inplace=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.byteswap" title="Permalink to this definition">¶</a></dt>
<dd><p>Swap the bytes of the array elements</p>
<p>Toggle between low-endian and big-endian data representation by
returning a byteswapped array, optionally swapped in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>optional</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, swap bytes in-place, default is <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>out</strong> – The byteswapped array. If <cite>inplace</cite> is <code class="docutils literal notranslate"><span class="pre">True</span></code>, this is
a view to self.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">A</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">256</span><span class="p">,</span> <span class="mi">8755</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">hex</span><span class="p">,</span> <span class="n">A</span><span class="p">))</span>
<span class="go">[&#39;0x1&#39;, &#39;0x100&#39;, &#39;0x2233&#39;]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">A</span><span class="o">.</span><span class="n">byteswap</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="go">array([  256,     1, 13090], dtype=int16)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">hex</span><span class="p">,</span> <span class="n">A</span><span class="p">))</span>
<span class="go">[&#39;0x100&#39;, &#39;0x1&#39;, &#39;0x3322&#39;]</span>
</pre></div>
</div>
<p>Arrays of strings are not swapped</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">A</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="s1">&#39;ceg&#39;</span><span class="p">,</span> <span class="s1">&#39;fac&#39;</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">A</span><span class="o">.</span><span class="n">byteswap</span><span class="p">()</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">UnicodeDecodeError</span>: <span class="n">...</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.choose">
<code class="sig-name descname">choose</code><span class="sig-paren">(</span><em class="sig-param">choices</em>, <em class="sig-param">out=None</em>, <em class="sig-param">mode='raise'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.choose" title="Permalink to this definition">¶</a></dt>
<dd><p>Use an index array to construct a new array from a set of choices.</p>
<p>Refer to <cite>numpy.choose</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.choose()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.clip">
<code class="sig-name descname">clip</code><span class="sig-paren">(</span><em class="sig-param">min=None</em>, <em class="sig-param">max=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.clip" title="Permalink to this definition">¶</a></dt>
<dd><p>Return an array whose values are limited to <code class="docutils literal notranslate"><span class="pre">[min,</span> <span class="pre">max]</span></code>.
One of max or min must be given.</p>
<p>Refer to <cite>numpy.clip</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.clip()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.compress">
<code class="sig-name descname">compress</code><span class="sig-paren">(</span><em class="sig-param">condition</em>, <em class="sig-param">axis=None</em>, <em class="sig-param">out=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.compress" title="Permalink to this definition">¶</a></dt>
<dd><p>Return <cite>a</cite> where condition is <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
<p>If condition is a <cite>MaskedArray</cite>, missing values are considered
as <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>condition</strong> (<em>var</em>) – Boolean 1-d array selecting which entries to return. If len(condition)
is less than the size of a along the axis, then output is truncated
to length of condition array.</p></li>
<li><p><strong>axis</strong> (<em>{None</em><em>, </em><em>int}</em><em>, </em><em>optional</em>) – Axis along which the operation must be performed.</p></li>
<li><p><strong>out</strong> (<em>{None</em><em>, </em><em>ndarray}</em><em>, </em><em>optional</em>) – Alternative output array in which to place the result. It must have
the same shape as the expected output but the type will be cast if
necessary.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>result</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">MaskedArray</span></code> object.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>MaskedArray</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Please note the difference with <a class="reference internal" href="#fanc.matrix.RegionMatrix.compressed" title="fanc.matrix.RegionMatrix.compressed"><code class="xref py py-meth docutils literal notranslate"><span class="pre">compressed()</span></code></a> !
The output of <a class="reference internal" href="#fanc.matrix.RegionMatrix.compress" title="fanc.matrix.RegionMatrix.compress"><code class="xref py py-meth docutils literal notranslate"><span class="pre">compress()</span></code></a> has a mask, the output of
<a class="reference internal" href="#fanc.matrix.RegionMatrix.compressed" title="fanc.matrix.RegionMatrix.compressed"><code class="xref py py-meth docutils literal notranslate"><span class="pre">compressed()</span></code></a> does not.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">],[</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">]],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[1, --, 3],</span>
<span class="go">        [--, 5, --],</span>
<span class="go">        [7, --, 9]],</span>
<span class="go">  mask=[[False,  True, False],</span>
<span class="go">        [ True, False,  True],</span>
<span class="go">        [False,  True, False]],</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">compress</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="go">masked_array(data=[1, 3],</span>
<span class="go">             mask=[False, False],</span>
<span class="go">       fill_value=999999)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">compress</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[1, 3],</span>
<span class="go">        [--, --],</span>
<span class="go">        [7, 9]],</span>
<span class="go">  mask=[[False, False],</span>
<span class="go">        [ True,  True],</span>
<span class="go">        [False, False]],</span>
<span class="go">  fill_value=999999)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.compressed">
<code class="sig-name descname">compressed</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.compressed" title="Permalink to this definition">¶</a></dt>
<dd><p>Return all the non-masked data as a 1-D array.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><strong>data</strong> – A new <cite>ndarray</cite> holding the non-masked data is returned.</p>
</dd>
<dt class="field-even">Return type</dt>
<dd class="field-even"><p>ndarray</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The result is <strong>not</strong> a MaskedArray!</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">5</span><span class="p">),</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mi">2</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">*</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">compressed</span><span class="p">()</span>
<span class="go">array([0, 1])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">compressed</span><span class="p">())</span>
<span class="go">&lt;class &#39;numpy.ndarray&#39;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.conj">
<code class="sig-name descname">conj</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.conj" title="Permalink to this definition">¶</a></dt>
<dd><p>Complex-conjugate all elements.</p>
<p>Refer to <cite>numpy.conjugate</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.conjugate()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.conjugate">
<code class="sig-name descname">conjugate</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.conjugate" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the complex conjugate, element-wise.</p>
<p>Refer to <cite>numpy.conjugate</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.conjugate()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><em class="sig-param">order='C'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a copy of the array.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>order</strong> (<em>{'C'</em><em>, </em><em>'F'</em><em>, </em><em>'A'</em><em>, </em><em>'K'}</em><em>, </em><em>optional</em>) – Controls the memory layout of the copy. ‘C’ means C-order,
‘F’ means F-order, ‘A’ means ‘F’ if <cite>a</cite> is Fortran contiguous,
‘C’ otherwise. ‘K’ means match the layout of <cite>a</cite> as closely
as possible. (Note that this function and <code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.copy()</span></code> are very
similar, but have different default values for their order=
arguments.)</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.copy()</span></code>, <code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.copyto()</span></code></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">]],</span> <span class="n">order</span><span class="o">=</span><span class="s1">&#39;F&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([[0, 0, 0],</span>
<span class="go">       [0, 0, 0]])</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">y</span>
<span class="go">array([[1, 2, 3],</span>
<span class="go">       [4, 5, 6]])</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">flags</span><span class="p">[</span><span class="s1">&#39;C_CONTIGUOUS&#39;</span><span class="p">]</span>
<span class="go">True</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.count">
<code class="sig-name descname">count</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.count" title="Permalink to this definition">¶</a></dt>
<dd><p>Count the non-masked elements of the array along the given axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>None</em><em> or </em><em>int</em><em> or </em><em>tuple of ints</em><em>, </em><em>optional</em>) – <p>Axis or axes along which the count is performed.
The default (<cite>axis</cite> = <cite>None</cite>) performs the count over all
the dimensions of the input array. <cite>axis</cite> may be negative, in
which case it counts from the last to the first axis.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.10.0.</span></p>
</div>
<p>If this is a tuple of ints, the count is performed on multiple
axes, instead of a single axis or all the axes as before.</p>
</p></li>
<li><p><strong>keepdims</strong> (<em>bool</em><em>, </em><em>optional</em>) – If this is set to True, the axes which are reduced are left
in the result as dimensions with size one. With this option,
the result will broadcast correctly against the array.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>result</strong> – An array with the same shape as the input array, with the specified
axis removed. If the array is a 0-d array, or if <cite>axis</cite> is None, a
scalar is returned.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray or scalar</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">count_masked()</span></code></dt><dd><p>Count masked elements in array or along a given axis.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy.ma</span> <span class="k">as</span> <span class="nn">ma</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">6</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">masked</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[0, 1, 2],</span>
<span class="go">        [--, --, --]],</span>
<span class="go">  mask=[[False, False, False],</span>
<span class="go">        [ True,  True,  True]],</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
<span class="go">3</span>
</pre></div>
</div>
<p>When the <cite>axis</cite> keyword is specified an array of appropriate size is
returned.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">array([1, 1, 1])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">array([3, 0])</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.ctypes">
<code class="sig-name descname">ctypes</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.ctypes" title="Permalink to this definition">¶</a></dt>
<dd><p>An object to simplify the interaction of the array with the ctypes
module.</p>
<p>This attribute creates an object that makes it easier to use arrays
when calling shared libraries with the ctypes module. The returned
object has, among others, data, shape, and strides attributes (see
Notes below) which themselves return ctypes objects that can be used
as arguments to a shared library.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>None</strong> – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>c</strong> – Possessing attributes data, shape, strides, etc.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Python object</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numpy.ctypeslib</span></code></p>
</div>
<p class="rubric">Notes</p>
<p>Below are the public attributes of this object which were documented
in “Guide to NumPy” (we have omitted undocumented public attributes,
as well as documented private attributes):</p>
<dl class="attribute">
<dt>
<code class="sig-prename descclassname">_ctypes.</code><code class="sig-name descname">data</code></dt>
<dd><p>A pointer to the memory area of the array as a Python integer.
This memory area may contain data that is not aligned, or not in correct
byte-order. The memory area may not even be writeable. The array
flags and data-type of this array should be respected when passing this
attribute to arbitrary C-code to avoid trouble that can include Python
crashing. User Beware! The value of this attribute is exactly the same
as <code class="docutils literal notranslate"><span class="pre">self._array_interface_['data'][0]</span></code>.</p>
<p>Note that unlike <cite>data_as</cite>, a reference will not be kept to the array:
code like <code class="docutils literal notranslate"><span class="pre">ctypes.c_void_p((a</span> <span class="pre">+</span> <span class="pre">b).ctypes.data)</span></code> will result in a
pointer to a deallocated array, and should be spelt
<code class="docutils literal notranslate"><span class="pre">(a</span> <span class="pre">+</span> <span class="pre">b).ctypes.data_as(ctypes.c_void_p)</span></code></p>
</dd></dl>

<dl class="attribute">
<dt>
<code class="sig-prename descclassname">_ctypes.</code><code class="sig-name descname">shape</code></dt>
<dd><p>A ctypes array of length self.ndim where
the basetype is the C-integer corresponding to <code class="docutils literal notranslate"><span class="pre">dtype('p')</span></code> on this
platform. This base-type could be <cite>ctypes.c_int</cite>, <cite>ctypes.c_long</cite>, or
<cite>ctypes.c_longlong</cite> depending on the platform.
The c_intp type is defined accordingly in <cite>numpy.ctypeslib</cite>.
The ctypes array contains the shape of the underlying array.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>(c_intp*self.ndim)</p>
</dd>
</dl>
</dd></dl>

<dl class="attribute">
<dt>
<code class="sig-prename descclassname">_ctypes.</code><code class="sig-name descname">strides</code></dt>
<dd><p>A ctypes array of length self.ndim where
the basetype is the same as for the shape attribute. This ctypes array
contains the strides information from the underlying array. This strides
information is important for showing how many bytes must be jumped to
get to the next element in the array.</p>
<dl class="field-list simple">
<dt class="field-odd">Type</dt>
<dd class="field-odd"><p>(c_intp*self.ndim)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-prename descclassname">_ctypes.</code><code class="sig-name descname">data_as</code><span class="sig-paren">(</span><em class="sig-param">obj</em><span class="sig-paren">)</span></dt>
<dd><p>Return the data pointer cast to a particular c-types object.
For example, calling <code class="docutils literal notranslate"><span class="pre">self._as_parameter_</span></code> is equivalent to
<code class="docutils literal notranslate"><span class="pre">self.data_as(ctypes.c_void_p)</span></code>. Perhaps you want to use the data as a
pointer to a ctypes array of floating-point data:
<code class="docutils literal notranslate"><span class="pre">self.data_as(ctypes.POINTER(ctypes.c_double))</span></code>.</p>
<p>The returned pointer will keep a reference to the array.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-prename descclassname">_ctypes.</code><code class="sig-name descname">shape_as</code><span class="sig-paren">(</span><em class="sig-param">obj</em><span class="sig-paren">)</span></dt>
<dd><p>Return the shape tuple as an array of some other c-types
type. For example: <code class="docutils literal notranslate"><span class="pre">self.shape_as(ctypes.c_short)</span></code>.</p>
</dd></dl>

<dl class="method">
<dt>
<code class="sig-prename descclassname">_ctypes.</code><code class="sig-name descname">strides_as</code><span class="sig-paren">(</span><em class="sig-param">obj</em><span class="sig-paren">)</span></dt>
<dd><p>Return the strides tuple as an array of some other
c-types type. For example: <code class="docutils literal notranslate"><span class="pre">self.strides_as(ctypes.c_longlong)</span></code>.</p>
</dd></dl>

<p>If the ctypes module is not available, then the ctypes attribute
of array objects still returns something useful, but ctypes objects
are not returned and errors may be raised instead. In particular,
the object will still have the <code class="docutils literal notranslate"><span class="pre">as_parameter</span></code> attribute which will
return an integer equal to the data attribute.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">ctypes</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([[0, 1],</span>
<span class="go">       [2, 3]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ctypes</span><span class="o">.</span><span class="n">data</span>
<span class="go">30439712</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ctypes</span><span class="o">.</span><span class="n">data_as</span><span class="p">(</span><span class="n">ctypes</span><span class="o">.</span><span class="n">POINTER</span><span class="p">(</span><span class="n">ctypes</span><span class="o">.</span><span class="n">c_long</span><span class="p">))</span>
<span class="go">&lt;ctypes.LP_c_long object at 0x01F01300&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ctypes</span><span class="o">.</span><span class="n">data_as</span><span class="p">(</span><span class="n">ctypes</span><span class="o">.</span><span class="n">POINTER</span><span class="p">(</span><span class="n">ctypes</span><span class="o">.</span><span class="n">c_long</span><span class="p">))</span><span class="o">.</span><span class="n">contents</span>
<span class="go">c_long(0)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ctypes</span><span class="o">.</span><span class="n">data_as</span><span class="p">(</span><span class="n">ctypes</span><span class="o">.</span><span class="n">POINTER</span><span class="p">(</span><span class="n">ctypes</span><span class="o">.</span><span class="n">c_longlong</span><span class="p">))</span><span class="o">.</span><span class="n">contents</span>
<span class="go">c_longlong(4294967296L)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ctypes</span><span class="o">.</span><span class="n">shape</span>
<span class="go">&lt;numpy.core._internal.c_long_Array_2 object at 0x01FFD580&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ctypes</span><span class="o">.</span><span class="n">shape_as</span><span class="p">(</span><span class="n">ctypes</span><span class="o">.</span><span class="n">c_long</span><span class="p">)</span>
<span class="go">&lt;numpy.core._internal.c_long_Array_2 object at 0x01FCE620&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ctypes</span><span class="o">.</span><span class="n">strides</span>
<span class="go">&lt;numpy.core._internal.c_long_Array_2 object at 0x01FCE620&gt;</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ctypes</span><span class="o">.</span><span class="n">strides_as</span><span class="p">(</span><span class="n">ctypes</span><span class="o">.</span><span class="n">c_longlong</span><span class="p">)</span>
<span class="go">&lt;numpy.core._internal.c_longlong_Array_2 object at 0x01F01300&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.cumprod">
<code class="sig-name descname">cumprod</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">dtype=None</em>, <em class="sig-param">out=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.cumprod" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the cumulative product of the array elements over the given axis.</p>
<p>Masked values are set to 1 internally during the computation.
However, their position is saved, and the result will be masked at
the same locations.</p>
<p>Refer to <cite>numpy.cumprod</cite> for full documentation.</p>
<p class="rubric">Notes</p>
<p>The mask is lost if <cite>out</cite> is not a valid MaskedArray !</p>
<p>Arithmetic is modular when using integer types, and no error is
raised on overflow.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.cumprod()</span></code></dt><dd><p>corresponding function for ndarrays</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.cumprod()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.cumsum">
<code class="sig-name descname">cumsum</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">dtype=None</em>, <em class="sig-param">out=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.cumsum" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the cumulative sum of the array elements over the given axis.</p>
<p>Masked values are set to 0 internally during the computation.
However, their position is saved, and the result will be masked at
the same locations.</p>
<p>Refer to <cite>numpy.cumsum</cite> for full documentation.</p>
<p class="rubric">Notes</p>
<p>The mask is lost if <cite>out</cite> is not a valid <code class="xref py py-class docutils literal notranslate"><span class="pre">MaskedArray</span></code> !</p>
<p>Arithmetic is modular when using integer types, and no error is
raised on overflow.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.cumsum()</span></code></dt><dd><p>corresponding function for ndarrays</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.cumsum()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">marr</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">10</span><span class="p">),</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">marr</span><span class="o">.</span><span class="n">cumsum</span><span class="p">()</span>
<span class="go">masked_array(data=[0, 1, 3, --, --, --, 9, 16, 24, 33],</span>
<span class="go">             mask=[False, False, False,  True,  True,  True, False, False,</span>
<span class="go">                   False, False],</span>
<span class="go">       fill_value=999999)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.data">
<em class="property">property </em><code class="sig-name descname">data</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.data" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the underlying data, as a view of the masked array.</p>
<p>If the underlying data is a subclass of <code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code>, it is
returned as such.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">matrix</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]]),</span> <span class="n">mask</span><span class="o">=</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">data</span>
<span class="go">matrix([[1, 2],</span>
<span class="go">        [3, 4]])</span>
</pre></div>
</div>
<p>The type of the data can be accessed through the <a class="reference internal" href="#fanc.matrix.RegionMatrix.baseclass" title="fanc.matrix.RegionMatrix.baseclass"><code class="xref py py-attr docutils literal notranslate"><span class="pre">baseclass</span></code></a>
attribute.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.diagonal">
<code class="sig-name descname">diagonal</code><span class="sig-paren">(</span><em class="sig-param">offset=0</em>, <em class="sig-param">axis1=0</em>, <em class="sig-param">axis2=1</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.diagonal" title="Permalink to this definition">¶</a></dt>
<dd><p>Return specified diagonals. In NumPy 1.9 the returned array is a
read-only view instead of a copy as in previous NumPy versions.  In
a future version the read-only restriction will be removed.</p>
<p>Refer to <code class="xref py py-func docutils literal notranslate"><span class="pre">numpy.diagonal()</span></code> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.diagonal()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.dot">
<code class="sig-name descname">dot</code><span class="sig-paren">(</span><em class="sig-param">b</em>, <em class="sig-param">out=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.dot" title="Permalink to this definition">¶</a></dt>
<dd><p>Masked dot product of two arrays. Note that <cite>out</cite> and <cite>strict</cite> are
located in different positions than in <cite>ma.dot</cite>. In order to
maintain compatibility with the functional version, it is
recommended that the optional arguments be treated as keyword only.
At some point that may be mandatory.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.10.0.</span></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>b</strong> (<em>masked_array_like</em>) – Inputs array.</p></li>
<li><p><strong>out</strong> (<em>masked_array</em><em>, </em><em>optional</em>) – Output argument. This must have the exact kind that would be
returned if it was not used. In particular, it must have the
right type, must be C-contiguous, and its dtype must be the
dtype that would be returned for <cite>ma.dot(a,b)</cite>. This is a
performance feature. Therefore, if these conditions are not
met, an exception is raised, instead of attempting to be
flexible.</p></li>
<li><p><strong>strict</strong> (<em>bool</em><em>, </em><em>optional</em>) – <p>Whether masked data are propagated (True) or set to 0 (False)
for the computation. Default is False.  Propagating the mask
means that if a masked value appears in a row or column, the
whole row or column is considered masked.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.10.2.</span></p>
</div>
</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.ma.dot()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.dtype">
<em class="property">property </em><code class="sig-name descname">dtype</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.dtype" title="Permalink to this definition">¶</a></dt>
<dd><p>Data-type of the array’s elements.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>None</strong> – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>d</strong></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>numpy dtype object</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numpy.dtype</span></code></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([[0, 1],</span>
<span class="go">       [2, 3]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">dtype</span>
<span class="go">dtype(&#39;int32&#39;)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">dtype</span><span class="p">)</span>
<span class="go">&lt;type &#39;numpy.dtype&#39;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.dump">
<code class="sig-name descname">dump</code><span class="sig-paren">(</span><em class="sig-param">file</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.dump" title="Permalink to this definition">¶</a></dt>
<dd><p>Dump a pickle of the array to the specified file.
The array can be read back with pickle.load or numpy.load.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>file</strong> (<em>str</em><em> or </em><em>Path</em>) – <p>A string naming the dump file.</p>
<div class="versionchanged">
<p><span class="versionmodified changed">Changed in version 1.17.0: </span><cite>pathlib.Path</cite> objects are now accepted.</p>
</div>
</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.dumps">
<code class="sig-name descname">dumps</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.dumps" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the pickle of the array as a string.
pickle.loads or numpy.loads will convert the string back to an array.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>None</strong> – </p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.fill">
<code class="sig-name descname">fill</code><span class="sig-paren">(</span><em class="sig-param">value</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.fill" title="Permalink to this definition">¶</a></dt>
<dd><p>Fill the array with a scalar value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>value</strong> (<em>scalar</em>) – All elements of <cite>a</cite> will be assigned this value.</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">fill</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">array([0, 0])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">empty</span><span class="p">(</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">fill</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">array([1.,  1.])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.fill_value">
<em class="property">property </em><code class="sig-name descname">fill_value</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.fill_value" title="Permalink to this definition">¶</a></dt>
<dd><p>The filling value of the masked array is a scalar. When setting, None
will set to a default based on the data type.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">dt</span> <span class="ow">in</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">complex128</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">dt</span><span class="p">)</span><span class="o">.</span><span class="n">get_fill_value</span><span class="p">()</span>
<span class="gp">...</span>
<span class="go">999999</span>
<span class="go">999999</span>
<span class="go">1e+20</span>
<span class="go">(1e+20+0j)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.</span><span class="p">],</span> <span class="n">fill_value</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill_value</span>
<span class="go">-inf</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill_value</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">pi</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill_value</span>
<span class="go">3.1415926535897931 # may vary</span>
</pre></div>
</div>
<p>Reset to default:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill_value</span> <span class="o">=</span> <span class="kc">None</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill_value</span>
<span class="go">1e+20</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.filled">
<code class="sig-name descname">filled</code><span class="sig-paren">(</span><em class="sig-param">fill_value=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.filled" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a copy of self, with masked values filled with a given value.
<strong>However</strong>, if there are no masked values to fill, self will be
returned instead as an ndarray.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fill_value</strong> (<em>scalar</em><em>, </em><em>optional</em>) – The value to use for invalid entries (None by default).
If None, the <cite>fill_value</cite> attribute of the array is used instead.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>filled_array</strong> – A copy of <code class="docutils literal notranslate"><span class="pre">self</span></code> with invalid entries replaced by <em>fill_value</em>
(be it the function argument or the attribute of <code class="docutils literal notranslate"><span class="pre">self</span></code>), or
<code class="docutils literal notranslate"><span class="pre">self</span></code> itself as an ndarray if there are no invalid entries to
be replaced.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The result is <strong>not</strong> a MaskedArray!</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">],</span> <span class="n">fill_value</span><span class="o">=-</span><span class="mi">999</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">filled</span><span class="p">()</span>
<span class="go">array([   1,    2, -999,    4, -999])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">type</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">filled</span><span class="p">())</span>
<span class="go">&lt;class &#39;numpy.ndarray&#39;&gt;</span>
</pre></div>
</div>
<p>Subclassing is preserved. This means that if, e.g., the data part of
the masked array is a recarray, <cite>filled</cite> returns a recarray:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="o">-</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">)],</span> <span class="n">dtype</span><span class="o">=</span><span class="s1">&#39;i8,i8&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">recarray</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="p">[(</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">),</span> <span class="p">(</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">m</span><span class="o">.</span><span class="n">filled</span><span class="p">()</span>
<span class="go">rec.array([(999999,      2), (    -3, 999999)],</span>
<span class="go">          dtype=[(&#39;f0&#39;, &#39;&lt;i8&#39;), (&#39;f1&#39;, &#39;&lt;i8&#39;)])</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.flags">
<code class="sig-name descname">flags</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.flags" title="Permalink to this definition">¶</a></dt>
<dd><p>Information about the memory layout of the array.</p>
<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.C_CONTIGUOUS">
<code class="sig-name descname">C_CONTIGUOUS</code><span class="sig-paren">(</span><em class="sig-param">C</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.C_CONTIGUOUS" title="Permalink to this definition">¶</a></dt>
<dd><p>The data is in a single, C-style contiguous segment.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.F_CONTIGUOUS">
<code class="sig-name descname">F_CONTIGUOUS</code><span class="sig-paren">(</span><em class="sig-param">F</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.F_CONTIGUOUS" title="Permalink to this definition">¶</a></dt>
<dd><p>The data is in a single, Fortran-style contiguous segment.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.OWNDATA">
<code class="sig-name descname">OWNDATA</code><span class="sig-paren">(</span><em class="sig-param">O</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.OWNDATA" title="Permalink to this definition">¶</a></dt>
<dd><p>The array owns the memory it uses or borrows it from another object.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.WRITEABLE">
<code class="sig-name descname">WRITEABLE</code><span class="sig-paren">(</span><em class="sig-param">W</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.WRITEABLE" title="Permalink to this definition">¶</a></dt>
<dd><p>The data area can be written to.  Setting this to False locks
the data, making it read-only.  A view (slice, etc.) inherits WRITEABLE
from its base array at creation time, but a view of a writeable
array may be subsequently locked while the base array remains writeable.
(The opposite is not true, in that a view of a locked array may not
be made writeable.  However, currently, locking a base object does not
lock any views that already reference it, so under that circumstance it
is possible to alter the contents of a locked array via a previously
created writeable view onto it.)  Attempting to change a non-writeable
array raises a RuntimeError exception.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.ALIGNED">
<code class="sig-name descname">ALIGNED</code><span class="sig-paren">(</span><em class="sig-param">A</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.ALIGNED" title="Permalink to this definition">¶</a></dt>
<dd><p>The data and all elements are aligned appropriately for the hardware.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.WRITEBACKIFCOPY">
<code class="sig-name descname">WRITEBACKIFCOPY</code><span class="sig-paren">(</span><em class="sig-param">X</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.WRITEBACKIFCOPY" title="Permalink to this definition">¶</a></dt>
<dd><p>This array is a copy of some other array. The C-API function
PyArray_ResolveWritebackIfCopy must be called before deallocating
to the base array will be updated with the contents of this array.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.UPDATEIFCOPY">
<code class="sig-name descname">UPDATEIFCOPY</code><span class="sig-paren">(</span><em class="sig-param">U</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.UPDATEIFCOPY" title="Permalink to this definition">¶</a></dt>
<dd><p>(Deprecated, use WRITEBACKIFCOPY) This array is a copy of some other array.
When this array is
deallocated, the base array will be updated with the contents of
this array.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.FNC">
<code class="sig-name descname">FNC</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.FNC" title="Permalink to this definition">¶</a></dt>
<dd><p>F_CONTIGUOUS and not C_CONTIGUOUS.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.FORC">
<code class="sig-name descname">FORC</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.FORC" title="Permalink to this definition">¶</a></dt>
<dd><p>F_CONTIGUOUS or C_CONTIGUOUS (one-segment test).</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.BEHAVED">
<code class="sig-name descname">BEHAVED</code><span class="sig-paren">(</span><em class="sig-param">B</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.BEHAVED" title="Permalink to this definition">¶</a></dt>
<dd><p>ALIGNED and WRITEABLE.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.CARRAY">
<code class="sig-name descname">CARRAY</code><span class="sig-paren">(</span><em class="sig-param">CA</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.CARRAY" title="Permalink to this definition">¶</a></dt>
<dd><p>BEHAVED and C_CONTIGUOUS.</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.FARRAY">
<code class="sig-name descname">FARRAY</code><span class="sig-paren">(</span><em class="sig-param">FA</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.FARRAY" title="Permalink to this definition">¶</a></dt>
<dd><p>BEHAVED and F_CONTIGUOUS and not C_CONTIGUOUS.</p>
</dd></dl>

<p class="rubric">Notes</p>
<p>The <cite>flags</cite> object can be accessed dictionary-like (as in <code class="docutils literal notranslate"><span class="pre">a.flags['WRITEABLE']</span></code>),
or by using lowercased attribute names (as in <code class="docutils literal notranslate"><span class="pre">a.flags.writeable</span></code>). Short flag
names are only supported in dictionary access.</p>
<p>Only the WRITEBACKIFCOPY, UPDATEIFCOPY, WRITEABLE, and ALIGNED flags can be
changed by the user, via direct assignment to the attribute or dictionary
entry, or by calling <cite>ndarray.setflags</cite>.</p>
<p>The array flags cannot be set arbitrarily:</p>
<ul class="simple">
<li><p>UPDATEIFCOPY can only be set <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
<li><p>WRITEBACKIFCOPY can only be set <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
<li><p>ALIGNED can only be set <code class="docutils literal notranslate"><span class="pre">True</span></code> if the data is truly aligned.</p></li>
<li><p>WRITEABLE can only be set <code class="docutils literal notranslate"><span class="pre">True</span></code> if the array owns its own memory
or the ultimate owner of the memory exposes a writeable buffer
interface or is a string.</p></li>
</ul>
<p>Arrays can be both C-style and Fortran-style contiguous simultaneously.
This is clear for 1-dimensional arrays, but can also be true for higher
dimensional arrays.</p>
<p>Even for contiguous arrays a stride for a given dimension
<code class="docutils literal notranslate"><span class="pre">arr.strides[dim]</span></code> may be <em>arbitrary</em> if <code class="docutils literal notranslate"><span class="pre">arr.shape[dim]</span> <span class="pre">==</span> <span class="pre">1</span></code>
or the array has no elements.
It does <em>not</em> generally hold that <code class="docutils literal notranslate"><span class="pre">self.strides[-1]</span> <span class="pre">==</span> <span class="pre">self.itemsize</span></code>
for C-style contiguous arrays or <code class="docutils literal notranslate"><span class="pre">self.strides[0]</span> <span class="pre">==</span> <span class="pre">self.itemsize</span></code> for
Fortran-style contiguous arrays is true.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.flat">
<em class="property">property </em><code class="sig-name descname">flat</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.flat" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a flat iterator, or set a flattened version of self to value.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.flatten">
<code class="sig-name descname">flatten</code><span class="sig-paren">(</span><em class="sig-param">order='C'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.flatten" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a copy of the array collapsed into one dimension.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>order</strong> (<em>{'C'</em><em>, </em><em>'F'</em><em>, </em><em>'A'</em><em>, </em><em>'K'}</em><em>, </em><em>optional</em>) – ‘C’ means to flatten in row-major (C-style) order.
‘F’ means to flatten in column-major (Fortran-
style) order. ‘A’ means to flatten in column-major
order if <cite>a</cite> is Fortran <em>contiguous</em> in memory,
row-major order otherwise. ‘K’ means to flatten
<cite>a</cite> in the order the elements occur in memory.
The default is ‘C’.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>y</strong> – A copy of the input array, flattened to one dimension.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#fanc.matrix.RegionMatrix.ravel" title="fanc.matrix.RegionMatrix.ravel"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ravel()</span></code></a></dt><dd><p>Return a flattened array.</p>
</dd>
<dt><a class="reference internal" href="#fanc.matrix.RegionMatrix.flat" title="fanc.matrix.RegionMatrix.flat"><code class="xref py py-meth docutils literal notranslate"><span class="pre">flat()</span></code></a></dt><dd><p>A 1-D flat iterator over the array.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">flatten</span><span class="p">()</span>
<span class="go">array([1, 2, 3, 4])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">flatten</span><span class="p">(</span><span class="s1">&#39;F&#39;</span><span class="p">)</span>
<span class="go">array([1, 3, 2, 4])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.get_fill_value">
<code class="sig-name descname">get_fill_value</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.get_fill_value" title="Permalink to this definition">¶</a></dt>
<dd><p>The filling value of the masked array is a scalar. When setting, None
will set to a default based on the data type.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">dt</span> <span class="ow">in</span> <span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">complex128</span><span class="p">]:</span>
<span class="gp">... </span>    <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">dt</span><span class="p">)</span><span class="o">.</span><span class="n">get_fill_value</span><span class="p">()</span>
<span class="gp">...</span>
<span class="go">999999</span>
<span class="go">999999</span>
<span class="go">1e+20</span>
<span class="go">(1e+20+0j)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span> <span class="mf">1.</span><span class="p">],</span> <span class="n">fill_value</span><span class="o">=-</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill_value</span>
<span class="go">-inf</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill_value</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">pi</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill_value</span>
<span class="go">3.1415926535897931 # may vary</span>
</pre></div>
</div>
<p>Reset to default:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill_value</span> <span class="o">=</span> <span class="kc">None</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">fill_value</span>
<span class="go">1e+20</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.get_imag">
<code class="sig-name descname">get_imag</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.get_imag" title="Permalink to this definition">¶</a></dt>
<dd><p>The imaginary part of the masked array.</p>
<p>This property is a view on the imaginary part of this <cite>MaskedArray</cite>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#fanc.matrix.RegionMatrix.real" title="fanc.matrix.RegionMatrix.real"><code class="xref py py-meth docutils literal notranslate"><span class="pre">real()</span></code></a></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="o">+</span><span class="mf">1.</span><span class="n">j</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="n">j</span><span class="p">,</span> <span class="mf">3.45</span><span class="o">+</span><span class="mf">1.6</span><span class="n">j</span><span class="p">],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">imag</span>
<span class="go">masked_array(data=[1.0, --, 1.6],</span>
<span class="go">             mask=[False,  True, False],</span>
<span class="go">       fill_value=1e+20)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.get_real">
<code class="sig-name descname">get_real</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.get_real" title="Permalink to this definition">¶</a></dt>
<dd><p>The real part of the masked array.</p>
<p>This property is a view on the real part of this <cite>MaskedArray</cite>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#fanc.matrix.RegionMatrix.imag" title="fanc.matrix.RegionMatrix.imag"><code class="xref py py-meth docutils literal notranslate"><span class="pre">imag()</span></code></a></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="o">+</span><span class="mf">1.</span><span class="n">j</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="n">j</span><span class="p">,</span> <span class="mf">3.45</span><span class="o">+</span><span class="mf">1.6</span><span class="n">j</span><span class="p">],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">real</span>
<span class="go">masked_array(data=[1.0, --, 3.45],</span>
<span class="go">             mask=[False,  True, False],</span>
<span class="go">       fill_value=1e+20)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.getfield">
<code class="sig-name descname">getfield</code><span class="sig-paren">(</span><em class="sig-param">dtype</em>, <em class="sig-param">offset=0</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.getfield" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a field of the given array as a certain type.</p>
<p>A field is a view of the array data with a given data-type. The values in
the view are determined by the given type and the offset into the current
array in bytes. The offset needs to be such that the view dtype fits in the
array dtype; for example an array of dtype complex128 has 16-byte elements.
If taking a view with a 32-bit integer (4 bytes), the offset needs to be
between 0 and 12 bytes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dtype</strong> (<em>str</em><em> or </em><em>dtype</em>) – The data type of the view. The dtype size of the view can not be larger
than that of the array itself.</p></li>
<li><p><strong>offset</strong> (<em>int</em>) – Number of bytes to skip before beginning the element view.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">([</span><span class="mf">1.</span><span class="o">+</span><span class="mf">1.</span><span class="n">j</span><span class="p">]</span><span class="o">*</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">+</span> <span class="mf">4.</span><span class="n">j</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([[1.+1.j,  0.+0.j],</span>
<span class="go">       [0.+0.j,  2.+4.j]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">getfield</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<span class="go">array([[1.,  0.],</span>
<span class="go">       [0.,  2.]])</span>
</pre></div>
</div>
<p>By choosing an offset of 8 bytes we can select the complex part of the
array for our view:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">getfield</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="mi">8</span><span class="p">)</span>
<span class="go">array([[1.,  0.],</span>
<span class="go">       [0.,  4.]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.harden_mask">
<code class="sig-name descname">harden_mask</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.harden_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Force the mask to hard.</p>
<p>Whether the mask of a masked array is hard or soft is determined by
its <cite>hardmask</cite> property. <cite>harden_mask</cite> sets <cite>hardmask</cite> to True.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#fanc.matrix.RegionMatrix.hardmask" title="fanc.matrix.RegionMatrix.hardmask"><code class="xref py py-meth docutils literal notranslate"><span class="pre">hardmask()</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.hardmask">
<em class="property">property </em><code class="sig-name descname">hardmask</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.hardmask" title="Permalink to this definition">¶</a></dt>
<dd><p>Hardness of the mask</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.ids">
<code class="sig-name descname">ids</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.ids" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the addresses of the data and mask areas.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>None</strong> – </p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
<span class="go">(166670640, 166659832) # may vary</span>
</pre></div>
</div>
<p>If the array has no mask, the address of <cite>nomask</cite> is returned. This address
is typically not close to the data in memory:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ids</span><span class="p">()</span>
<span class="go">(166691080, 3083169284L) # may vary</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.imag">
<em class="property">property </em><code class="sig-name descname">imag</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.imag" title="Permalink to this definition">¶</a></dt>
<dd><p>The imaginary part of the masked array.</p>
<p>This property is a view on the imaginary part of this <cite>MaskedArray</cite>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#fanc.matrix.RegionMatrix.real" title="fanc.matrix.RegionMatrix.real"><code class="xref py py-obj docutils literal notranslate"><span class="pre">real</span></code></a></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="o">+</span><span class="mf">1.</span><span class="n">j</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="n">j</span><span class="p">,</span> <span class="mf">3.45</span><span class="o">+</span><span class="mf">1.6</span><span class="n">j</span><span class="p">],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">imag</span>
<span class="go">masked_array(data=[1.0, --, 1.6],</span>
<span class="go">             mask=[False,  True, False],</span>
<span class="go">       fill_value=1e+20)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.iscontiguous">
<code class="sig-name descname">iscontiguous</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.iscontiguous" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a boolean indicating whether the data is contiguous.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>None</strong> – </p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">iscontiguous</span><span class="p">()</span>
<span class="go">True</span>
</pre></div>
</div>
<p><cite>iscontiguous</cite> returns one of the flags of the masked array:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">flags</span>
<span class="go">  C_CONTIGUOUS : True</span>
<span class="go">  F_CONTIGUOUS : True</span>
<span class="go">  OWNDATA : False</span>
<span class="go">  WRITEABLE : True</span>
<span class="go">  ALIGNED : True</span>
<span class="go">  WRITEBACKIFCOPY : False</span>
<span class="go">  UPDATEIFCOPY : False</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.item">
<code class="sig-name descname">item</code><span class="sig-paren">(</span><em class="sig-param">*args</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.item" title="Permalink to this definition">¶</a></dt>
<dd><p>Copy an element of an array to a standard Python scalar and return it.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>*args</strong> (<em>Arguments</em><em> (</em><em>variable number and type</em><em>)</em>) – <ul class="simple">
<li><p>none: in this case, the method only works for arrays
with one element (<cite>a.size == 1</cite>), which element is
copied into a standard Python scalar object and returned.</p></li>
<li><p>int_type: this argument is interpreted as a flat index into
the array, specifying which element to copy and return.</p></li>
<li><p>tuple of int_types: functions as does a single int_type argument,
except that the argument is interpreted as an nd-index into the
array.</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>z</strong> – A copy of the specified element of the array as a suitable
Python scalar</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>Standard Python scalar object</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>When the data type of <cite>a</cite> is longdouble or clongdouble, item() returns
a scalar array object because there is no available Python scalar that
would not lose information. Void arrays return a buffer object for item(),
unless fields are defined, in which case a tuple is returned.</p>
<p><cite>item</cite> is very similar to a[args], except, instead of an array scalar,
a standard Python scalar is returned. This can be useful for speeding up
access to elements of the array and doing arithmetic on elements of the
array using Python’s optimized math.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([[2, 2, 6],</span>
<span class="go">       [1, 3, 6],</span>
<span class="go">       [1, 0, 1]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">item</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">item</span><span class="p">(</span><span class="mi">7</span><span class="p">)</span>
<span class="go">0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">item</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="go">2</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">item</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="go">1</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.itemset">
<code class="sig-name descname">itemset</code><span class="sig-paren">(</span><em class="sig-param">*args</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.itemset" title="Permalink to this definition">¶</a></dt>
<dd><p>Insert scalar into an array (scalar is cast to array’s dtype, if possible)</p>
<p>There must be at least 1 argument, and define the last argument
as <em>item</em>.  Then, <code class="docutils literal notranslate"><span class="pre">a.itemset(*args)</span></code> is equivalent to but faster
than <code class="docutils literal notranslate"><span class="pre">a[args]</span> <span class="pre">=</span> <span class="pre">item</span></code>.  The item should be a scalar value and <cite>args</cite>
must select a single item in the array <cite>a</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>*args</strong> (<em>Arguments</em>) – If one argument: a scalar, only used in case <cite>a</cite> is of size 1.
If two arguments: the last argument is the value to be set
and must be a scalar, the first argument specifies a single array
element location. It is either an int or a tuple.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Compared to indexing syntax, <cite>itemset</cite> provides some speed increase
for placing a scalar into a particular location in an <cite>ndarray</cite>,
if you must do this.  However, generally this is discouraged:
among other problems, it complicates the appearance of the code.
Also, when using <cite>itemset</cite> (and <cite>item</cite>) inside a loop, be sure
to assign the methods to a local variable to avoid the attribute
look-up at each loop iteration.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">seed</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([[2, 2, 6],</span>
<span class="go">       [1, 3, 6],</span>
<span class="go">       [1, 0, 1]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">itemset</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">itemset</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="mi">9</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([[2, 2, 6],</span>
<span class="go">       [1, 0, 6],</span>
<span class="go">       [1, 0, 9]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.itemsize">
<code class="sig-name descname">itemsize</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.itemsize" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of one array element in bytes.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">itemsize</span>
<span class="go">8</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">complex128</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">itemsize</span>
<span class="go">16</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.mask">
<em class="property">property </em><code class="sig-name descname">mask</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Current mask.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.max">
<code class="sig-name descname">max</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">fill_value=None</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.max" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the maximum along a given axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>{None</em><em>, </em><em>int}</em><em>, </em><em>optional</em>) – Axis along which to operate.  By default, <code class="docutils literal notranslate"><span class="pre">axis</span></code> is None and the
flattened input is used.</p></li>
<li><p><strong>out</strong> (<em>array_like</em><em>, </em><em>optional</em>) – Alternative output array in which to place the result.  Must
be of the same shape and buffer length as the expected output.</p></li>
<li><p><strong>fill_value</strong> (<em>{var}</em><em>, </em><em>optional</em>) – Value used to fill in the masked values.
If None, use the output of maximum_fill_value().</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>amax</strong> – New array holding the result.
If <code class="docutils literal notranslate"><span class="pre">out</span></code> was specified, <code class="docutils literal notranslate"><span class="pre">out</span></code> is returned.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>array_like</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">maximum_fill_value()</span></code></dt><dd><p>Returns the maximum filling value for a given datatype.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.mean">
<code class="sig-name descname">mean</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">dtype=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.mean" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the average of the array elements along given axis.</p>
<p>Masked entries are ignored, and result elements which are not
finite will be masked.</p>
<p>Refer to <cite>numpy.mean</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.mean()</span></code></dt><dd><p>corresponding function for ndarrays</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.mean()</span></code></dt><dd><p>Equivalent function</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.ma.average()</span></code></dt><dd><p>Weighted average.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">masked_array(data=[1, 2, --],</span>
<span class="go">             mask=[False, False,  True],</span>
<span class="go">       fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">mean</span><span class="p">()</span>
<span class="go">1.5</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.min">
<code class="sig-name descname">min</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">fill_value=None</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.min" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the minimum along a given axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>{None</em><em>, </em><em>int}</em><em>, </em><em>optional</em>) – Axis along which to operate.  By default, <code class="docutils literal notranslate"><span class="pre">axis</span></code> is None and the
flattened input is used.</p></li>
<li><p><strong>out</strong> (<em>array_like</em><em>, </em><em>optional</em>) – Alternative output array in which to place the result.  Must be of
the same shape and buffer length as the expected output.</p></li>
<li><p><strong>fill_value</strong> (<em>{var}</em><em>, </em><em>optional</em>) – Value used to fill in the masked values.
If None, use the output of <cite>minimum_fill_value</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>amin</strong> – New array holding the result.
If <code class="docutils literal notranslate"><span class="pre">out</span></code> was specified, <code class="docutils literal notranslate"><span class="pre">out</span></code> is returned.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>array_like</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimum_fill_value()</span></code></dt><dd><p>Returns the minimum filling value for a given datatype.</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.mini">
<code class="sig-name descname">mini</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.mini" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the array minimum along the specified axis.</p>
<div class="deprecated">
<p><span class="versionmodified deprecated">Deprecated since version 1.13.0: </span>This function is identical to both:</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">self.min(keepdims=True,</span> <span class="pre">axis=axis).squeeze(axis=axis)</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">np.ma.minimum.reduce(self,</span> <span class="pre">axis=axis)</span></code></p></li>
</ul>
</div></blockquote>
<p>Typically though, <code class="docutils literal notranslate"><span class="pre">self.min(axis=axis)</span></code> is sufficient.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>axis</strong> (<em>int</em><em>, </em><em>optional</em>) – The axis along which to find the minima. Default is None, in which case
the minimum value in the whole array is returned.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>min</strong> – If <cite>axis</cite> is None, the result is a scalar. Otherwise, if <cite>axis</cite> is
given and the array is at least 2-D, the result is a masked array with
dimension one smaller than the array on which <cite>mini</cite> is called.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>scalar or MaskedArray</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">6</span><span class="p">),</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span> <span class="p">,</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span> <span class="p">,</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[0, --],</span>
<span class="go">        [2, 3],</span>
<span class="go">        [4, --]],</span>
<span class="go">  mask=[[False,  True],</span>
<span class="go">        [False, False],</span>
<span class="go">        [False,  True]],</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">mini</span><span class="p">()</span>
<span class="go">masked_array(data=0,</span>
<span class="go">             mask=False,</span>
<span class="go">       fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">mini</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">masked_array(data=[0, 3],</span>
<span class="go">             mask=[False, False],</span>
<span class="go">       fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">mini</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">masked_array(data=[0, 2, 4],</span>
<span class="go">             mask=[False, False, False],</span>
<span class="go">       fill_value=999999)</span>
</pre></div>
</div>
<p>There is a small difference between <cite>mini</cite> and <cite>min</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">mini</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">masked_array(data=3,</span>
<span class="go">             mask=False,</span>
<span class="go">       fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">3</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.nbytes">
<code class="sig-name descname">nbytes</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.nbytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Total bytes consumed by the elements of the array.</p>
<p class="rubric">Notes</p>
<p>Does not include memory consumed by non-element attributes of the
array object.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">3</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">2</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">complex128</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">nbytes</span>
<span class="go">480</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span> <span class="o">*</span> <span class="n">x</span><span class="o">.</span><span class="n">itemsize</span>
<span class="go">480</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.ndim">
<code class="sig-name descname">ndim</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.ndim" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of array dimensions.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ndim</span>
<span class="go">1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">ndim</span>
<span class="go">3</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.newbyteorder">
<code class="sig-name descname">newbyteorder</code><span class="sig-paren">(</span><em class="sig-param">new_order='S'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.newbyteorder" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the array with the same data viewed with a different byte order.</p>
<p>Equivalent to:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">arr</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">arr</span><span class="o">.</span><span class="n">dtype</span><span class="o">.</span><span class="n">newbytorder</span><span class="p">(</span><span class="n">new_order</span><span class="p">))</span>
</pre></div>
</div>
<p>Changes are also made in all fields and sub-arrays of the array data
type.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>new_order</strong> (<em>string</em><em>, </em><em>optional</em>) – <p>Byte order to force; a value from the byte order specifications
below. <cite>new_order</cite> codes can be any of:</p>
<ul class="simple">
<li><p>’S’ - swap dtype from current to opposite endian</p></li>
<li><p>{‘&lt;’, ‘L’} - little endian</p></li>
<li><p>{‘&gt;’, ‘B’} - big endian</p></li>
<li><p>{‘=’, ‘N’} - native order</p></li>
<li><p>{‘|’, ‘I’} - ignore (no change to byte order)</p></li>
</ul>
<p>The default value (‘S’) results in swapping the current
byte order. The code does a case-insensitive check on the first
letter of <cite>new_order</cite> for the alternatives above.  For example,
any of ‘B’ or ‘b’ or ‘biggish’ are valid to specify big-endian.</p>
</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>new_arr</strong> – New array object with the dtype reflecting given change to the
byte order.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.nonzero">
<code class="sig-name descname">nonzero</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.nonzero" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the indices of unmasked elements that are not zero.</p>
<p>Returns a tuple of arrays, one for each dimension, containing the
indices of the non-zero elements in that dimension. The corresponding
non-zero values can be obtained with:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">a</span><span class="p">[</span><span class="n">a</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()]</span>
</pre></div>
</div>
<p>To group the indices by element, rather than dimension, use
instead:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">a</span><span class="o">.</span><span class="n">nonzero</span><span class="p">())</span>
</pre></div>
</div>
<p>The result of this is always a 2d array, with a row for each non-zero
element.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>None</strong> – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>tuple_of_arrays</strong> – Indices of elements that are non-zero.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.nonzero()</span></code></dt><dd><p>Function operating on ndarrays.</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">flatnonzero()</span></code></dt><dd><p>Return indices that are non-zero in the flattened version of the input array.</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.nonzero()</span></code></dt><dd><p>Equivalent ndarray method.</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">count_nonzero()</span></code></dt><dd><p>Counts the number of non-zero elements in the input array.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="kn">import</span> <span class="nn">numpy.ma</span> <span class="k">as</span> <span class="nn">ma</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="mi">3</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[1., 0., 0.],</span>
<span class="go">        [0., 1., 0.],</span>
<span class="go">        [0., 0., 1.]],</span>
<span class="go">  mask=False,</span>
<span class="go">  fill_value=1e+20)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()</span>
<span class="go">(array([0, 1, 2]), array([0, 1, 2]))</span>
</pre></div>
</div>
<p>Masked elements are ignored.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">masked</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[1.0, 0.0, 0.0],</span>
<span class="go">        [0.0, --, 0.0],</span>
<span class="go">        [0.0, 0.0, 1.0]],</span>
<span class="go">  mask=[[False, False, False],</span>
<span class="go">        [False,  True, False],</span>
<span class="go">        [False, False, False]],</span>
<span class="go">  fill_value=1e+20)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()</span>
<span class="go">(array([0, 2]), array([0, 2]))</span>
</pre></div>
</div>
<p>Indices can also be grouped by element.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">nonzero</span><span class="p">())</span>
<span class="go">array([[0, 0],</span>
<span class="go">       [2, 2]])</span>
</pre></div>
</div>
<p>A common use for <code class="docutils literal notranslate"><span class="pre">nonzero</span></code> is to find the indices of an array, where
a condition is True.  Given an array <cite>a</cite>, the condition <cite>a</cite> &gt; 3 is a
boolean array and since False is interpreted as 0, ma.nonzero(a &gt; 3)
yields the indices of the <cite>a</cite> where the condition is true.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">],[</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">&gt;</span> <span class="mi">3</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[False, False, False],</span>
<span class="go">        [ True,  True,  True],</span>
<span class="go">        [ True,  True,  True]],</span>
<span class="go">  mask=False,</span>
<span class="go">  fill_value=True)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">ma</span><span class="o">.</span><span class="n">nonzero</span><span class="p">(</span><span class="n">a</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">(array([1, 1, 1, 2, 2, 2]), array([0, 1, 2, 0, 1, 2]))</span>
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">nonzero</span></code> method of the condition array can also be called.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="p">(</span><span class="n">a</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">)</span><span class="o">.</span><span class="n">nonzero</span><span class="p">()</span>
<span class="go">(array([1, 1, 1, 2, 2, 2]), array([0, 1, 2, 0, 1, 2]))</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.partition">
<code class="sig-name descname">partition</code><span class="sig-paren">(</span><em class="sig-param">kth</em>, <em class="sig-param">axis=-1</em>, <em class="sig-param">kind='introselect'</em>, <em class="sig-param">order=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.partition" title="Permalink to this definition">¶</a></dt>
<dd><p>Rearranges the elements in the array in such a way that the value of the
element in kth position is in the position it would be in a sorted array.
All elements smaller than the kth element are moved before this element and
all equal or greater are moved behind it. The ordering of the elements in
the two partitions is undefined.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.8.0.</span></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>kth</strong> (<em>int</em><em> or </em><em>sequence of ints</em>) – Element index to partition by. The kth element value will be in its
final sorted position and all smaller elements will be moved before it
and all equal or greater elements behind it.
The order of all elements in the partitions is undefined.
If provided with a sequence of kth it will partition all elements
indexed by kth of them into their sorted position at once.</p></li>
<li><p><strong>axis</strong> (<em>int</em><em>, </em><em>optional</em>) – Axis along which to sort. Default is -1, which means sort along the
last axis.</p></li>
<li><p><strong>kind</strong> (<em>{'introselect'}</em><em>, </em><em>optional</em>) – Selection algorithm. Default is ‘introselect’.</p></li>
<li><p><strong>order</strong> (<em>str</em><em> or </em><em>list of str</em><em>, </em><em>optional</em>) – When <cite>a</cite> is an array with fields defined, this argument specifies
which fields to compare first, second, etc. A single field can
be specified as a string, and not all fields need to be specified,
but unspecified fields will still be used, in the order in which
they come up in the dtype, to break ties.</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.partition()</span></code></dt><dd><p>Return a parititioned copy of an array.</p>
</dd>
<dt><a class="reference internal" href="#fanc.matrix.RegionMatrix.argpartition" title="fanc.matrix.RegionMatrix.argpartition"><code class="xref py py-meth docutils literal notranslate"><span class="pre">argpartition()</span></code></a></dt><dd><p>Indirect partition.</p>
</dd>
<dt><a class="reference internal" href="#fanc.matrix.RegionMatrix.sort" title="fanc.matrix.RegionMatrix.sort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sort()</span></code></a></dt><dd><p>Full sort.</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>See <code class="docutils literal notranslate"><span class="pre">np.partition</span></code> for notes on the different algorithms.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">partition</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">array([2, 1, 3, 4])</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">partition</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">array([1, 2, 3, 4])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.prod">
<code class="sig-name descname">prod</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">dtype=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.prod" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the product of the array elements over the given axis.</p>
<p>Masked elements are set to 1 internally for computation.</p>
<p>Refer to <cite>numpy.prod</cite> for full documentation.</p>
<p class="rubric">Notes</p>
<p>Arithmetic is modular when using integer types, and no error is raised
on overflow.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.prod()</span></code></dt><dd><p>corresponding function for ndarrays</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.prod()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.product">
<code class="sig-name descname">product</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">dtype=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.product" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the product of the array elements over the given axis.</p>
<p>Masked elements are set to 1 internally for computation.</p>
<p>Refer to <cite>numpy.prod</cite> for full documentation.</p>
<p class="rubric">Notes</p>
<p>Arithmetic is modular when using integer types, and no error is raised
on overflow.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.prod()</span></code></dt><dd><p>corresponding function for ndarrays</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.prod()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.ptp">
<code class="sig-name descname">ptp</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">fill_value=None</em>, <em class="sig-param">keepdims=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.ptp" title="Permalink to this definition">¶</a></dt>
<dd><p>Return (maximum - minimum) along the given dimension
(i.e. peak-to-peak value).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>axis</strong> (<em>{None</em><em>, </em><em>int}</em><em>, </em><em>optional</em>) – Axis along which to find the peaks.  If None (default) the
flattened array is used.</p></li>
<li><p><strong>out</strong> (<em>{None</em><em>, </em><em>array_like}</em><em>, </em><em>optional</em>) – Alternative output array in which to place the result. It must
have the same shape and buffer length as the expected output
but the type will be cast if necessary.</p></li>
<li><p><strong>fill_value</strong> (<em>{var}</em><em>, </em><em>optional</em>) – Value used to fill in the masked values.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>ptp</strong> – A new array holding the result, unless <code class="docutils literal notranslate"><span class="pre">out</span></code> was
specified, in which case a reference to <code class="docutils literal notranslate"><span class="pre">out</span></code> is returned.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.put">
<code class="sig-name descname">put</code><span class="sig-paren">(</span><em class="sig-param">indices</em>, <em class="sig-param">values</em>, <em class="sig-param">mode='raise'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.put" title="Permalink to this definition">¶</a></dt>
<dd><p>Set storage-indexed locations to corresponding values.</p>
<p>Sets self._data.flat[n] = values[n] for each n in indices.
If <cite>values</cite> is shorter than <cite>indices</cite> then it will repeat.
If <cite>values</cite> has some masked values, the initial mask is updated
in consequence, else the corresponding values are unmasked.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>indices</strong> (<em>1-D array_like</em>) – Target indices, interpreted as integers.</p></li>
<li><p><strong>values</strong> (<em>array_like</em>) – Values to place in self._data copy at target indices.</p></li>
<li><p><strong>mode</strong> (<em>{'raise'</em><em>, </em><em>'wrap'</em><em>, </em><em>'clip'}</em><em>, </em><em>optional</em>) – Specifies how out-of-bounds indices will behave.
‘raise’ : raise an error.
‘wrap’ : wrap around.
‘clip’ : clip to the range.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p><cite>values</cite> can be a scalar or length 1 array.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">],[</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">]],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[1, --, 3],</span>
<span class="go">        [--, 5, --],</span>
<span class="go">        [7, --, 9]],</span>
<span class="go">  mask=[[False,  True, False],</span>
<span class="go">        [ True, False,  True],</span>
<span class="go">        [False,  True, False]],</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">put</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="mi">4</span><span class="p">,</span><span class="mi">8</span><span class="p">],[</span><span class="mi">10</span><span class="p">,</span><span class="mi">20</span><span class="p">,</span><span class="mi">30</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[10, --, 3],</span>
<span class="go">        [--, 20, --],</span>
<span class="go">        [7, --, 30]],</span>
<span class="go">  mask=[[False,  True, False],</span>
<span class="go">        [ True, False,  True],</span>
<span class="go">        [False,  True, False]],</span>
<span class="go">  fill_value=999999)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">put</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">999</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[10, --, 3],</span>
<span class="go">        [--, 999, --],</span>
<span class="go">        [7, --, 30]],</span>
<span class="go">  mask=[[False,  True, False],</span>
<span class="go">        [ True, False,  True],</span>
<span class="go">        [False,  True, False]],</span>
<span class="go">  fill_value=999999)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.ravel">
<code class="sig-name descname">ravel</code><span class="sig-paren">(</span><em class="sig-param">order='C'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.ravel" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a 1D version of self, as a view.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>order</strong> (<em>{'C'</em><em>, </em><em>'F'</em><em>, </em><em>'A'</em><em>, </em><em>'K'}</em><em>, </em><em>optional</em>) – The elements of <cite>a</cite> are read using this index order. ‘C’ means to
index the elements in C-like order, with the last axis index
changing fastest, back to the first axis index changing slowest.
‘F’ means to index the elements in Fortran-like index order, with
the first index changing fastest, and the last index changing
slowest. Note that the ‘C’ and ‘F’ options take no account of the
memory layout of the underlying array, and only refer to the order
of axis indexing.  ‘A’ means to read the elements in Fortran-like
index order if <cite>m</cite> is Fortran <em>contiguous</em> in memory, C-like order
otherwise.  ‘K’ means to read the elements in the order they occur
in memory, except for reversing the data when strides are negative.
By default, ‘C’ index order is used.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Output view is of shape <code class="docutils literal notranslate"><span class="pre">(self.size,)</span></code> (or
<code class="docutils literal notranslate"><span class="pre">(np.ma.product(self.shape),)</span></code>).</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>MaskedArray</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">],[</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">]],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[1, --, 3],</span>
<span class="go">        [--, 5, --],</span>
<span class="go">        [7, --, 9]],</span>
<span class="go">  mask=[[False,  True, False],</span>
<span class="go">        [ True, False,  True],</span>
<span class="go">        [False,  True, False]],</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">ravel</span><span class="p">()</span>
<span class="go">masked_array(data=[1, --, 3, --, 5, --, 7, --, 9],</span>
<span class="go">             mask=[False,  True, False,  True, False,  True, False,  True,</span>
<span class="go">                   False],</span>
<span class="go">       fill_value=999999)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.real">
<em class="property">property </em><code class="sig-name descname">real</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.real" title="Permalink to this definition">¶</a></dt>
<dd><p>The real part of the masked array.</p>
<p>This property is a view on the real part of this <cite>MaskedArray</cite>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#fanc.matrix.RegionMatrix.imag" title="fanc.matrix.RegionMatrix.imag"><code class="xref py py-obj docutils literal notranslate"><span class="pre">imag</span></code></a></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="o">+</span><span class="mf">1.</span><span class="n">j</span><span class="p">,</span> <span class="o">-</span><span class="mi">2</span><span class="n">j</span><span class="p">,</span> <span class="mf">3.45</span><span class="o">+</span><span class="mf">1.6</span><span class="n">j</span><span class="p">],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">real</span>
<span class="go">masked_array(data=[1.0, --, 3.45],</span>
<span class="go">             mask=[False,  True, False],</span>
<span class="go">       fill_value=1e+20)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.recordmask">
<em class="property">property </em><code class="sig-name descname">recordmask</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.recordmask" title="Permalink to this definition">¶</a></dt>
<dd><p>Get or set the mask of the array if it has no named fields. For
structured arrays, returns a ndarray of booleans where entries are
<code class="docutils literal notranslate"><span class="pre">True</span></code> if <strong>all</strong> the fields are masked, <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">),</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="p">)],</span>
<span class="gp">... </span>        <span class="n">mask</span><span class="o">=</span><span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">),</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)],</span>
<span class="gp">... </span>       <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="nb">int</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">recordmask</span>
<span class="go">array([False, False,  True, False, False])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.repeat">
<code class="sig-name descname">repeat</code><span class="sig-paren">(</span><em class="sig-param">repeats</em>, <em class="sig-param">axis=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.repeat" title="Permalink to this definition">¶</a></dt>
<dd><p>Repeat elements of an array.</p>
<p>Refer to <cite>numpy.repeat</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.repeat()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.reshape">
<code class="sig-name descname">reshape</code><span class="sig-paren">(</span><em class="sig-param">*s</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.reshape" title="Permalink to this definition">¶</a></dt>
<dd><p>Give a new shape to the array without changing its data.</p>
<p>Returns a masked array containing the same data, but with a new shape.
The result is a view on the original array; if this is not possible, a
ValueError is raised.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>shape</strong> (<em>int</em><em> or </em><em>tuple of ints</em>) – The new shape should be compatible with the original shape. If an
integer is supplied, then the result will be a 1-D array of that
length.</p></li>
<li><p><strong>order</strong> (<em>{'C'</em><em>, </em><em>'F'}</em><em>, </em><em>optional</em>) – Determines whether the array data should be viewed as in C
(row-major) or FORTRAN (column-major) order.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>reshaped_array</strong> – A new view on the array.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>array</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><a class="reference internal" href="#fanc.matrix.RegionMatrix.reshape" title="fanc.matrix.RegionMatrix.reshape"><code class="xref py py-meth docutils literal notranslate"><span class="pre">reshape()</span></code></a></dt><dd><p>Equivalent function in the masked array module.</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.ndarray.reshape()</span></code></dt><dd><p>Equivalent method on ndarray object.</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.reshape()</span></code></dt><dd><p>Equivalent function in the NumPy module.</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>The reshaping operation cannot guarantee that a copy will not be made,
to modify the shape in place, use <code class="docutils literal notranslate"><span class="pre">a.shape</span> <span class="pre">=</span> <span class="pre">s</span></code></p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">],[</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">]],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[--, 2],</span>
<span class="go">        [3, --]],</span>
<span class="go">  mask=[[ True, False],</span>
<span class="go">        [False,  True]],</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">reshape</span><span class="p">((</span><span class="mi">4</span><span class="p">,</span><span class="mi">1</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[--],</span>
<span class="go">        [2],</span>
<span class="go">        [3],</span>
<span class="go">        [--]],</span>
<span class="go">  mask=[[ True],</span>
<span class="go">        [False],</span>
<span class="go">        [False],</span>
<span class="go">        [ True]],</span>
<span class="go">  fill_value=999999)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.resize">
<code class="sig-name descname">resize</code><span class="sig-paren">(</span><em class="sig-param">newshape</em>, <em class="sig-param">refcheck=True</em>, <em class="sig-param">order=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.resize" title="Permalink to this definition">¶</a></dt>
<dd><div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This method does nothing, except raise a ValueError exception. A
masked array does not own its data and therefore cannot safely be
resized in place. Use the <cite>numpy.ma.resize</cite> function instead.</p>
</div>
<p>This method is difficult to implement safely and may be deprecated in
future releases of NumPy.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.round">
<code class="sig-name descname">round</code><span class="sig-paren">(</span><em class="sig-param">decimals=0</em>, <em class="sig-param">out=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.round" title="Permalink to this definition">¶</a></dt>
<dd><p>Return each element rounded to the given number of decimals.</p>
<p>Refer to <cite>numpy.around</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.around()</span></code></dt><dd><p>corresponding function for ndarrays</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.around()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.searchsorted">
<code class="sig-name descname">searchsorted</code><span class="sig-paren">(</span><em class="sig-param">v</em>, <em class="sig-param">side='left'</em>, <em class="sig-param">sorter=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.searchsorted" title="Permalink to this definition">¶</a></dt>
<dd><p>Find indices where elements of v should be inserted in a to maintain order.</p>
<p>For full documentation, see <cite>numpy.searchsorted</cite></p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.searchsorted()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.setfield">
<code class="sig-name descname">setfield</code><span class="sig-paren">(</span><em class="sig-param">val</em>, <em class="sig-param">dtype</em>, <em class="sig-param">offset=0</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.setfield" title="Permalink to this definition">¶</a></dt>
<dd><p>Put a value into a specified place in a field defined by a data-type.</p>
<p>Place <cite>val</cite> into <cite>a</cite>’s field defined by <cite>dtype</cite> and beginning <cite>offset</cite>
bytes into the field.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>val</strong> (<em>object</em>) – Value to be placed in field.</p></li>
<li><p><strong>dtype</strong> (<em>dtype object</em>) – Data-type of the field in which to place <cite>val</cite>.</p></li>
<li><p><strong>offset</strong> (<em>int</em><em>, </em><em>optional</em>) – The number of bytes into the field at which to place <cite>val</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#fanc.matrix.RegionMatrix.getfield" title="fanc.matrix.RegionMatrix.getfield"><code class="xref py py-meth docutils literal notranslate"><span class="pre">getfield()</span></code></a></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">getfield</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<span class="go">array([[1.,  0.,  0.],</span>
<span class="go">       [0.,  1.,  0.],</span>
<span class="go">       [0.,  0.,  1.]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">setfield</span><span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">getfield</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="go">array([[3, 3, 3],</span>
<span class="go">       [3, 3, 3],</span>
<span class="go">       [3, 3, 3]], dtype=int32)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([[1.0e+000, 1.5e-323, 1.5e-323],</span>
<span class="go">       [1.5e-323, 1.0e+000, 1.5e-323],</span>
<span class="go">       [1.5e-323, 1.5e-323, 1.0e+000]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">setfield</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">eye</span><span class="p">(</span><span class="mi">3</span><span class="p">),</span> <span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([[1.,  0.,  0.],</span>
<span class="go">       [0.,  1.,  0.],</span>
<span class="go">       [0.,  0.,  1.]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.setflags">
<code class="sig-name descname">setflags</code><span class="sig-paren">(</span><em class="sig-param">write=None</em>, <em class="sig-param">align=None</em>, <em class="sig-param">uic=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.setflags" title="Permalink to this definition">¶</a></dt>
<dd><p>Set array flags WRITEABLE, ALIGNED, (WRITEBACKIFCOPY and UPDATEIFCOPY),
respectively.</p>
<p>These Boolean-valued flags affect how numpy interprets the memory
area used by <cite>a</cite> (see Notes below). The ALIGNED flag can only
be set to True if the data is actually aligned according to the type.
The WRITEBACKIFCOPY and (deprecated) UPDATEIFCOPY flags can never be set
to True. The flag WRITEABLE can only be set to True if the array owns its
own memory, or the ultimate owner of the memory exposes a writeable buffer
interface, or is a string. (The exception for string is made so that
unpickling can be done without copying memory.)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>write</strong> (<em>bool</em><em>, </em><em>optional</em>) – Describes whether or not <cite>a</cite> can be written to.</p></li>
<li><p><strong>align</strong> (<em>bool</em><em>, </em><em>optional</em>) – Describes whether or not <cite>a</cite> is aligned properly for its type.</p></li>
<li><p><strong>uic</strong> (<em>bool</em><em>, </em><em>optional</em>) – Describes whether or not <cite>a</cite> is a copy of another “base” array.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>Array flags provide information about how the memory area used
for the array is to be interpreted. There are 7 Boolean flags
in use, only four of which can be changed by the user:
WRITEBACKIFCOPY, UPDATEIFCOPY, WRITEABLE, and ALIGNED.</p>
<p>WRITEABLE (W) the data area can be written to;</p>
<p>ALIGNED (A) the data and strides are aligned appropriately for the hardware
(as determined by the compiler);</p>
<p>UPDATEIFCOPY (U) (deprecated), replaced by WRITEBACKIFCOPY;</p>
<p>WRITEBACKIFCOPY (X) this array is a copy of some other array (referenced
by .base). When the C-API function PyArray_ResolveWritebackIfCopy is
called, the base array will be updated with the contents of this array.</p>
<p>All flags can be accessed using the single (upper case) letter as well
as the full name.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">7</span><span class="p">],</span>
<span class="gp">... </span>              <span class="p">[</span><span class="mi">2</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">],</span>
<span class="gp">... </span>              <span class="p">[</span><span class="mi">8</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">9</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span>
<span class="go">array([[3, 1, 7],</span>
<span class="go">       [2, 0, 0],</span>
<span class="go">       [8, 5, 9]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">flags</span>
<span class="go">  C_CONTIGUOUS : True</span>
<span class="go">  F_CONTIGUOUS : False</span>
<span class="go">  OWNDATA : True</span>
<span class="go">  WRITEABLE : True</span>
<span class="go">  ALIGNED : True</span>
<span class="go">  WRITEBACKIFCOPY : False</span>
<span class="go">  UPDATEIFCOPY : False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">setflags</span><span class="p">(</span><span class="n">write</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">align</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">flags</span>
<span class="go">  C_CONTIGUOUS : True</span>
<span class="go">  F_CONTIGUOUS : False</span>
<span class="go">  OWNDATA : True</span>
<span class="go">  WRITEABLE : False</span>
<span class="go">  ALIGNED : False</span>
<span class="go">  WRITEBACKIFCOPY : False</span>
<span class="go">  UPDATEIFCOPY : False</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">setflags</span><span class="p">(</span><span class="n">uic</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
<span class="gr">ValueError</span>: <span class="n">cannot set WRITEBACKIFCOPY flag to True</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.shape">
<em class="property">property </em><code class="sig-name descname">shape</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.shape" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple of array dimensions.</p>
<p>The shape property is usually used to get the current shape of an array,
but may also be used to reshape the array in-place by assigning a tuple of
array dimensions to it.  As with <cite>numpy.reshape</cite>, one of the new shape
dimensions can be -1, in which case its value is inferred from the size of
the array and the remaining dimensions. Reshaping an array in-place will
fail if a copy is required.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(4,)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">shape</span>
<span class="go">(2, 3, 4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">shape</span> <span class="o">=</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span>
<span class="go">array([[ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],</span>
<span class="go">       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.],</span>
<span class="go">       [ 0.,  0.,  0.,  0.,  0.,  0.,  0.,  0.]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">shape</span> <span class="o">=</span> <span class="p">(</span><span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
<span class="gr">ValueError</span>: <span class="n">total size of new array must be unchanged</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">4</span><span class="p">,</span><span class="mi">2</span><span class="p">))[::</span><span class="mi">2</span><span class="p">]</span><span class="o">.</span><span class="n">shape</span> <span class="o">=</span> <span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,)</span>
<span class="gt">Traceback (most recent call last):</span>
  File <span class="nb">&quot;&lt;stdin&gt;&quot;</span>, line <span class="m">1</span>, in <span class="n">&lt;module&gt;</span>
<span class="gr">AttributeError</span>: <span class="n">incompatible shape for a non-contiguous array</span>
</pre></div>
</div>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">numpy.reshape</span></code></dt><dd><p>similar function</p>
</dd>
<dt><code class="xref py py-obj docutils literal notranslate"><span class="pre">ndarray.reshape</span></code></dt><dd><p>similar method</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.sharedmask">
<em class="property">property </em><code class="sig-name descname">sharedmask</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.sharedmask" title="Permalink to this definition">¶</a></dt>
<dd><p>Share status of the mask (read-only).</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.shrink_mask">
<code class="sig-name descname">shrink_mask</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.shrink_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Reduce a mask to nomask when possible.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>None</strong> – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>None</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span> <span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">mask</span>
<span class="go">array([[False, False],</span>
<span class="go">       [False, False]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">shrink_mask</span><span class="p">()</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[1, 2],</span>
<span class="go">        [3, 4]],</span>
<span class="go">  mask=False,</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">mask</span>
<span class="go">False</span>
</pre></div>
</div>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.size">
<code class="sig-name descname">size</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.size" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of elements in the array.</p>
<p>Equal to <code class="docutils literal notranslate"><span class="pre">np.prod(a.shape)</span></code>, i.e., the product of the array’s
dimensions.</p>
<p class="rubric">Notes</p>
<p><cite>a.size</cite> returns a standard arbitrary precision Python integer. This
may not be the case with other methods of obtaining the same value
(like the suggested <code class="docutils literal notranslate"><span class="pre">np.prod(a.shape)</span></code>, which returns an instance
of <code class="docutils literal notranslate"><span class="pre">np.int_</span></code>), and may be relevant if the value is used further in
calculations that may overflow a fixed size integer type.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">3</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">2</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">complex128</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">size</span>
<span class="go">30</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">prod</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="go">30</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.soften_mask">
<code class="sig-name descname">soften_mask</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.soften_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Force the mask to soft.</p>
<p>Whether the mask of a masked array is hard or soft is determined by
its <cite>hardmask</cite> property. <cite>soften_mask</cite> sets <cite>hardmask</cite> to False.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#fanc.matrix.RegionMatrix.hardmask" title="fanc.matrix.RegionMatrix.hardmask"><code class="xref py py-meth docutils literal notranslate"><span class="pre">hardmask()</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.sort">
<code class="sig-name descname">sort</code><span class="sig-paren">(</span><em class="sig-param">axis=-1</em>, <em class="sig-param">kind=None</em>, <em class="sig-param">order=None</em>, <em class="sig-param">endwith=True</em>, <em class="sig-param">fill_value=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.sort" title="Permalink to this definition">¶</a></dt>
<dd><p>Sort the array, in-place</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<em>array_like</em>) – Array to be sorted.</p></li>
<li><p><strong>axis</strong> (<em>int</em><em>, </em><em>optional</em>) – Axis along which to sort. If None, the array is flattened before
sorting. The default is -1, which sorts along the last axis.</p></li>
<li><p><strong>kind</strong> (<em>{'quicksort'</em><em>, </em><em>'mergesort'</em><em>, </em><em>'heapsort'</em><em>, </em><em>'stable'}</em><em>, </em><em>optional</em>) – The sorting algorithm used.</p></li>
<li><p><strong>order</strong> (<em>list</em><em>, </em><em>optional</em>) – When <cite>a</cite> is a structured array, this argument specifies which fields
to compare first, second, and so on.  This list does not need to
include all of the fields.</p></li>
<li><p><strong>endwith</strong> (<em>{True</em><em>, </em><em>False}</em><em>, </em><em>optional</em>) – Whether missing values (if any) should be treated as the largest values
(True) or the smallest values (False)
When the array contains unmasked values sorting at the same extremes of the
datatype, the ordering of these values and the masked values is
undefined.</p></li>
<li><p><strong>fill_value</strong> (<em>{var}</em><em>, </em><em>optional</em>) – Value used internally for the masked values.
If <code class="docutils literal notranslate"><span class="pre">fill_value</span></code> is not None, it supersedes <code class="docutils literal notranslate"><span class="pre">endwith</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>sorted_array</strong> – Array of the same type and shape as <cite>a</cite>.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.sort()</span></code></dt><dd><p>Method to sort an array in-place.</p>
</dd>
<dt><a class="reference internal" href="#fanc.matrix.RegionMatrix.argsort" title="fanc.matrix.RegionMatrix.argsort"><code class="xref py py-meth docutils literal notranslate"><span class="pre">argsort()</span></code></a></dt><dd><p>Indirect sort.</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">lexsort()</span></code></dt><dd><p>Indirect stable sort on multiple keys.</p>
</dd>
<dt><a class="reference internal" href="#fanc.matrix.RegionMatrix.searchsorted" title="fanc.matrix.RegionMatrix.searchsorted"><code class="xref py py-meth docutils literal notranslate"><span class="pre">searchsorted()</span></code></a></dt><dd><p>Find elements in a sorted array.</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>See <code class="docutils literal notranslate"><span class="pre">sort</span></code> for notes on the different sorting algorithms.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span><span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Default</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">masked_array(data=[1, 3, 5, --, --],</span>
<span class="go">             mask=[False, False, False,  True,  True],</span>
<span class="go">       fill_value=999999)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span><span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># Put missing values in the front</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">endwith</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">masked_array(data=[--, --, 1, 3, 5],</span>
<span class="go">             mask=[ True,  True, False, False, False],</span>
<span class="go">       fill_value=999999)</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span><span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1"># fill_value takes over endwith</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">endwith</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">fill_value</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">masked_array(data=[1, --, --, 3, 5],</span>
<span class="go">             mask=[False,  True,  True, False, False],</span>
<span class="go">       fill_value=999999)</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.squeeze">
<code class="sig-name descname">squeeze</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.squeeze" title="Permalink to this definition">¶</a></dt>
<dd><p>Remove single-dimensional entries from the shape of <cite>a</cite>.</p>
<p>Refer to <cite>numpy.squeeze</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.squeeze()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.std">
<code class="sig-name descname">std</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">dtype=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">ddof=0</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.std" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the standard deviation of the array elements along given axis.</p>
<p>Masked entries are ignored.</p>
<p>Refer to <cite>numpy.std</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.std()</span></code></dt><dd><p>corresponding function for ndarrays</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.std()</span></code></dt><dd><p>Equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="attribute">
<dt id="fanc.matrix.RegionMatrix.strides">
<code class="sig-name descname">strides</code><a class="headerlink" href="#fanc.matrix.RegionMatrix.strides" title="Permalink to this definition">¶</a></dt>
<dd><p>Tuple of bytes to step in each dimension when traversing an array.</p>
<p>The byte offset of element <code class="docutils literal notranslate"><span class="pre">(i[0],</span> <span class="pre">i[1],</span> <span class="pre">...,</span> <span class="pre">i[n])</span></code> in an array <cite>a</cite>
is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">offset</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="o">*</span> <span class="n">a</span><span class="o">.</span><span class="n">strides</span><span class="p">)</span>
</pre></div>
</div>
<p>A more detailed explanation of strides can be found in the
“ndarray.rst” file in the NumPy reference guide.</p>
<p class="rubric">Notes</p>
<p>Imagine an array of 32-bit integers (each 4 bytes):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">],</span>
              <span class="p">[</span><span class="mi">5</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">]],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int32</span><span class="p">)</span>
</pre></div>
</div>
<p>This array is stored in memory as 40 bytes, one after the other
(known as a contiguous block of memory).  The strides of an array tell
us how many bytes we have to skip in memory to move to the next position
along a certain axis.  For example, we have to skip 4 bytes (1 value) to
move to the next column, but 20 bytes (5 values) to get to the same
position in the next row.  As such, the strides for the array <cite>x</cite> will be
<code class="docutils literal notranslate"><span class="pre">(20,</span> <span class="pre">4)</span></code>.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-obj docutils literal notranslate"><span class="pre">numpy.lib.stride_tricks.as_strided</span></code></p>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">2</span><span class="o">*</span><span class="mi">3</span><span class="o">*</span><span class="mi">4</span><span class="p">),</span> <span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span>
<span class="go">array([[[ 0,  1,  2,  3],</span>
<span class="go">        [ 4,  5,  6,  7],</span>
<span class="go">        [ 8,  9, 10, 11]],</span>
<span class="go">       [[12, 13, 14, 15],</span>
<span class="go">        [16, 17, 18, 19],</span>
<span class="go">        [20, 21, 22, 23]]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">strides</span>
<span class="go">(48, 16, 4)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span>
<span class="go">17</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">offset</span><span class="o">=</span><span class="nb">sum</span><span class="p">(</span><span class="n">y</span><span class="o">.</span><span class="n">strides</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">)))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">offset</span><span class="o">/</span><span class="n">y</span><span class="o">.</span><span class="n">itemsize</span>
<span class="go">17</span>
</pre></div>
</div>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">5</span><span class="o">*</span><span class="mi">6</span><span class="o">*</span><span class="mi">7</span><span class="o">*</span><span class="mi">8</span><span class="p">),</span> <span class="p">(</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">,</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">))</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">,</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">strides</span>
<span class="go">(32, 4, 224, 1344)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">i</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">3</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">2</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">offset</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">i</span> <span class="o">*</span> <span class="n">x</span><span class="o">.</span><span class="n">strides</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="p">[</span><span class="mi">3</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">2</span><span class="p">]</span>
<span class="go">813</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">offset</span> <span class="o">/</span> <span class="n">x</span><span class="o">.</span><span class="n">itemsize</span>
<span class="go">813</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.sum">
<code class="sig-name descname">sum</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">dtype=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.sum" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the sum of the array elements over the given axis.</p>
<p>Masked elements are set to 0 internally.</p>
<p>Refer to <cite>numpy.sum</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.sum()</span></code></dt><dd><p>corresponding function for ndarrays</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.sum()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">],[</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">]],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[1, --, 3],</span>
<span class="go">        [--, 5, --],</span>
<span class="go">        [7, --, 9]],</span>
<span class="go">  mask=[[False,  True, False],</span>
<span class="go">        [ True, False,  True],</span>
<span class="go">        [False,  True, False]],</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">sum</span><span class="p">()</span>
<span class="go">25</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">masked_array(data=[4, 5, 16],</span>
<span class="go">             mask=[False, False, False],</span>
<span class="go">       fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">masked_array(data=[8, 5, 12],</span>
<span class="go">             mask=[False, False, False],</span>
<span class="go">       fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int64</span><span class="p">)[</span><span class="mi">0</span><span class="p">]))</span>
<span class="go">&lt;class &#39;numpy.int64&#39;&gt;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.swapaxes">
<code class="sig-name descname">swapaxes</code><span class="sig-paren">(</span><em class="sig-param">axis1</em>, <em class="sig-param">axis2</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.swapaxes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a view of the array with <cite>axis1</cite> and <cite>axis2</cite> interchanged.</p>
<p>Refer to <cite>numpy.swapaxes</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.swapaxes()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.take">
<code class="sig-name descname">take</code><span class="sig-paren">(</span><em class="sig-param">indices</em>, <em class="sig-param">axis=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">mode='raise'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.take" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.tobytes">
<code class="sig-name descname">tobytes</code><span class="sig-paren">(</span><em class="sig-param">fill_value=None</em>, <em class="sig-param">order='C'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.tobytes" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the array data as a string containing the raw bytes in the array.</p>
<p>The array is filled with a fill value before the string conversion.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.9.0.</span></p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fill_value</strong> (<em>scalar</em><em>, </em><em>optional</em>) – Value used to fill in the masked values. Default is None, in which
case <cite>MaskedArray.fill_value</cite> is used.</p></li>
<li><p><strong>order</strong> (<em>{'C'</em><em>,</em><em>'F'</em><em>,</em><em>'A'}</em><em>, </em><em>optional</em>) – <p>Order of the data item in the copy. Default is ‘C’.</p>
<ul>
<li><p>’C’   – C order (row major).</p></li>
<li><p>’F’   – Fortran order (column major).</p></li>
<li><p>’A’   – Any, current order of array.</p></li>
<li><p>None  – Same as ‘A’.</p></li>
</ul>
</p></li>
</ul>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.tobytes()</span></code>, <a class="reference internal" href="#fanc.matrix.RegionMatrix.tolist" title="fanc.matrix.RegionMatrix.tolist"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tolist()</span></code></a>, <a class="reference internal" href="#fanc.matrix.RegionMatrix.tofile" title="fanc.matrix.RegionMatrix.tofile"><code class="xref py py-meth docutils literal notranslate"><span class="pre">tofile()</span></code></a></p>
</div>
<p class="rubric">Notes</p>
<p>As for <cite>ndarray.tobytes</cite>, information about the shape, dtype, etc.,
but also about <cite>fill_value</cite>, will be lost.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]]),</span> <span class="n">mask</span><span class="o">=</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">],</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">tobytes</span><span class="p">()</span>
<span class="go">b&#39;\x01\x00\x00\x00\x00\x00\x00\x00?B\x0f\x00\x00\x00\x00\x00?B\x0f\x00\x00\x00\x00\x00\x04\x00\x00\x00\x00\x00\x00\x00&#39;</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.tofile">
<code class="sig-name descname">tofile</code><span class="sig-paren">(</span><em class="sig-param">fid</em>, <em class="sig-param">sep=''</em>, <em class="sig-param">format='%s'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.tofile" title="Permalink to this definition">¶</a></dt>
<dd><p>Save a masked array to a file in binary format.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This function is not implemented yet.</p>
</div>
<dl class="field-list simple">
<dt class="field-odd">Raises</dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – When <cite>tofile</cite> is called.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.toflex">
<code class="sig-name descname">toflex</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.toflex" title="Permalink to this definition">¶</a></dt>
<dd><p>Transforms a masked array into a flexible-type array.</p>
<p>The flexible type array that is returned will have two fields:</p>
<ul class="simple">
<li><p>the <code class="docutils literal notranslate"><span class="pre">_data</span></code> field stores the <code class="docutils literal notranslate"><span class="pre">_data</span></code> part of the array.</p></li>
<li><p>the <code class="docutils literal notranslate"><span class="pre">_mask</span></code> field stores the <code class="docutils literal notranslate"><span class="pre">_mask</span></code> part of the array.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>None</strong> – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>record</strong> – A new flexible-type <cite>ndarray</cite> with two fields: the first element
containing a value, the second element containing the corresponding
mask boolean. The returned record shape matches self.shape.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>A side-effect of transforming a masked array into a flexible <cite>ndarray</cite> is
that meta information (<code class="docutils literal notranslate"><span class="pre">fill_value</span></code>, …) will be lost.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">],[</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">]],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[1, --, 3],</span>
<span class="go">        [--, 5, --],</span>
<span class="go">        [7, --, 9]],</span>
<span class="go">  mask=[[False,  True, False],</span>
<span class="go">        [ True, False,  True],</span>
<span class="go">        [False,  True, False]],</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">toflex</span><span class="p">()</span>
<span class="go">array([[(1, False), (2,  True), (3, False)],</span>
<span class="go">       [(4,  True), (5, False), (6,  True)],</span>
<span class="go">       [(7, False), (8,  True), (9, False)]],</span>
<span class="go">      dtype=[(&#39;_data&#39;, &#39;&lt;i8&#39;), (&#39;_mask&#39;, &#39;?&#39;)])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.tolist">
<code class="sig-name descname">tolist</code><span class="sig-paren">(</span><em class="sig-param">fill_value=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.tolist" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the data portion of the masked array as a hierarchical Python list.</p>
<p>Data items are converted to the nearest compatible Python type.
Masked values are converted to <cite>fill_value</cite>. If <cite>fill_value</cite> is None,
the corresponding entries in the output list will be <code class="docutils literal notranslate"><span class="pre">None</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>fill_value</strong> (<em>scalar</em><em>, </em><em>optional</em>) – The value to use for invalid entries. Default is None.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>result</strong> – The Python list representation of the masked array.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">],</span> <span class="p">[</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">]],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="go">[[1, None, 3], [None, 5, None], [7, None, 9]]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">tolist</span><span class="p">(</span><span class="o">-</span><span class="mi">999</span><span class="p">)</span>
<span class="go">[[1, -999, 3], [-999, 5, -999], [7, -999, 9]]</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.torecords">
<code class="sig-name descname">torecords</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.torecords" title="Permalink to this definition">¶</a></dt>
<dd><p>Transforms a masked array into a flexible-type array.</p>
<p>The flexible type array that is returned will have two fields:</p>
<ul class="simple">
<li><p>the <code class="docutils literal notranslate"><span class="pre">_data</span></code> field stores the <code class="docutils literal notranslate"><span class="pre">_data</span></code> part of the array.</p></li>
<li><p>the <code class="docutils literal notranslate"><span class="pre">_mask</span></code> field stores the <code class="docutils literal notranslate"><span class="pre">_mask</span></code> part of the array.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>None</strong> – </p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>record</strong> – A new flexible-type <cite>ndarray</cite> with two fields: the first element
containing a value, the second element containing the corresponding
mask boolean. The returned record shape matches self.shape.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>A side-effect of transforming a masked array into a flexible <cite>ndarray</cite> is
that meta information (<code class="docutils literal notranslate"><span class="pre">fill_value</span></code>, …) will be lost.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">ma</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">],[</span><span class="mi">7</span><span class="p">,</span><span class="mi">8</span><span class="p">,</span><span class="mi">9</span><span class="p">]],</span> <span class="n">mask</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="mi">4</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">masked_array(</span>
<span class="go">  data=[[1, --, 3],</span>
<span class="go">        [--, 5, --],</span>
<span class="go">        [7, --, 9]],</span>
<span class="go">  mask=[[False,  True, False],</span>
<span class="go">        [ True, False,  True],</span>
<span class="go">        [False,  True, False]],</span>
<span class="go">  fill_value=999999)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="o">.</span><span class="n">toflex</span><span class="p">()</span>
<span class="go">array([[(1, False), (2,  True), (3, False)],</span>
<span class="go">       [(4,  True), (5, False), (6,  True)],</span>
<span class="go">       [(7, False), (8,  True), (9, False)]],</span>
<span class="go">      dtype=[(&#39;_data&#39;, &#39;&lt;i8&#39;), (&#39;_mask&#39;, &#39;?&#39;)])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.tostring">
<code class="sig-name descname">tostring</code><span class="sig-paren">(</span><em class="sig-param">fill_value=None</em>, <em class="sig-param">order='C'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.tostring" title="Permalink to this definition">¶</a></dt>
<dd><p>This function is a compatibility alias for tobytes. Despite its name it
returns bytes not strings.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.trace">
<code class="sig-name descname">trace</code><span class="sig-paren">(</span><em class="sig-param">offset=0</em>, <em class="sig-param">axis1=0</em>, <em class="sig-param">axis2=1</em>, <em class="sig-param">dtype=None</em>, <em class="sig-param">out=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.trace" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the sum along diagonals of the array.</p>
<p>Refer to <cite>numpy.trace</cite> for full documentation.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.trace()</span></code></dt><dd><p>equivalent function</p>
</dd>
</dl>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.transpose">
<code class="sig-name descname">transpose</code><span class="sig-paren">(</span><em class="sig-param">*axes</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.transpose" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a view of the array with axes transposed.</p>
<p>For a 1-D array this has no effect, as a transposed vector is simply the
same vector. To convert a 1-D array into a 2D column vector, an additional
dimension must be added. <cite>np.atleast2d(a).T</cite> achieves this, as does
<cite>a[:, np.newaxis]</cite>.
For a 2-D array, this is a standard matrix transpose.
For an n-D array, if axes are given, their order indicates how the
axes are permuted (see Examples). If axes are not provided and
<code class="docutils literal notranslate"><span class="pre">a.shape</span> <span class="pre">=</span> <span class="pre">(i[0],</span> <span class="pre">i[1],</span> <span class="pre">...</span> <span class="pre">i[n-2],</span> <span class="pre">i[n-1])</span></code>, then
<code class="docutils literal notranslate"><span class="pre">a.transpose().shape</span> <span class="pre">=</span> <span class="pre">(i[n-1],</span> <span class="pre">i[n-2],</span> <span class="pre">...</span> <span class="pre">i[1],</span> <span class="pre">i[0])</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>axes</strong> (None, tuple of ints, or <cite>n</cite> ints) – <ul class="simple">
<li><p>None or no argument: reverses the order of the axes.</p></li>
<li><p>tuple of ints: <cite>i</cite> in the <cite>j</cite>-th place in the tuple means <cite>a</cite>’s
<cite>i</cite>-th axis becomes <cite>a.transpose()</cite>’s <cite>j</cite>-th axis.</p></li>
<li><p><cite>n</cite> ints: same as an n-tuple of the same ints (this form is
intended simply as a “convenience” alternative to the tuple form)</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>out</strong> – View of <cite>a</cite>, with axes suitably permuted.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.T()</span></code></dt><dd><p>Array property returning the array transposed.</p>
</dd>
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">ndarray.reshape()</span></code></dt><dd><p>Give a new shape to an array without changing its data.</p>
</dd>
</dl>
</div>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span>
<span class="go">array([[1, 2],</span>
<span class="go">       [3, 4]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">transpose</span><span class="p">()</span>
<span class="go">array([[1, 3],</span>
<span class="go">       [2, 4]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">transpose</span><span class="p">((</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
<span class="go">array([[1, 3],</span>
<span class="go">       [2, 4]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="o">.</span><span class="n">transpose</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span>
<span class="go">array([[1, 3],</span>
<span class="go">       [2, 4]])</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.unshare_mask">
<code class="sig-name descname">unshare_mask</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.unshare_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Copy the mask and set the sharedmask flag to False.</p>
<p>Whether the mask is shared between masked arrays can be seen from
the <cite>sharedmask</cite> property. <cite>unshare_mask</cite> ensures the mask is not shared.
A copy of the mask is only made if it was shared.</p>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#fanc.matrix.RegionMatrix.sharedmask" title="fanc.matrix.RegionMatrix.sharedmask"><code class="xref py py-meth docutils literal notranslate"><span class="pre">sharedmask()</span></code></a></p>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.var">
<code class="sig-name descname">var</code><span class="sig-paren">(</span><em class="sig-param">axis=None</em>, <em class="sig-param">dtype=None</em>, <em class="sig-param">out=None</em>, <em class="sig-param">ddof=0</em>, <em class="sig-param">keepdims=&lt;no value&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.var" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the variance along the specified axis.</p>
<p>Returns the variance of the array elements, a measure of the spread of a
distribution.  The variance is computed for the flattened array by
default, otherwise over the specified axis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>a</strong> (<em>array_like</em>) – Array containing numbers whose variance is desired.  If <cite>a</cite> is not an
array, a conversion is attempted.</p></li>
<li><p><strong>axis</strong> (<em>None</em><em> or </em><em>int</em><em> or </em><em>tuple of ints</em><em>, </em><em>optional</em>) – <p>Axis or axes along which the variance is computed.  The default is to
compute the variance of the flattened array.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.7.0.</span></p>
</div>
<p>If this is a tuple of ints, a variance is performed over multiple axes,
instead of a single axis or all the axes as before.</p>
</p></li>
<li><p><strong>dtype</strong> (<em>data-type</em><em>, </em><em>optional</em>) – Type to use in computing the variance.  For arrays of integer type
the default is <cite>float32</cite>; for arrays of float types it is the same as
the array type.</p></li>
<li><p><strong>out</strong> (<em>ndarray</em><em>, </em><em>optional</em>) – Alternate output array in which to place the result.  It must have
the same shape as the expected output, but the type is cast if
necessary.</p></li>
<li><p><strong>ddof</strong> (<em>int</em><em>, </em><em>optional</em>) – “Delta Degrees of Freedom”: the divisor used in the calculation is
<code class="docutils literal notranslate"><span class="pre">N</span> <span class="pre">-</span> <span class="pre">ddof</span></code>, where <code class="docutils literal notranslate"><span class="pre">N</span></code> represents the number of elements. By
default <cite>ddof</cite> is zero.</p></li>
<li><p><strong>keepdims</strong> (<em>bool</em><em>, </em><em>optional</em>) – <p>If this is set to True, the axes which are reduced are left
in the result as dimensions with size one. With this option,
the result will broadcast correctly against the input array.</p>
<p>If the default value is passed, then <cite>keepdims</cite> will not be
passed through to the <cite>var</cite> method of sub-classes of
<cite>ndarray</cite>, however any non-default value will be.  If the
sub-class’ method does not implement <cite>keepdims</cite> any
exceptions will be raised.</p>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><strong>variance</strong> – If <code class="docutils literal notranslate"><span class="pre">out=None</span></code>, returns a new array containing the variance;
otherwise, a reference to the output array is returned.</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>ndarray, see dtype parameter above</p>
</dd>
</dl>
<div class="admonition seealso">
<p class="admonition-title">See also</p>
<p><a class="reference internal" href="#fanc.matrix.RegionMatrix.std" title="fanc.matrix.RegionMatrix.std"><code class="xref py py-meth docutils literal notranslate"><span class="pre">std()</span></code></a>, <a class="reference internal" href="#fanc.matrix.RegionMatrix.mean" title="fanc.matrix.RegionMatrix.mean"><code class="xref py py-meth docutils literal notranslate"><span class="pre">mean()</span></code></a>, <code class="xref py py-meth docutils literal notranslate"><span class="pre">nanmean()</span></code>, <code class="xref py py-meth docutils literal notranslate"><span class="pre">nanstd()</span></code>, <code class="xref py py-meth docutils literal notranslate"><span class="pre">nanvar()</span></code></p>
<dl class="simple">
<dt><code class="xref py py-meth docutils literal notranslate"><span class="pre">numpy.doc.ufuncs()</span></code></dt><dd><p>Section “Output arguments”</p>
</dd>
</dl>
</div>
<p class="rubric">Notes</p>
<p>The variance is the average of the squared deviations from the mean,
i.e.,  <code class="docutils literal notranslate"><span class="pre">var</span> <span class="pre">=</span> <span class="pre">mean(abs(x</span> <span class="pre">-</span> <span class="pre">x.mean())**2)</span></code>.</p>
<p>The mean is normally calculated as <code class="docutils literal notranslate"><span class="pre">x.sum()</span> <span class="pre">/</span> <span class="pre">N</span></code>, where <code class="docutils literal notranslate"><span class="pre">N</span> <span class="pre">=</span> <span class="pre">len(x)</span></code>.
If, however, <cite>ddof</cite> is specified, the divisor <code class="docutils literal notranslate"><span class="pre">N</span> <span class="pre">-</span> <span class="pre">ddof</span></code> is used
instead.  In standard statistical practice, <code class="docutils literal notranslate"><span class="pre">ddof=1</span></code> provides an
unbiased estimator of the variance of a hypothetical infinite population.
<code class="docutils literal notranslate"><span class="pre">ddof=0</span></code> provides a maximum likelihood estimate of the variance for
normally distributed variables.</p>
<p>Note that for complex numbers, the absolute value is taken before
squaring, so that the result is always real and nonnegative.</p>
<p>For floating-point input, the variance is computed using the same
precision the input has.  Depending on the input data, this can cause
the results to be inaccurate, especially for <cite>float32</cite> (see example
below).  Specifying a higher-accuracy accumulator using the <code class="docutils literal notranslate"><span class="pre">dtype</span></code>
keyword can alleviate this issue.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">],</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">]])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="go">1.25</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">array([1.,  1.])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">array([0.25,  0.25])</span>
</pre></div>
</div>
<p>In single precision, var() can be inaccurate:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">a</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">2</span><span class="p">,</span> <span class="mi">512</span><span class="o">*</span><span class="mi">512</span><span class="p">),</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">a</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="p">:]</span> <span class="o">=</span> <span class="mf">0.1</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">a</span><span class="p">)</span>
<span class="go">0.20250003</span>
</pre></div>
</div>
<p>Computing the variance in float64 is more accurate:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">var</span><span class="p">(</span><span class="n">a</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">float64</span><span class="p">)</span>
<span class="go">0.20249999932944759 # may vary</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">((</span><span class="mi">1</span><span class="o">-</span><span class="mf">0.55</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span> <span class="o">+</span> <span class="p">(</span><span class="mf">0.1</span><span class="o">-</span><span class="mf">0.55</span><span class="p">)</span><span class="o">**</span><span class="mi">2</span><span class="p">)</span><span class="o">/</span><span class="mi">2</span>
<span class="go">0.2025</span>
</pre></div>
</div>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrix.view">
<code class="sig-name descname">view</code><span class="sig-paren">(</span><em class="sig-param">dtype=None</em>, <em class="sig-param">type=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrix.view" title="Permalink to this definition">¶</a></dt>
<dd><p>New view of array with the same data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dtype</strong> (<em>data-type</em><em> or </em><em>ndarray sub-class</em><em>, </em><em>optional</em>) – Data-type descriptor of the returned view, e.g., float32 or int16. The
default, None, results in the view having the same data-type as <cite>a</cite>.
This argument can also be specified as an ndarray sub-class, which
then specifies the type of the returned object (this is equivalent to
setting the <code class="docutils literal notranslate"><span class="pre">type</span></code> parameter).</p></li>
<li><p><strong>type</strong> (<em>Python type</em><em>, </em><em>optional</em>) – Type of the returned view, e.g., ndarray or matrix.  Again, the
default None results in type preservation.</p></li>
</ul>
</dd>
</dl>
<p class="rubric">Notes</p>
<p><code class="docutils literal notranslate"><span class="pre">a.view()</span></code> is used two different ways:</p>
<p><code class="docutils literal notranslate"><span class="pre">a.view(some_dtype)</span></code> or <code class="docutils literal notranslate"><span class="pre">a.view(dtype=some_dtype)</span></code> constructs a view
of the array’s memory with a different data-type.  This can cause a
reinterpretation of the bytes of memory.</p>
<p><code class="docutils literal notranslate"><span class="pre">a.view(ndarray_subclass)</span></code> or <code class="docutils literal notranslate"><span class="pre">a.view(type=ndarray_subclass)</span></code> just
returns an instance of <cite>ndarray_subclass</cite> that looks at the same array
(same shape, dtype, etc.)  This does not cause a reinterpretation of the
memory.</p>
<p>For <code class="docutils literal notranslate"><span class="pre">a.view(some_dtype)</span></code>, if <code class="docutils literal notranslate"><span class="pre">some_dtype</span></code> has a different number of
bytes per entry than the previous dtype (for example, converting a
regular array to a structured array), then the behavior of the view
cannot be predicted just from the superficial appearance of <code class="docutils literal notranslate"><span class="pre">a</span></code> (shown
by <code class="docutils literal notranslate"><span class="pre">print(a)</span></code>). It also depends on exactly how <code class="docutils literal notranslate"><span class="pre">a</span></code> is stored in
memory. Therefore if <code class="docutils literal notranslate"><span class="pre">a</span></code> is C-ordered versus fortran-ordered, versus
defined as a slice or transpose, etc., the view may give different
results.</p>
<p class="rubric">Examples</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">)],</span> <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)])</span>
</pre></div>
</div>
<p>Viewing array data using a different type and dtype:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">matrix</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span>
<span class="go">matrix([[513]], dtype=int16)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="nb">print</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">y</span><span class="p">))</span>
<span class="go">&lt;class &#39;numpy.matrix&#39;&gt;</span>
</pre></div>
</div>
<p>Creating a view on a structured array so it can be used in calculations</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">),(</span><span class="mi">3</span><span class="p">,</span><span class="mi">4</span><span class="p">)],</span> <span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;a&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;b&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">xv</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int8</span><span class="p">)</span><span class="o">.</span><span class="n">reshape</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">xv</span>
<span class="go">array([[1, 2],</span>
<span class="go">       [3, 4]], dtype=int8)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">xv</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="go">array([2.,  3.])</span>
</pre></div>
</div>
<p>Making changes to the view changes the underlying array</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">xv</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">20</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">x</span>
<span class="go">array([(1, 20), (3,  4)], dtype=[(&#39;a&#39;, &#39;i1&#39;), (&#39;b&#39;, &#39;i1&#39;)])</span>
</pre></div>
</div>
<p>Using a view to convert an array to a recarray:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">z</span> <span class="o">=</span> <span class="n">x</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">recarray</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">z</span><span class="o">.</span><span class="n">a</span>
<span class="go">array([1, 3], dtype=int8)</span>
</pre></div>
</div>
<p>Views share data:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">z</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="go">(9, 10)</span>
</pre></div>
</div>
<p>Views that change the dtype size (bytes per entry) should normally be
avoided on arrays defined by slices, transposes, fortran-ordering, etc.:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">x</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([[</span><span class="mi">1</span><span class="p">,</span><span class="mi">2</span><span class="p">,</span><span class="mi">3</span><span class="p">],[</span><span class="mi">4</span><span class="p">,</span><span class="mi">5</span><span class="p">,</span><span class="mi">6</span><span class="p">]],</span> <span class="n">dtype</span><span class="o">=</span><span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span> <span class="o">=</span> <span class="n">x</span><span class="p">[:,</span> <span class="mi">0</span><span class="p">:</span><span class="mi">2</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span>
<span class="go">array([[1, 2],</span>
<span class="go">       [4, 5]], dtype=int16)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">y</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;width&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;length&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">)])</span>
<span class="gt">Traceback (most recent call last):</span>
    <span class="o">...</span>
<span class="gr">ValueError</span>: <span class="n">To change to a dtype of a different size, the array must be C-contiguous</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">z</span> <span class="o">=</span> <span class="n">y</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">z</span><span class="o">.</span><span class="n">view</span><span class="p">(</span><span class="n">dtype</span><span class="o">=</span><span class="p">[(</span><span class="s1">&#39;width&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">),</span> <span class="p">(</span><span class="s1">&#39;length&#39;</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">int16</span><span class="p">)])</span>
<span class="go">array([[(1, 2)],</span>
<span class="go">       [(4, 5)]], dtype=[(&#39;width&#39;, &#39;&lt;i2&#39;), (&#39;length&#39;, &#39;&lt;i2&#39;)])</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.matrix.RegionMatrixContainer">
<em class="property">class </em><code class="sig-prename descclassname">fanc.matrix.</code><code class="sig-name descname">RegionMatrixContainer</code><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.matrix.RegionPairsContainer</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.regions.RegionBasedWithBins</span></code></p>
<p>Class representing matrices where pixels correspond to genomic region pairs.</p>
<p>This is the common interface for all matrix-based classes, such as
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> or <code class="xref py py-class docutils literal notranslate"><span class="pre">FoldChangeMatrix</span></code>. It provides
access to specialised matrix methods, most importantly
<a class="reference internal" href="#fanc.matrix.RegionMatrixContainer.matrix" title="fanc.matrix.RegionMatrixContainer.matrix"><code class="xref py py-func docutils literal notranslate"><span class="pre">matrix()</span></code></a>, which assembles <code class="xref py py-mod docutils literal notranslate"><span class="pre">numpy</span></code>
arrays from the list of pairwise contacts stored in each object.</p>
<p>It inherits all region methods from <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code>,
and all edge/contact methods from <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>.
You can use the same type of keys for <a class="reference internal" href="#fanc.matrix.RegionMatrixContainer.matrix" title="fanc.matrix.RegionMatrixContainer.matrix"><code class="xref py py-func docutils literal notranslate"><span class="pre">matrix()</span></code></a>
that you would use for <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>, and additionally
have the option to retrieve the observed/expected matrix.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>
<span class="n">hic</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;output/hic/binned/fanc_example_1mb.hic&quot;</span><span class="p">)</span>

<span class="c1"># get the whole-genome matrix</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">hic</span><span class="o">.</span><span class="n">matrix</span><span class="p">()</span>
<span class="nb">type</span><span class="p">(</span><span class="n">m</span><span class="p">)</span>  <span class="c1"># fanc.matrix.RegionMatrix</span>
<span class="nb">isinstance</span><span class="p">(</span><span class="n">m</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">)</span>  <span class="c1"># True</span>
<span class="n">m</span><span class="o">.</span><span class="n">shape</span>  <span class="c1"># 139, 139</span>

<span class="c1"># get just the chromosome 18 intra-chromosomal matrix</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">hic</span><span class="o">.</span><span class="n">matrix</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr18&#39;</span><span class="p">))</span>
<span class="n">m</span><span class="o">.</span><span class="n">shape</span>  <span class="c1"># 79, 79</span>

<span class="c1"># get all rows of the whole-genome matrix</span>
<span class="c1"># corresponding to chromosome 18</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">hic</span><span class="o">.</span><span class="n">matrix</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">)</span>
<span class="n">m</span><span class="o">.</span><span class="n">shape</span>  <span class="c1"># 79, 139</span>

<span class="c1"># get unnormalised chromosome 18 matrix</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">hic</span><span class="o">.</span><span class="n">matrix</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr18&#39;</span><span class="p">),</span> <span class="n">norm</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="c1"># get chromosome 18 O/E matrix</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">hic</span><span class="o">.</span><span class="n">matrix</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr18&#39;</span><span class="p">),</span> <span class="n">oe</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="c1"># get log2-transformed chromosome 18 O/E matrix</span>
<span class="n">m</span> <span class="o">=</span> <span class="n">hic</span><span class="o">.</span><span class="n">matrix</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr18&#39;</span><span class="p">),</span> <span class="n">oe</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">log</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.add_contact">
<code class="sig-name descname">add_contact</code><span class="sig-paren">(</span><em class="sig-param">contact</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.add_contact" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.add_edge" title="fanc.matrix.RegionPairsContainer.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contact</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.add_contacts">
<code class="sig-name descname">add_contacts</code><span class="sig-paren">(</span><em class="sig-param">contacts</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.add_contacts" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#fanc.matrix.RegionPairsTable.add_edges" title="fanc.matrix.RegionPairsTable.add_edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edges()</span></code></a></p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.add_edge">
<code class="sig-name descname">add_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">check_nodes_exist=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.add_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an edge / contact between two regions to this object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>, dict with at least the
attributes source and sink, optionally weight,
or a list of length 2 (source, sink) or 3
(source, sink, weight).</p></li>
<li><p><strong>check_nodes_exist</strong> – Make sure that there are nodes
that match source and sink indexes</p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.add_edge_from_dict">
<code class="sig-name descname">add_edge_from_dict</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.add_edge_from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from dict input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – dict with at least the keys “source”
and “sink”. Additional keys will be loaded
as edge attributes</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.add_edge_from_edge">
<code class="sig-name descname">add_edge_from_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.add_edge_from_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.add_edge_from_list">
<code class="sig-name descname">add_edge_from_list</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.add_edge_from_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from list or tuple input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – List or tuple. Should be of length 2
(source, sink) or 3 (source, sink, weight)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.add_edge_simple">
<code class="sig-name descname">add_edge_simple</code><span class="sig-paren">(</span><em class="sig-param">source</em>, <em class="sig-param">sink</em>, <em class="sig-param">weight=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.add_edge_simple" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> – Source region index</p></li>
<li><p><strong>sink</strong> – Sink region index</p></li>
<li><p><strong>weight</strong> – Weight of the edge</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.add_edges">
<code class="sig-name descname">add_edges</code><span class="sig-paren">(</span><em class="sig-param">edges</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.add_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Bulk-add edges from a list.</p>
<p>List items can be any of the supported edge types,
list, tuple, dict, or <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>. Repeatedly
calls <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.add_edge" title="fanc.matrix.RegionPairsContainer.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a>, so
may be inefficient for large amounts of data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edges</strong> – List (or iterator) of edges. See
<a class="reference internal" href="#fanc.matrix.RegionMatrixTable.add_edge" title="fanc.matrix.RegionMatrixTable.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a>
for details</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.add_region">
<code class="sig-name descname">add_region</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.add_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.bin_intervals">
<em class="property">static </em><code class="sig-name descname">bin_intervals</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bins</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.bin_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.bin_intervals_equidistant">
<em class="property">static </em><code class="sig-name descname">bin_intervals_equidistant</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bin_size</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.bin_intervals_equidistant" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.bin_size">
<em class="property">property </em><code class="sig-name descname">bin_size</code><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.bin_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.binned_regions">
<code class="sig-name descname">binned_regions</code><span class="sig-paren">(</span><em class="sig-param">region=None</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.binned_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.bins_to_distance">
<code class="sig-name descname">bins_to_distance</code><span class="sig-paren">(</span><em class="sig-param">bins</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.bins_to_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.chromosome_bins">
<em class="property">property </em><code class="sig-name descname">chromosome_bins</code><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.chromosome_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.chromosome_lengths">
<em class="property">property </em><code class="sig-name descname">chromosome_lengths</code><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.chromosome_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.chromosomes">
<code class="sig-name descname">chromosomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of chromosome names.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.distance_to_bins">
<code class="sig-name descname">distance_to_bins</code><span class="sig-paren">(</span><em class="sig-param">distance</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.distance_to_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.edge_data">
<code class="sig-name descname">edge_data</code><span class="sig-paren">(</span><em class="sig-param">attribute</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.edge_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over specific edge attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> – Name of the attribute, e.g. “weight”</p></li>
<li><p><strong>args</strong> – Positional arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator over edge attribute</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.edge_subset">
<code class="sig-name descname">edge_subset</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.edge_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a subset of edges.</p>
<p>This is an alias for <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>generator (<a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.edges">
<em class="property">property </em><code class="sig-name descname">edges</code><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over contacts / edges.</p>
<p><a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> is the central function of
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>. Here, we will use the
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> implementation for demonstration purposes,
but the usage is exactly the same for all compatible
objects implementing <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>, including
<a class="reference internal" href="compatibility/juicer.html#fanc.compatibility.juicer.JuicerHic" title="fanc.compatibility.juicer.JuicerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">JuicerHic</span></code></a> and
<a class="reference internal" href="compatibility/cooler.html#fanc.compatibility.cooler.CoolerHic" title="fanc.compatibility.cooler.CoolerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">CoolerHic</span></code></a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>

<span class="c1"># file from FAN-C examples</span>
<span class="n">hic</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;output/hic/binned/fanc_example_1mb.hic&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can easily find the number of edges in the sample
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">)</span>  <span class="c1"># 8695</span>
</pre></div>
</div>
<p>When used in an iterator context, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
iterates over all edges in the <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Calling <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> as a method has the
same effect:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># note the &#39;()&#39;</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Rather than iterate over all edges in the object, we can select only a subset.
If the key is a string or a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, all non-zero edges connecting
the region described by the key to any other region are returned. If the key is a
tuple of strings or <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, only edges between the two regions
are returned.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># select all edges between chromosome 19</span>
<span class="c1"># and any other region:</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s2">&quot;chr19&quot;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>

<span class="c1"># select all edges that are only on</span>
<span class="c1"># chromosome 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr19&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 90--116; bias: 0.00021173151730025176; sink_node: chr19:37000001-38000000; source_node: chr19:11000001-12000000; weight: 0.009104455243910825</span>
    <span class="c1"># 135--135; bias: 0.00018003890596887822; sink_node: chr19:56000001-57000000; source_node: chr19:56000001-57000000; weight: 0.10028167062466517</span>
    <span class="c1"># 123--123; bias: 0.00011063368998965993; sink_node: chr19:44000001-45000000; source_node: chr19:44000001-45000000; weight: 0.1386240135570439</span>
    <span class="c1"># 92--93; bias: 0.00040851066434864896; sink_node: chr19:14000001-15000000; source_node: chr19:13000001-14000000; weight: 0.10090213409411629</span>
    <span class="c1"># ...</span>

<span class="c1"># select inter-chromosomal edges</span>
<span class="c1"># between chromosomes 18 and 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>By default, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> will retrieve all edge attributes,
which can be slow when iterating over a lot of edges. This is why all file-based FAN-C
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a> objects support lazy loading, where attributes
are only read on demand.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">,</span> <span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42 42 0.12291311562018173 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #0&gt;</span>
    <span class="c1"># 24 28 0.025205961072838057 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #1&gt;</span>
    <span class="c1"># 5 76 0.00961709840049876 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #2&gt;</span>
    <span class="c1"># 66 68 0.03876763316345468 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #3&gt;</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The lazy iterator reuses the <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> object in every iteration,
and overwrites the <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> attributes. Therefore <strong>do not</strong> use
lazy iterators if you need to store edge objects for later access.
For example, the following code works as expected
<code class="code docutils literal notranslate"><span class="pre">list(hic.edges())</span></code>, with all <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> objects stored in the
list, while this code <code class="code docutils literal notranslate"><span class="pre">list(hic.edges(lazy=True))</span></code>
will result in a list of identical <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> objects. Always ensure
you do all edge processing in the loop when working with lazy iterators!</p>
</div>
<p>When working with normalised contact frequencies, such as obtained through
matrix balancing in the example above, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
automatically returns normalised edge weights. In addition, the <code class="code docutils literal notranslate"><span class="pre">bias</span></code>
attribute will (typically) have a value different from 1.</p>
<p>When you are interested in the raw contact frequency, use the <code class="code docutils literal notranslate"><span class="pre">norm=False</span></code>
parameter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">)</span>
    <span class="c1"># 42 42 2120.0</span>
    <span class="c1"># 24 28 388.0</span>
    <span class="c1"># 5 76 94.0</span>
    <span class="c1"># 66 68 470.0</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>You can also choose to omit all intra- or inter-chromosomal edges using
<code class="code docutils literal notranslate"><span class="pre">intra_chromosomal=False</span></code> or <code class="code docutils literal notranslate"><span class="pre">inter_chromosomal=False</span></code>, respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Iterator over <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> or equivalent.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.edges_dict">
<code class="sig-name descname">edges_dict</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.edges_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Edges iterator with access by bracket notation.</p>
<p>This iterator <strong>always</strong> returns unnormalised edges.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict or dict-like iterator</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.expected_values">
<code class="sig-name descname">expected_values</code><span class="sig-paren">(</span><em class="sig-param">selected_chromosome=None</em>, <em class="sig-param">norm=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.expected_values" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the expected values for genomic contacts at all distances.</p>
<p>This calculates the expected values between genomic regions
separated by a specific distance. Expected values are calculated
as the average weight of edges between region pairs with the same
genomic separation, taking into account unmappable regions.</p>
<p>It will return a tuple with three values: a list of genome-wide
intra-chromosomal expected values (list index corresponds to number
of separating bins), a dict with chromosome names as keys and
intra-chromosomal expected values specific to each chromosome, and
a float for inter-chromosomal expected value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_chromosome</strong> – (optional) Chromosome name. If provided,
will only return expected values for this
chromosome.</p></li>
<li><p><strong>norm</strong> – If False, will calculate the expected values on the
unnormalised matrix.</p></li>
<li><p><strong>args</strong> – Not used in this context</p></li>
<li><p><strong>kwargs</strong> – Not used in this context</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of intra-chromosomal expected values,
dict of intra-chromosomal expected values by chromosome,
inter-chromosomal expected value</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.expected_values_and_marginals">
<code class="sig-name descname">expected_values_and_marginals</code><span class="sig-paren">(</span><em class="sig-param">selected_chromosome=None</em>, <em class="sig-param">norm=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.expected_values_and_marginals" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the expected values for genomic contacts at all distances
and the whole matrix marginals.</p>
<p>This calculates the expected values between genomic regions
separated by a specific distance. Expected values are calculated
as the average weight of edges between region pairs with the same
genomic separation, taking into account unmappable regions.</p>
<p>It will return a tuple with three values: a list of genome-wide
intra-chromosomal expected values (list index corresponds to number
of separating bins), a dict with chromosome names as keys and
intra-chromosomal expected values specific to each chromosome, and
a float for inter-chromosomal expected value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_chromosome</strong> – (optional) Chromosome name. If provided,
will only return expected values for this
chromosome.</p></li>
<li><p><strong>norm</strong> – If False, will calculate the expected values on the
unnormalised matrix.</p></li>
<li><p><strong>args</strong> – Not used in this context</p></li>
<li><p><strong>kwargs</strong> – Not used in this context</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of intra-chromosomal expected values,
dict of intra-chromosomal expected values by chromosome,
inter-chromosomal expected value</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.find_region">
<code class="sig-name descname">find_region</code><span class="sig-paren">(</span><em class="sig-param">query_regions</em>, <em class="sig-param">_regions_dict=None</em>, <em class="sig-param">_region_ends=None</em>, <em class="sig-param">_chromosomes=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.find_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.intervals">
<code class="sig-name descname">intervals</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.mappable">
<code class="sig-name descname">mappable</code><span class="sig-paren">(</span><em class="sig-param">region=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.mappable" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the mappability of regions in this object.</p>
<p>A “mappable” region has at least one contact to another region
in the genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">array</span></code> where True means mappable
and False unmappable</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.marginals">
<code class="sig-name descname">marginals</code><span class="sig-paren">(</span><em class="sig-param">masked=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.marginals" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the marginals vector of this Hic matrix.</p>
<p>Sums up all contacts for each bin of the Hi-C matrix.
Unmappable regoins will be masked in the returned vector unless
the <code class="code docutils literal notranslate"><span class="pre">masked</span></code> parameter is set to <code class="code docutils literal notranslate"><span class="pre">False</span></code>.</p>
<p>By default, corrected matrix entries are summed up.
To get uncorrected matrix marginals use <code class="code docutils literal notranslate"><span class="pre">norm=False</span></code>.
Generally, all parameters accepted by <a class="reference internal" href="#fanc.matrix.RegionMatrixContainer.edges" title="fanc.matrix.RegionMatrixContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
are supported.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>masked</strong> – Use a numpy masked array to mask entries
corresponding to unmappable regions</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.matrix">
<code class="sig-name descname">matrix</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">log=False</em>, <em class="sig-param">default_value=None</em>, <em class="sig-param">mask=True</em>, <em class="sig-param">log_base=2</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Assemble a <a class="reference internal" href="#fanc.matrix.RegionMatrix" title="fanc.matrix.RegionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrix</span></code></a> from region pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Matrix selector. See <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
for all supported key types</p></li>
<li><p><strong>log</strong> – If True, log-transform the matrix entries. Also see log_base</p></li>
<li><p><strong>log_base</strong> – Base of the log transformation. Default: 2; only used when
log=True</p></li>
<li><p><strong>default_value</strong> – (optional) set the default value of matrix entries
that have no associated edge/contact</p></li>
<li><p><strong>mask</strong> – If False, do not mask unmappable regions</p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<a class="reference internal" href="#fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries" title="fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries"><code class="xref py py-func docutils literal notranslate"><span class="pre">regions_and_matrix_entries()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<a class="reference internal" href="#fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries" title="fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries"><code class="xref py py-func docutils literal notranslate"><span class="pre">regions_and_matrix_entries()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.matrix.RegionMatrix" title="fanc.matrix.RegionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrix</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.merge">
<em class="property">classmethod </em><code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">pairs</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge two or more <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pairs</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code> of <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a></p></li>
<li><p><strong>args</strong> – Positional arguments passed to constructor of this
class</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to constructor of this
class</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.possible_contacts">
<code class="sig-name descname">possible_contacts</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.possible_contacts" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the possible number of contacts in the genome.</p>
<p>This calculates the number of potential region pairs in
a genome for any possible separation distance, taking into
account the existence of unmappable regions.</p>
<p>It will calculate one number for inter-chromosomal pairs,
return a list with the number of possible pairs where the
list index corresponds to the number of bins separating two regions,
and a dictionary of lists for each chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>possible intra-chromosomal pairs,
possible intra-chromosomal pairs by chromosome,
possible inter-chromosomal pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.region_bins">
<code class="sig-name descname">region_bins</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.region_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.region_intervals">
<code class="sig-name descname">region_intervals</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">score_field='score'</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.region_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.region_subset">
<code class="sig-name descname">region_subset</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.region_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.regions">
<em class="property">property </em><code class="sig-name descname">regions</code><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.regions_and_edges">
<code class="sig-name descname">regions_and_edges</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.regions_and_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenient access to regions and edges selected by key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Edge selector, see <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>args</strong> – Positional arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of row regions, list of col regions, iterator over edges</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries">
<code class="sig-name descname">regions_and_matrix_entries</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">oe=False</em>, <em class="sig-param">oe_per_chromosome=True</em>, <em class="sig-param">score_field=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenient access to non-zero matrix entries and associated regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Edge key, see <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>oe</strong> – If True, will divide observed values by their expected value
at the given distance. False by default</p></li>
<li><p><strong>oe_per_chromosome</strong> – If True (default), will do a per-chromosome O/E
calculation rather than using the whole matrix
to obtain expected values</p></li>
<li><p><strong>score_field</strong> – (optional) any edge attribute that returns a number
can be specified here for filling the matrix. Usually
this is defined by the <code class="code docutils literal notranslate"><span class="pre">_default_score_field</span></code>
attribute of the matrix class.</p></li>
<li><p><strong>args</strong> – Positional arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of row regions, list of col regions, iterator over (i, j, weight) tuples</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.regions_dict">
<em class="property">property </em><code class="sig-name descname">regions_dict</code><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.regions_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.regions_identical">
<em class="property">static </em><code class="sig-name descname">regions_identical</code><span class="sig-paren">(</span><em class="sig-param">pairs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.regions_identical" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the regions in all objects in the list are identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pairs</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code>
objects</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if chromosome, start, and end are identical between
all regions in the same list positions.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.scaling_factor">
<code class="sig-name descname">scaling_factor</code><span class="sig-paren">(</span><em class="sig-param">matrix</em>, <em class="sig-param">weight_column=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.scaling_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the scaling factor to another matrix.</p>
<p>Calculates the ratio between the number of contacts in
this Hic object to the number of contacts in another
Hic object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> object</p></li>
<li><p><strong>weight_column</strong> – Name of the column to calculate the scaling factor on</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.to_bed">
<code class="sig-name descname">to_bed</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.to_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.to_bigwig">
<code class="sig-name descname">to_bigwig</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.to_bigwig" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixContainer.to_gff">
<code class="sig-name descname">to_gff</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixContainer.to_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.matrix.RegionMatrixTable">
<em class="property">class </em><code class="sig-prename descclassname">fanc.matrix.</code><code class="sig-name descname">RegionMatrixTable</code><span class="sig-paren">(</span><em class="sig-param">file_name=None</em>, <em class="sig-param">mode='a'</em>, <em class="sig-param">tmpdir=None</em>, <em class="sig-param">partition_strategy='auto'</em>, <em class="sig-param">additional_region_fields=None</em>, <em class="sig-param">additional_edge_fields=None</em>, <em class="sig-param">default_score_field='weight'</em>, <em class="sig-param">default_value=0.0</em>, <em class="sig-param">_table_name_regions='regions'</em>, <em class="sig-param">_table_name_edges='edges'</em>, <em class="sig-param">_table_name_expected_values='expected_values'</em>, <em class="sig-param">_edge_buffer_size='3G'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.matrix.RegionMatrixContainer" title="fanc.matrix.RegionMatrixContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.matrix.RegionMatrixContainer</span></code></a>, <a class="reference internal" href="#fanc.matrix.RegionPairsTable" title="fanc.matrix.RegionPairsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.matrix.RegionPairsTable</span></code></a></p>
<p>HDF5 implementation of the <a class="reference internal" href="#fanc.matrix.RegionMatrixContainer" title="fanc.matrix.RegionMatrixContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrixContainer</span></code></a> interface.</p>
<dl class="class">
<dt id="fanc.matrix.RegionMatrixTable.ChromosomeDescription">
<em class="property">class </em><code class="sig-name descname">ChromosomeDescription</code><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.ChromosomeDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="class">
<dt id="fanc.matrix.RegionMatrixTable.MaskDescription">
<em class="property">class </em><code class="sig-name descname">MaskDescription</code><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.MaskDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
</dd></dl>

<dl class="class">
<dt id="fanc.matrix.RegionMatrixTable.RegionDescription">
<em class="property">class </em><code class="sig-name descname">RegionDescription</code><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.RegionDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_contact">
<code class="sig-name descname">add_contact</code><span class="sig-paren">(</span><em class="sig-param">contact</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_contact" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.add_edge" title="fanc.matrix.RegionPairsContainer.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contact</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_contacts">
<code class="sig-name descname">add_contacts</code><span class="sig-paren">(</span><em class="sig-param">contacts</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_contacts" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#fanc.matrix.RegionPairsTable.add_edges" title="fanc.matrix.RegionPairsTable.add_edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edges()</span></code></a></p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_edge">
<code class="sig-name descname">add_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">check_nodes_exist=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an edge / contact between two regions to this object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>, dict with at least the
attributes source and sink, optionally weight,
or a list of length 2 (source, sink) or 3
(source, sink, weight).</p></li>
<li><p><strong>check_nodes_exist</strong> – Make sure that there are nodes
that match source and sink indexes</p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_edge_from_dict">
<code class="sig-name descname">add_edge_from_dict</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_edge_from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from dict input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – dict with at least the keys “source”
and “sink”. Additional keys will be loaded
as edge attributes</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_edge_from_edge">
<code class="sig-name descname">add_edge_from_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_edge_from_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_edge_from_list">
<code class="sig-name descname">add_edge_from_list</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_edge_from_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from list or tuple input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – List or tuple. Should be of length 2
(source, sink) or 3 (source, sink, weight)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_edge_simple">
<code class="sig-name descname">add_edge_simple</code><span class="sig-paren">(</span><em class="sig-param">source</em>, <em class="sig-param">sink</em>, <em class="sig-param">weight=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_edge_simple" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> – Source region index</p></li>
<li><p><strong>sink</strong> – Sink region index</p></li>
<li><p><strong>weight</strong> – Weight of the edge</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_edges">
<code class="sig-name descname">add_edges</code><span class="sig-paren">(</span><em class="sig-param">edges</em>, <em class="sig-param">flush=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Bulk-add edges from a list.</p>
<p>List items can be any of the supported edge types,
list, tuple, dict, or <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>. Repeatedly
calls <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.add_edge" title="fanc.matrix.RegionPairsContainer.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a>, so
may be inefficient for large amounts of data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edges</strong> – List (or iterator) of edges. See
<a class="reference internal" href="#fanc.matrix.RegionMatrixTable.add_edge" title="fanc.matrix.RegionMatrixTable.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a>
for details</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_mask_description">
<code class="sig-name descname">add_mask_description</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">description</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_mask_description" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a mask description to the _mask table and return its ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – name of the mask</p></li>
<li><p><strong>description</strong> (<em>str</em>) – description of the mask</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>id of the mask</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_region">
<code class="sig-name descname">add_region</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.add_regions">
<code class="sig-name descname">add_regions</code><span class="sig-paren">(</span><em class="sig-param">regions</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.add_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.bin_intervals">
<em class="property">static </em><code class="sig-name descname">bin_intervals</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bins</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.bin_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.bin_intervals_equidistant">
<em class="property">static </em><code class="sig-name descname">bin_intervals_equidistant</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bin_size</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.bin_intervals_equidistant" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.bin_size">
<em class="property">property </em><code class="sig-name descname">bin_size</code><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.bin_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.binned_regions">
<code class="sig-name descname">binned_regions</code><span class="sig-paren">(</span><em class="sig-param">region=None</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.binned_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.bins_to_distance">
<code class="sig-name descname">bins_to_distance</code><span class="sig-paren">(</span><em class="sig-param">bins</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.bins_to_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.chromosome_bins">
<em class="property">property </em><code class="sig-name descname">chromosome_bins</code><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.chromosome_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.chromosome_lengths">
<em class="property">property </em><code class="sig-name descname">chromosome_lengths</code><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.chromosome_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.chromosomes">
<code class="sig-name descname">chromosomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.close">
<code class="sig-name descname">close</code><span class="sig-paren">(</span><em class="sig-param">copy_tmp=True</em>, <em class="sig-param">remove_tmp=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.distance_to_bins">
<code class="sig-name descname">distance_to_bins</code><span class="sig-paren">(</span><em class="sig-param">distance</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.distance_to_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.downsample">
<code class="sig-name descname">downsample</code><span class="sig-paren">(</span><em class="sig-param">n</em>, <em class="sig-param">file_name=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.downsample" title="Permalink to this definition">¶</a></dt>
<dd><p>Sample edges from this object.</p>
<p>Sampling is always done on uncorrected Hi-C matrices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> – Sample size or reference object. If n &lt; 1 will be interpreted as
a fraction of total reads in this object.</p></li>
<li><p><strong>file_name</strong> – Output file name for down-sampled object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.matrix.RegionPairsTable" title="fanc.matrix.RegionPairsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.edge_data">
<code class="sig-name descname">edge_data</code><span class="sig-paren">(</span><em class="sig-param">attribute</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.edge_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over specific edge attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> – Name of the attribute, e.g. “weight”</p></li>
<li><p><strong>args</strong> – Positional arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator over edge attribute</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.edge_subset">
<code class="sig-name descname">edge_subset</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.edge_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a subset of edges.</p>
<p>This is an alias for <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>generator (<a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.edges">
<em class="property">property </em><code class="sig-name descname">edges</code><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over contacts / edges.</p>
<p><a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> is the central function of
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>. Here, we will use the
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> implementation for demonstration purposes,
but the usage is exactly the same for all compatible
objects implementing <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>, including
<a class="reference internal" href="compatibility/juicer.html#fanc.compatibility.juicer.JuicerHic" title="fanc.compatibility.juicer.JuicerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">JuicerHic</span></code></a> and
<a class="reference internal" href="compatibility/cooler.html#fanc.compatibility.cooler.CoolerHic" title="fanc.compatibility.cooler.CoolerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">CoolerHic</span></code></a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>

<span class="c1"># file from FAN-C examples</span>
<span class="n">hic</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;output/hic/binned/fanc_example_1mb.hic&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can easily find the number of edges in the sample
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">)</span>  <span class="c1"># 8695</span>
</pre></div>
</div>
<p>When used in an iterator context, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
iterates over all edges in the <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Calling <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> as a method has the
same effect:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># note the &#39;()&#39;</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Rather than iterate over all edges in the object, we can select only a subset.
If the key is a string or a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, all non-zero edges connecting
the region described by the key to any other region are returned. If the key is a
tuple of strings or <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, only edges between the two regions
are returned.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># select all edges between chromosome 19</span>
<span class="c1"># and any other region:</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s2">&quot;chr19&quot;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>

<span class="c1"># select all edges that are only on</span>
<span class="c1"># chromosome 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr19&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 90--116; bias: 0.00021173151730025176; sink_node: chr19:37000001-38000000; source_node: chr19:11000001-12000000; weight: 0.009104455243910825</span>
    <span class="c1"># 135--135; bias: 0.00018003890596887822; sink_node: chr19:56000001-57000000; source_node: chr19:56000001-57000000; weight: 0.10028167062466517</span>
    <span class="c1"># 123--123; bias: 0.00011063368998965993; sink_node: chr19:44000001-45000000; source_node: chr19:44000001-45000000; weight: 0.1386240135570439</span>
    <span class="c1"># 92--93; bias: 0.00040851066434864896; sink_node: chr19:14000001-15000000; source_node: chr19:13000001-14000000; weight: 0.10090213409411629</span>
    <span class="c1"># ...</span>

<span class="c1"># select inter-chromosomal edges</span>
<span class="c1"># between chromosomes 18 and 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>By default, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> will retrieve all edge attributes,
which can be slow when iterating over a lot of edges. This is why all file-based FAN-C
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a> objects support lazy loading, where attributes
are only read on demand.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">,</span> <span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42 42 0.12291311562018173 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #0&gt;</span>
    <span class="c1"># 24 28 0.025205961072838057 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #1&gt;</span>
    <span class="c1"># 5 76 0.00961709840049876 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #2&gt;</span>
    <span class="c1"># 66 68 0.03876763316345468 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #3&gt;</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The lazy iterator reuses the <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> object in every iteration,
and overwrites the <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> attributes. Therefore <strong>do not</strong> use
lazy iterators if you need to store edge objects for later access.
For example, the following code works as expected
<code class="code docutils literal notranslate"><span class="pre">list(hic.edges())</span></code>, with all <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> objects stored in the
list, while this code <code class="code docutils literal notranslate"><span class="pre">list(hic.edges(lazy=True))</span></code>
will result in a list of identical <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> objects. Always ensure
you do all edge processing in the loop when working with lazy iterators!</p>
</div>
<p>When working with normalised contact frequencies, such as obtained through
matrix balancing in the example above, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
automatically returns normalised edge weights. In addition, the <code class="code docutils literal notranslate"><span class="pre">bias</span></code>
attribute will (typically) have a value different from 1.</p>
<p>When you are interested in the raw contact frequency, use the <code class="code docutils literal notranslate"><span class="pre">norm=False</span></code>
parameter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">)</span>
    <span class="c1"># 42 42 2120.0</span>
    <span class="c1"># 24 28 388.0</span>
    <span class="c1"># 5 76 94.0</span>
    <span class="c1"># 66 68 470.0</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>You can also choose to omit all intra- or inter-chromosomal edges using
<code class="code docutils literal notranslate"><span class="pre">intra_chromosomal=False</span></code> or <code class="code docutils literal notranslate"><span class="pre">inter_chromosomal=False</span></code>, respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Iterator over <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> or equivalent.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.edges_dict">
<code class="sig-name descname">edges_dict</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.edges_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Edges iterator with access by bracket notation.</p>
<p>This iterator <strong>always</strong> returns unnormalised edges.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict or dict-like iterator</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.expected_values">
<code class="sig-name descname">expected_values</code><span class="sig-paren">(</span><em class="sig-param">selected_chromosome=None</em>, <em class="sig-param">norm=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.expected_values" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the expected values for genomic contacts at all distances.</p>
<p>This calculates the expected values between genomic regions
separated by a specific distance. Expected values are calculated
as the average weight of edges between region pairs with the same
genomic separation, taking into account unmappable regions.</p>
<p>It will return a tuple with three values: a list of genome-wide
intra-chromosomal expected values (list index corresponds to number
of separating bins), a dict with chromosome names as keys and
intra-chromosomal expected values specific to each chromosome, and
a float for inter-chromosomal expected value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_chromosome</strong> – (optional) Chromosome name. If provided,
will only return expected values for this
chromosome.</p></li>
<li><p><strong>norm</strong> – If False, will calculate the expected values on the
unnormalised matrix.</p></li>
<li><p><strong>args</strong> – Not used in this context</p></li>
<li><p><strong>kwargs</strong> – Not used in this context</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of intra-chromosomal expected values,
dict of intra-chromosomal expected values by chromosome,
inter-chromosomal expected value</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.expected_values_and_marginals">
<code class="sig-name descname">expected_values_and_marginals</code><span class="sig-paren">(</span><em class="sig-param">selected_chromosome=None</em>, <em class="sig-param">norm=True</em>, <em class="sig-param">force=False</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.expected_values_and_marginals" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the expected values for genomic contacts at all distances
and the whole matrix marginals.</p>
<p>This calculates the expected values between genomic regions
separated by a specific distance. Expected values are calculated
as the average weight of edges between region pairs with the same
genomic separation, taking into account unmappable regions.</p>
<p>It will return a tuple with three values: a list of genome-wide
intra-chromosomal expected values (list index corresponds to number
of separating bins), a dict with chromosome names as keys and
intra-chromosomal expected values specific to each chromosome, and
a float for inter-chromosomal expected value.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>selected_chromosome</strong> – (optional) Chromosome name. If provided,
will only return expected values for this
chromosome.</p></li>
<li><p><strong>norm</strong> – If False, will calculate the expected values on the
unnormalised matrix.</p></li>
<li><p><strong>args</strong> – Not used in this context</p></li>
<li><p><strong>kwargs</strong> – Not used in this context</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of intra-chromosomal expected values,
dict of intra-chromosomal expected values by chromosome,
inter-chromosomal expected value</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.filter">
<code class="sig-name descname">filter</code><span class="sig-paren">(</span><em class="sig-param">edge_filter</em>, <em class="sig-param">queue=False</em>, <em class="sig-param">log_progress=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter edges in this object by using a
<code class="xref py py-class docutils literal notranslate"><span class="pre">MaskFilter</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge_filter</strong> – Class implementing <code class="xref py py-class docutils literal notranslate"><span class="pre">MaskFilter</span></code>.</p></li>
<li><p><strong>queue</strong> – If True, filter will be queued and can be executed
along with other queued filters using
<a class="reference internal" href="#fanc.matrix.RegionPairsTable.run_queued_filters" title="fanc.matrix.RegionPairsTable.run_queued_filters"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_queued_filters()</span></code></a></p></li>
<li><p><strong>log_progress</strong> – If true, process iterating through all edges
will be continuously reported.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.find_region">
<code class="sig-name descname">find_region</code><span class="sig-paren">(</span><em class="sig-param">query_regions</em>, <em class="sig-param">_regions_dict=None</em>, <em class="sig-param">_region_ends=None</em>, <em class="sig-param">_chromosomes=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.find_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.flush">
<code class="sig-name descname">flush</code><span class="sig-paren">(</span><em class="sig-param">silent=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.flush" title="Permalink to this definition">¶</a></dt>
<dd><p>Write data to file and flush buffers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>silent</strong> – do not print flush progress</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param">key</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Search _mask table for key and return Mask.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<em>int</em>) – search by mask name</p></li>
<li><p><strong>key</strong> – search by mask ID</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Mask</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.get_masks">
<code class="sig-name descname">get_masks</code><span class="sig-paren">(</span><em class="sig-param">ix</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.get_masks" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract mask IDs encoded in parameter and return masks.</p>
<p>IDs are powers of 2, so a single int field in the table can hold
multiple masks by simply adding up the IDs. Similar principle to
UNIX chmod (although that uses base 8)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ix</strong> (<em>int</em>) – integer that is the sum of powers of 2. Note that this value
is not necessarily itself a power of 2.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of Masks extracted from ix</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list (Mask)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.intervals">
<code class="sig-name descname">intervals</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.mappable">
<code class="sig-name descname">mappable</code><span class="sig-paren">(</span><em class="sig-param">region=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.mappable" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the mappability of regions in this object.</p>
<p>A “mappable” region has at least one contact to another region
in the genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">array</span></code> where True means mappable
and False unmappable</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.marginals">
<code class="sig-name descname">marginals</code><span class="sig-paren">(</span><em class="sig-param">masked=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.marginals" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the marginals vector of this Hic matrix.</p>
<p>Sums up all contacts for each bin of the Hi-C matrix.
Unmappable regoins will be masked in the returned vector unless
the <code class="code docutils literal notranslate"><span class="pre">masked</span></code> parameter is set to <code class="code docutils literal notranslate"><span class="pre">False</span></code>.</p>
<p>By default, corrected matrix entries are summed up.
To get uncorrected matrix marginals use <code class="code docutils literal notranslate"><span class="pre">norm=False</span></code>.
Generally, all parameters accepted by <a class="reference internal" href="#fanc.matrix.RegionMatrixContainer.edges" title="fanc.matrix.RegionMatrixContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
are supported.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>masked</strong> – Use a numpy masked array to mask entries
corresponding to unmappable regions</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.matrix">
<code class="sig-name descname">matrix</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">log=False</em>, <em class="sig-param">default_value=None</em>, <em class="sig-param">mask=True</em>, <em class="sig-param">log_base=2</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Assemble a <a class="reference internal" href="#fanc.matrix.RegionMatrix" title="fanc.matrix.RegionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrix</span></code></a> from region pairs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Matrix selector. See <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
for all supported key types</p></li>
<li><p><strong>log</strong> – If True, log-transform the matrix entries. Also see log_base</p></li>
<li><p><strong>log_base</strong> – Base of the log transformation. Default: 2; only used when
log=True</p></li>
<li><p><strong>default_value</strong> – (optional) set the default value of matrix entries
that have no associated edge/contact</p></li>
<li><p><strong>mask</strong> – If False, do not mask unmappable regions</p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<a class="reference internal" href="#fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries" title="fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries"><code class="xref py py-func docutils literal notranslate"><span class="pre">regions_and_matrix_entries()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<a class="reference internal" href="#fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries" title="fanc.matrix.RegionMatrixContainer.regions_and_matrix_entries"><code class="xref py py-func docutils literal notranslate"><span class="pre">regions_and_matrix_entries()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.matrix.RegionMatrix" title="fanc.matrix.RegionMatrix"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrix</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.merge">
<em class="property">classmethod </em><code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">matrices</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge multiple <a class="reference internal" href="#fanc.matrix.RegionMatrixContainer" title="fanc.matrix.RegionMatrixContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrixContainer</span></code></a> objects.</p>
<p>Merging is done by adding the weight of edges in each object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>matrices</strong> – list of <a class="reference internal" href="#fanc.matrix.RegionMatrixContainer" title="fanc.matrix.RegionMatrixContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrixContainer</span></code></a></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>merged <a class="reference internal" href="#fanc.matrix.RegionMatrixContainer" title="fanc.matrix.RegionMatrixContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMatrixContainer</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.possible_contacts">
<code class="sig-name descname">possible_contacts</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.possible_contacts" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculate the possible number of contacts in the genome.</p>
<p>This calculates the number of potential region pairs in
a genome for any possible separation distance, taking into
account the existence of unmappable regions.</p>
<p>It will calculate one number for inter-chromosomal pairs,
return a list with the number of possible pairs where the
list index corresponds to the number of bins separating two regions,
and a dictionary of lists for each chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>possible intra-chromosomal pairs,
possible intra-chromosomal pairs by chromosome,
possible inter-chromosomal pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.region_bins">
<code class="sig-name descname">region_bins</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.region_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.region_data">
<code class="sig-name descname">region_data</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">value=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.region_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.region_intervals">
<code class="sig-name descname">region_intervals</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">score_field='score'</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.region_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.region_subset">
<code class="sig-name descname">region_subset</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.region_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.regions">
<em class="property">property </em><code class="sig-name descname">regions</code><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.regions_and_edges">
<code class="sig-name descname">regions_and_edges</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.regions_and_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenient access to regions and edges selected by key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Edge selector, see <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>args</strong> – Positional arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of row regions, list of col regions, iterator over edges</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.regions_and_matrix_entries">
<code class="sig-name descname">regions_and_matrix_entries</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">oe=False</em>, <em class="sig-param">oe_per_chromosome=True</em>, <em class="sig-param">score_field=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.regions_and_matrix_entries" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenient access to non-zero matrix entries and associated regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Edge key, see <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>oe</strong> – If True, will divide observed values by their expected value
at the given distance. False by default</p></li>
<li><p><strong>oe_per_chromosome</strong> – If True (default), will do a per-chromosome O/E
calculation rather than using the whole matrix
to obtain expected values</p></li>
<li><p><strong>score_field</strong> – (optional) any edge attribute that returns a number
can be specified here for filling the matrix. Usually
this is defined by the <code class="code docutils literal notranslate"><span class="pre">_default_score_field</span></code>
attribute of the matrix class.</p></li>
<li><p><strong>args</strong> – Positional arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of row regions, list of col regions, iterator over (i, j, weight) tuples</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.regions_dict">
<em class="property">property </em><code class="sig-name descname">regions_dict</code><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.regions_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.regions_identical">
<em class="property">static </em><code class="sig-name descname">regions_identical</code><span class="sig-paren">(</span><em class="sig-param">pairs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.regions_identical" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the regions in all objects in the list are identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pairs</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code>
objects</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if chromosome, start, and end are identical between
all regions in the same list positions.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.run_queued_filters">
<code class="sig-name descname">run_queued_filters</code><span class="sig-paren">(</span><em class="sig-param">log_progress=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.run_queued_filters" title="Permalink to this definition">¶</a></dt>
<dd><p>Run queued filters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>log_progress</strong> – If true, process iterating through all edges
will be continuously reported.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.scaling_factor">
<code class="sig-name descname">scaling_factor</code><span class="sig-paren">(</span><em class="sig-param">matrix</em>, <em class="sig-param">weight_column=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.scaling_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the scaling factor to another matrix.</p>
<p>Calculates the ratio between the number of contacts in
this Hic object to the number of contacts in another
Hic object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>matrix</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> object</p></li>
<li><p><strong>weight_column</strong> – Name of the column to calculate the scaling factor on</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.subset">
<code class="sig-name descname">subset</code><span class="sig-paren">(</span><em class="sig-param">*regions</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Subset a Hic object by specifying one or more subset regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>regions</strong> – string or GenomicRegion object(s)</p></li>
<li><p><strong>kwargs</strong> – Supports
file_name: destination file name of subset Hic object;
tmpdir: if True works in tmp until object is closed</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.to_bed">
<code class="sig-name descname">to_bed</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.to_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.to_bigwig">
<code class="sig-name descname">to_bigwig</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.to_bigwig" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionMatrixTable.to_gff">
<code class="sig-name descname">to_gff</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionMatrixTable.to_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.matrix.RegionPairsContainer">
<em class="property">class </em><code class="sig-prename descclassname">fanc.matrix.</code><code class="sig-name descname">RegionPairsContainer</code><a class="headerlink" href="#fanc.matrix.RegionPairsContainer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">genomic_regions.regions.RegionBased</span></code></p>
<p>Class representing pairs of genomic regions.</p>
<p>This is the basic interface for all pair and matrix classes in this module.
It inherits all methods from <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code>, and is
therefore based on a list of genomic regions (<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>)
representing the underlying genome. You can use the
<code class="xref py py-func docutils literal notranslate"><span class="pre">regions()</span></code> method to access genomic regions
in a intuitive fashion, for example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">region</span> <span class="ow">in</span> <span class="n">rpc</span><span class="o">.</span><span class="n">regions</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">):</span>
    <span class="c1"># do something with region</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">region</span><span class="p">)</span>
</pre></div>
</div>
<p>For more details on region access, see the <code class="code docutils literal notranslate"><span class="pre">genomic_regions</span></code>
documentation, on which this module is built.</p>
<p><a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a> adds methods for <em>pairs</em> of genomic regions
on top of the <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code> methods for individual
regions. In the nomenclature of this module, which borrows from network
analysis terminology, a pair of regions is represented by an <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># iterate over all region pairs / edges in chr1</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">rpc</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s2">&quot;chr1&quot;</span><span class="p">,</span> <span class="s2">&quot;chr1&quot;</span><span class="p">)):</span>
    <span class="c1"># do something with edge / region pair</span>
    <span class="n">region1</span> <span class="o">=</span> <span class="n">edge</span><span class="o">.</span><span class="n">source_region</span>
    <span class="n">region2</span> <span class="o">=</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink_region</span>
</pre></div>
</div>
<p>for more details see the <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> method help.</p>
<p>This class itself is only an interface and cannot actually be used to add
regions and region pairs. Implementations of this interface, i.e. subclasses
such as <a class="reference internal" href="#fanc.matrix.RegionPairsTable" title="fanc.matrix.RegionPairsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsTable</span></code></a> must override various hidden methods
to give them full functionality.</p>
<ul class="simple">
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code> is used to save region pairs / edges
to the object. It receives a single <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> as input and should
return the index of the added edge.</p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">_edges_iter()</span></code> is required by
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>. It is used to iterate over all
edges in the object in no particular order. It should return a generator
of <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> objects representing all region pairs in the object.</p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">_edges_subset()</span></code> is also used by
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>. It is used to iterate over a subset of
edges in this object. It receives as input a <code class="code docutils literal notranslate"><span class="pre">key</span></code> representing the requested
subset (further described in <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>), and
two lists of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects, <code class="code docutils literal notranslate"><span class="pre">row_regions</span></code> and
<code class="code docutils literal notranslate"><span class="pre">col_regions</span></code> representing the two dimensions of regions selected
by <code class="code docutils literal notranslate"><span class="pre">key</span></code>. It should return an iterator over <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> objects.</p></li>
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">_edges_getitem()</span></code> is used by
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> for retrieval of edges by bracket notation.
For integer input, it should return a single <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>, for <code class="xref py py-class docutils literal notranslate"><span class="pre">slice</span></code>
input a list of <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> objects.</p></li>
</ul>
<p>The above methods cover all the basic <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a> functionality,
but for speed improvements you may also want to override the following method,
which by default iterates over all edges</p>
<ul class="simple">
<li><p><code class="xref py py-func docutils literal notranslate"><span class="pre">_edges_length()</span></code> which returns the total number of
edges in the object</p></li>
</ul>
<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.add_contact">
<code class="sig-name descname">add_contact</code><span class="sig-paren">(</span><em class="sig-param">contact</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.add_contact" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.add_edge" title="fanc.matrix.RegionPairsContainer.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contact</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.add_contacts">
<code class="sig-name descname">add_contacts</code><span class="sig-paren">(</span><em class="sig-param">contacts</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.add_contacts" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#fanc.matrix.RegionPairsTable.add_edges" title="fanc.matrix.RegionPairsTable.add_edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edges()</span></code></a></p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.add_edge">
<code class="sig-name descname">add_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">check_nodes_exist=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.add_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an edge / contact between two regions to this object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>, dict with at least the
attributes source and sink, optionally weight,
or a list of length 2 (source, sink) or 3
(source, sink, weight).</p></li>
<li><p><strong>check_nodes_exist</strong> – Make sure that there are nodes
that match source and sink indexes</p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.add_edge_from_dict">
<code class="sig-name descname">add_edge_from_dict</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.add_edge_from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from dict input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – dict with at least the keys “source”
and “sink”. Additional keys will be loaded
as edge attributes</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.add_edge_from_edge">
<code class="sig-name descname">add_edge_from_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.add_edge_from_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.add_edge_from_list">
<code class="sig-name descname">add_edge_from_list</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.add_edge_from_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from list or tuple input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – List or tuple. Should be of length 2
(source, sink) or 3 (source, sink, weight)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.add_edge_simple">
<code class="sig-name descname">add_edge_simple</code><span class="sig-paren">(</span><em class="sig-param">source</em>, <em class="sig-param">sink</em>, <em class="sig-param">weight=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.add_edge_simple" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> – Source region index</p></li>
<li><p><strong>sink</strong> – Sink region index</p></li>
<li><p><strong>weight</strong> – Weight of the edge</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.add_edges">
<code class="sig-name descname">add_edges</code><span class="sig-paren">(</span><em class="sig-param">edges</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.add_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Bulk-add edges from a list.</p>
<p>List items can be any of the supported edge types,
list, tuple, dict, or <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>. Repeatedly
calls <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.add_edge" title="fanc.matrix.RegionPairsContainer.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a>, so
may be inefficient for large amounts of data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edges</strong> – List (or iterator) of edges. See
<a class="reference internal" href="#fanc.matrix.RegionMatrixTable.add_edge" title="fanc.matrix.RegionMatrixTable.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a>
for details</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.add_region">
<code class="sig-name descname">add_region</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.add_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.bin_intervals">
<em class="property">static </em><code class="sig-name descname">bin_intervals</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bins</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.bin_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.bin_intervals_equidistant">
<em class="property">static </em><code class="sig-name descname">bin_intervals_equidistant</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bin_size</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.bin_intervals_equidistant" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.binned_regions">
<code class="sig-name descname">binned_regions</code><span class="sig-paren">(</span><em class="sig-param">region=None</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.binned_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.chromosome_lengths">
<em class="property">property </em><code class="sig-name descname">chromosome_lengths</code><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.chromosome_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.chromosomes">
<code class="sig-name descname">chromosomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a list of chromosome names.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.edge_data">
<code class="sig-name descname">edge_data</code><span class="sig-paren">(</span><em class="sig-param">attribute</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.edge_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over specific edge attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> – Name of the attribute, e.g. “weight”</p></li>
<li><p><strong>args</strong> – Positional arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator over edge attribute</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.edge_subset">
<code class="sig-name descname">edge_subset</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.edge_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a subset of edges.</p>
<p>This is an alias for <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>generator (<a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.edges">
<em class="property">property </em><code class="sig-name descname">edges</code><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over contacts / edges.</p>
<p><a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> is the central function of
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>. Here, we will use the
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> implementation for demonstration purposes,
but the usage is exactly the same for all compatible
objects implementing <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>, including
<a class="reference internal" href="compatibility/juicer.html#fanc.compatibility.juicer.JuicerHic" title="fanc.compatibility.juicer.JuicerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">JuicerHic</span></code></a> and
<a class="reference internal" href="compatibility/cooler.html#fanc.compatibility.cooler.CoolerHic" title="fanc.compatibility.cooler.CoolerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">CoolerHic</span></code></a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>

<span class="c1"># file from FAN-C examples</span>
<span class="n">hic</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;output/hic/binned/fanc_example_1mb.hic&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can easily find the number of edges in the sample
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">)</span>  <span class="c1"># 8695</span>
</pre></div>
</div>
<p>When used in an iterator context, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
iterates over all edges in the <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Calling <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> as a method has the
same effect:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># note the &#39;()&#39;</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Rather than iterate over all edges in the object, we can select only a subset.
If the key is a string or a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, all non-zero edges connecting
the region described by the key to any other region are returned. If the key is a
tuple of strings or <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, only edges between the two regions
are returned.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># select all edges between chromosome 19</span>
<span class="c1"># and any other region:</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s2">&quot;chr19&quot;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>

<span class="c1"># select all edges that are only on</span>
<span class="c1"># chromosome 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr19&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 90--116; bias: 0.00021173151730025176; sink_node: chr19:37000001-38000000; source_node: chr19:11000001-12000000; weight: 0.009104455243910825</span>
    <span class="c1"># 135--135; bias: 0.00018003890596887822; sink_node: chr19:56000001-57000000; source_node: chr19:56000001-57000000; weight: 0.10028167062466517</span>
    <span class="c1"># 123--123; bias: 0.00011063368998965993; sink_node: chr19:44000001-45000000; source_node: chr19:44000001-45000000; weight: 0.1386240135570439</span>
    <span class="c1"># 92--93; bias: 0.00040851066434864896; sink_node: chr19:14000001-15000000; source_node: chr19:13000001-14000000; weight: 0.10090213409411629</span>
    <span class="c1"># ...</span>

<span class="c1"># select inter-chromosomal edges</span>
<span class="c1"># between chromosomes 18 and 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>By default, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> will retrieve all edge attributes,
which can be slow when iterating over a lot of edges. This is why all file-based FAN-C
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a> objects support lazy loading, where attributes
are only read on demand.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">,</span> <span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42 42 0.12291311562018173 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #0&gt;</span>
    <span class="c1"># 24 28 0.025205961072838057 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #1&gt;</span>
    <span class="c1"># 5 76 0.00961709840049876 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #2&gt;</span>
    <span class="c1"># 66 68 0.03876763316345468 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #3&gt;</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The lazy iterator reuses the <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> object in every iteration,
and overwrites the <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> attributes. Therefore <strong>do not</strong> use
lazy iterators if you need to store edge objects for later access.
For example, the following code works as expected
<code class="code docutils literal notranslate"><span class="pre">list(hic.edges())</span></code>, with all <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> objects stored in the
list, while this code <code class="code docutils literal notranslate"><span class="pre">list(hic.edges(lazy=True))</span></code>
will result in a list of identical <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> objects. Always ensure
you do all edge processing in the loop when working with lazy iterators!</p>
</div>
<p>When working with normalised contact frequencies, such as obtained through
matrix balancing in the example above, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
automatically returns normalised edge weights. In addition, the <code class="code docutils literal notranslate"><span class="pre">bias</span></code>
attribute will (typically) have a value different from 1.</p>
<p>When you are interested in the raw contact frequency, use the <code class="code docutils literal notranslate"><span class="pre">norm=False</span></code>
parameter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">)</span>
    <span class="c1"># 42 42 2120.0</span>
    <span class="c1"># 24 28 388.0</span>
    <span class="c1"># 5 76 94.0</span>
    <span class="c1"># 66 68 470.0</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>You can also choose to omit all intra- or inter-chromosomal edges using
<code class="code docutils literal notranslate"><span class="pre">intra_chromosomal=False</span></code> or <code class="code docutils literal notranslate"><span class="pre">inter_chromosomal=False</span></code>, respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Iterator over <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> or equivalent.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.edges_dict">
<code class="sig-name descname">edges_dict</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.edges_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Edges iterator with access by bracket notation.</p>
<p>This iterator <strong>always</strong> returns unnormalised edges.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict or dict-like iterator</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.find_region">
<code class="sig-name descname">find_region</code><span class="sig-paren">(</span><em class="sig-param">query_regions</em>, <em class="sig-param">_regions_dict=None</em>, <em class="sig-param">_region_ends=None</em>, <em class="sig-param">_chromosomes=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.find_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.intervals">
<code class="sig-name descname">intervals</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.mappable">
<code class="sig-name descname">mappable</code><span class="sig-paren">(</span><em class="sig-param">region=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.mappable" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the mappability of regions in this object.</p>
<p>A “mappable” region has at least one contact to another region
in the genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">array</span></code> where True means mappable
and False unmappable</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.merge">
<em class="property">classmethod </em><code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">pairs</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge two or more <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>pairs</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code> of <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a></p></li>
<li><p><strong>args</strong> – Positional arguments passed to constructor of this
class</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to constructor of this
class</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.region_bins">
<code class="sig-name descname">region_bins</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.region_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a genomic region and returns a slice of the bin
indices that are covered by the region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>slice</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.region_intervals">
<code class="sig-name descname">region_intervals</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">score_field='score'</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.region_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.region_subset">
<code class="sig-name descname">region_subset</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.region_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.regions">
<em class="property">property </em><code class="sig-name descname">regions</code><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.regions_and_edges">
<code class="sig-name descname">regions_and_edges</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.regions_and_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenient access to regions and edges selected by key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Edge selector, see <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>args</strong> – Positional arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of row regions, list of col regions, iterator over edges</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.regions_dict">
<em class="property">property </em><code class="sig-name descname">regions_dict</code><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.regions_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.regions_identical">
<em class="property">static </em><code class="sig-name descname">regions_identical</code><span class="sig-paren">(</span><em class="sig-param">pairs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.regions_identical" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the regions in all objects in the list are identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pairs</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code>
objects</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if chromosome, start, and end are identical between
all regions in the same list positions.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.to_bed">
<code class="sig-name descname">to_bed</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.to_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.to_bigwig">
<code class="sig-name descname">to_bigwig</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.to_bigwig" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsContainer.to_gff">
<code class="sig-name descname">to_gff</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsContainer.to_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.matrix.RegionPairsTable">
<em class="property">class </em><code class="sig-prename descclassname">fanc.matrix.</code><code class="sig-name descname">RegionPairsTable</code><span class="sig-paren">(</span><em class="sig-param">file_name=None</em>, <em class="sig-param">mode='a'</em>, <em class="sig-param">tmpdir=None</em>, <em class="sig-param">additional_region_fields=None</em>, <em class="sig-param">additional_edge_fields=None</em>, <em class="sig-param">partition_strategy='auto'</em>, <em class="sig-param">_table_name_regions='regions'</em>, <em class="sig-param">_table_name_edges='edges'</em>, <em class="sig-param">_edge_buffer_size='3G'</em>, <em class="sig-param">_edge_table_prefix='chrpair_'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.matrix.RegionPairsContainer</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.general.Maskable</span></code>, <a class="reference internal" href="regions.html#fanc.regions.RegionsTable" title="fanc.regions.RegionsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.regions.RegionsTable</span></code></a></p>
<p>HDF5 implementation of the <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a> interface.</p>
<dl class="class">
<dt id="fanc.matrix.RegionPairsTable.ChromosomeDescription">
<em class="property">class </em><code class="sig-name descname">ChromosomeDescription</code><a class="headerlink" href="#fanc.matrix.RegionPairsTable.ChromosomeDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="class">
<dt id="fanc.matrix.RegionPairsTable.MaskDescription">
<em class="property">class </em><code class="sig-name descname">MaskDescription</code><a class="headerlink" href="#fanc.matrix.RegionPairsTable.MaskDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
</dd></dl>

<dl class="class">
<dt id="fanc.matrix.RegionPairsTable.RegionDescription">
<em class="property">class </em><code class="sig-name descname">RegionDescription</code><a class="headerlink" href="#fanc.matrix.RegionPairsTable.RegionDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_contact">
<code class="sig-name descname">add_contact</code><span class="sig-paren">(</span><em class="sig-param">contact</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_contact" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.add_edge" title="fanc.matrix.RegionPairsContainer.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>contact</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_contacts">
<code class="sig-name descname">add_contacts</code><span class="sig-paren">(</span><em class="sig-param">contacts</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_contacts" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for <a class="reference internal" href="#fanc.matrix.RegionPairsTable.add_edges" title="fanc.matrix.RegionPairsTable.add_edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edges()</span></code></a></p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_edge">
<code class="sig-name descname">add_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">check_nodes_exist=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Add an edge / contact between two regions to this object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>, dict with at least the
attributes source and sink, optionally weight,
or a list of length 2 (source, sink) or 3
(source, sink, weight).</p></li>
<li><p><strong>check_nodes_exist</strong> – Make sure that there are nodes
that match source and sink indexes</p></li>
<li><p><strong>args</strong> – Positional arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to
<code class="xref py py-func docutils literal notranslate"><span class="pre">_add_edge()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_edge_from_dict">
<code class="sig-name descname">add_edge_from_dict</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_edge_from_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from dict input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – dict with at least the keys “source”
and “sink”. Additional keys will be loaded
as edge attributes</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_edge_from_edge">
<code class="sig-name descname">add_edge_from_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_edge_from_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_edge_from_list">
<code class="sig-name descname">add_edge_from_list</code><span class="sig-paren">(</span><em class="sig-param">edge</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_edge_from_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from list or tuple input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – List or tuple. Should be of length 2
(source, sink) or 3 (source, sink, weight)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_edge_simple">
<code class="sig-name descname">add_edge_simple</code><span class="sig-paren">(</span><em class="sig-param">source</em>, <em class="sig-param">sink</em>, <em class="sig-param">weight=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_edge_simple" title="Permalink to this definition">¶</a></dt>
<dd><p>Direct method to add an edge from <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> input.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source</strong> – Source region index</p></li>
<li><p><strong>sink</strong> – Sink region index</p></li>
<li><p><strong>weight</strong> – Weight of the edge</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_edges">
<code class="sig-name descname">add_edges</code><span class="sig-paren">(</span><em class="sig-param">edges</em>, <em class="sig-param">flush=True</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Bulk-add edges from a list.</p>
<p>List items can be any of the supported edge types,
list, tuple, dict, or <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>. Repeatedly
calls <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.add_edge" title="fanc.matrix.RegionPairsContainer.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a>, so
may be inefficient for large amounts of data.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edges</strong> – List (or iterator) of edges. See
<a class="reference internal" href="#fanc.matrix.RegionMatrixTable.add_edge" title="fanc.matrix.RegionMatrixTable.add_edge"><code class="xref py py-func docutils literal notranslate"><span class="pre">add_edge()</span></code></a>
for details</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_mask_description">
<code class="sig-name descname">add_mask_description</code><span class="sig-paren">(</span><em class="sig-param">name</em>, <em class="sig-param">description</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_mask_description" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a mask description to the _mask table and return its ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>name</strong> (<em>str</em>) – name of the mask</p></li>
<li><p><strong>description</strong> (<em>str</em>) – description of the mask</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>id of the mask</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_region">
<code class="sig-name descname">add_region</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.add_regions">
<code class="sig-name descname">add_regions</code><span class="sig-paren">(</span><em class="sig-param">regions</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.add_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.bin_intervals">
<em class="property">static </em><code class="sig-name descname">bin_intervals</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bins</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.bin_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.bin_intervals_equidistant">
<em class="property">static </em><code class="sig-name descname">bin_intervals_equidistant</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bin_size</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.bin_intervals_equidistant" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.bin_size">
<em class="property">property </em><code class="sig-name descname">bin_size</code><a class="headerlink" href="#fanc.matrix.RegionPairsTable.bin_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.binned_regions">
<code class="sig-name descname">binned_regions</code><span class="sig-paren">(</span><em class="sig-param">region=None</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.binned_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.bins_to_distance">
<code class="sig-name descname">bins_to_distance</code><span class="sig-paren">(</span><em class="sig-param">bins</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.bins_to_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.chromosome_bins">
<em class="property">property </em><code class="sig-name descname">chromosome_bins</code><a class="headerlink" href="#fanc.matrix.RegionPairsTable.chromosome_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.chromosome_lengths">
<em class="property">property </em><code class="sig-name descname">chromosome_lengths</code><a class="headerlink" href="#fanc.matrix.RegionPairsTable.chromosome_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.chromosomes">
<code class="sig-name descname">chromosomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.close">
<code class="sig-name descname">close</code><span class="sig-paren">(</span><em class="sig-param">copy_tmp=True</em>, <em class="sig-param">remove_tmp=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.distance_to_bins">
<code class="sig-name descname">distance_to_bins</code><span class="sig-paren">(</span><em class="sig-param">distance</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.distance_to_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.downsample">
<code class="sig-name descname">downsample</code><span class="sig-paren">(</span><em class="sig-param">n</em>, <em class="sig-param">file_name=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.downsample" title="Permalink to this definition">¶</a></dt>
<dd><p>Sample edges from this object.</p>
<p>Sampling is always done on uncorrected Hi-C matrices.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n</strong> – Sample size or reference object. If n &lt; 1 will be interpreted as
a fraction of total reads in this object.</p></li>
<li><p><strong>file_name</strong> – Output file name for down-sampled object.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.matrix.RegionPairsTable" title="fanc.matrix.RegionPairsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.edge_data">
<code class="sig-name descname">edge_data</code><span class="sig-paren">(</span><em class="sig-param">attribute</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.edge_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over specific edge attribute.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> – Name of the attribute, e.g. “weight”</p></li>
<li><p><strong>args</strong> – Positional arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator over edge attribute</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.edge_subset">
<code class="sig-name descname">edge_subset</code><span class="sig-paren">(</span><em class="sig-param">key=None</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.edge_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a subset of edges.</p>
<p>This is an alias for <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>generator (<a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.edges">
<em class="property">property </em><code class="sig-name descname">edges</code><a class="headerlink" href="#fanc.matrix.RegionPairsTable.edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over contacts / edges.</p>
<p><a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> is the central function of
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>. Here, we will use the
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> implementation for demonstration purposes,
but the usage is exactly the same for all compatible
objects implementing <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>, including
<a class="reference internal" href="compatibility/juicer.html#fanc.compatibility.juicer.JuicerHic" title="fanc.compatibility.juicer.JuicerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">JuicerHic</span></code></a> and
<a class="reference internal" href="compatibility/cooler.html#fanc.compatibility.cooler.CoolerHic" title="fanc.compatibility.cooler.CoolerHic"><code class="xref py py-class docutils literal notranslate"><span class="pre">CoolerHic</span></code></a>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>

<span class="c1"># file from FAN-C examples</span>
<span class="n">hic</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;output/hic/binned/fanc_example_1mb.hic&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>We can easily find the number of edges in the sample
<code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code> object:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="nb">len</span><span class="p">(</span><span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">)</span>  <span class="c1"># 8695</span>
</pre></div>
</div>
<p>When used in an iterator context, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
iterates over all edges in the <a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">:</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Calling <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> as a method has the
same effect:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># note the &#39;()&#39;</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">():</span>
    <span class="c1"># do something with edge</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42--42; bias: 5.797788472650082e-05; sink_node: chr18:42000001-43000000; source_node: chr18:42000001-43000000; weight: 0.12291311562018173</span>
    <span class="c1"># 24--28; bias: 6.496381719803623e-05; sink_node: chr18:28000001-29000000; source_node: chr18:24000001-25000000; weight: 0.025205961072838057</span>
    <span class="c1"># 5--76; bias: 0.00010230955745211447; sink_node: chr18:76000001-77000000; source_node: chr18:5000001-6000000; weight: 0.00961709840049876</span>
    <span class="c1"># 66--68; bias: 8.248432587969082e-05; sink_node: chr18:68000001-69000000; source_node: chr18:66000001-67000000; weight: 0.03876763316345468</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>Rather than iterate over all edges in the object, we can select only a subset.
If the key is a string or a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, all non-zero edges connecting
the region described by the key to any other region are returned. If the key is a
tuple of strings or <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, only edges between the two regions
are returned.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># select all edges between chromosome 19</span>
<span class="c1"># and any other region:</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s2">&quot;chr19&quot;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>

<span class="c1"># select all edges that are only on</span>
<span class="c1"># chromosome 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr19&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 90--116; bias: 0.00021173151730025176; sink_node: chr19:37000001-38000000; source_node: chr19:11000001-12000000; weight: 0.009104455243910825</span>
    <span class="c1"># 135--135; bias: 0.00018003890596887822; sink_node: chr19:56000001-57000000; source_node: chr19:56000001-57000000; weight: 0.10028167062466517</span>
    <span class="c1"># 123--123; bias: 0.00011063368998965993; sink_node: chr19:44000001-45000000; source_node: chr19:44000001-45000000; weight: 0.1386240135570439</span>
    <span class="c1"># 92--93; bias: 0.00040851066434864896; sink_node: chr19:14000001-15000000; source_node: chr19:13000001-14000000; weight: 0.10090213409411629</span>
    <span class="c1"># ...</span>

<span class="c1"># select inter-chromosomal edges</span>
<span class="c1"># between chromosomes 18 and 19</span>
<span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr19&#39;</span><span class="p">)):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 49--106; bias: 0.00026372303696871666; sink_node: chr19:27000001-28000000; source_node: chr18:49000001-50000000; weight: 0.003692122517562033</span>
    <span class="c1"># 6--82; bias: 0.00021923129703834945; sink_node: chr19:3000001-4000000; source_node: chr18:6000001-7000000; weight: 0.0008769251881533978</span>
    <span class="c1"># 47--107; bias: 0.00012820949175399097; sink_node: chr19:28000001-29000000; source_node: chr18:47000001-48000000; weight: 0.0015385139010478917</span>
    <span class="c1"># 38--112; bias: 0.0001493344481069762; sink_node: chr19:33000001-34000000; source_node: chr18:38000001-39000000; weight: 0.0005973377924279048</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>By default, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a> will retrieve all edge attributes,
which can be slow when iterating over a lot of edges. This is why all file-based FAN-C
<a class="reference internal" href="#fanc.matrix.RegionPairsContainer" title="fanc.matrix.RegionPairsContainer"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsContainer</span></code></a> objects support lazy loading, where attributes
are only read on demand.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">,</span> <span class="n">edge</span><span class="p">)</span>
    <span class="c1"># 42 42 0.12291311562018173 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #0&gt;</span>
    <span class="c1"># 24 28 0.025205961072838057 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #1&gt;</span>
    <span class="c1"># 5 76 0.00961709840049876 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #2&gt;</span>
    <span class="c1"># 66 68 0.03876763316345468 &lt;fanc.matrix.LazyEdge for row /edges/chrpair_0_0.row (Row), pointing to row #3&gt;</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The lazy iterator reuses the <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> object in every iteration,
and overwrites the <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> attributes. Therefore <strong>do not</strong> use
lazy iterators if you need to store edge objects for later access.
For example, the following code works as expected
<code class="code docutils literal notranslate"><span class="pre">list(hic.edges())</span></code>, with all <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> objects stored in the
list, while this code <code class="code docutils literal notranslate"><span class="pre">list(hic.edges(lazy=True))</span></code>
will result in a list of identical <a class="reference internal" href="#fanc.matrix.LazyEdge" title="fanc.matrix.LazyEdge"><code class="xref py py-class docutils literal notranslate"><span class="pre">LazyEdge</span></code></a> objects. Always ensure
you do all edge processing in the loop when working with lazy iterators!</p>
</div>
<p>When working with normalised contact frequencies, such as obtained through
matrix balancing in the example above, <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a>
automatically returns normalised edge weights. In addition, the <code class="code docutils literal notranslate"><span class="pre">bias</span></code>
attribute will (typically) have a value different from 1.</p>
<p>When you are interested in the raw contact frequency, use the <code class="code docutils literal notranslate"><span class="pre">norm=False</span></code>
parameter:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">edge</span> <span class="ow">in</span> <span class="n">hic</span><span class="o">.</span><span class="n">edges</span><span class="p">(</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="n">lazy</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">norm</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">edge</span><span class="o">.</span><span class="n">source</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">sink</span><span class="p">,</span> <span class="n">edge</span><span class="o">.</span><span class="n">weight</span><span class="p">)</span>
    <span class="c1"># 42 42 2120.0</span>
    <span class="c1"># 24 28 388.0</span>
    <span class="c1"># 5 76 94.0</span>
    <span class="c1"># 66 68 470.0</span>
    <span class="c1"># ...</span>
</pre></div>
</div>
<p>You can also choose to omit all intra- or inter-chromosomal edges using
<code class="code docutils literal notranslate"><span class="pre">intra_chromosomal=False</span></code> or <code class="code docutils literal notranslate"><span class="pre">inter_chromosomal=False</span></code>, respectively.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>Iterator over <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> or equivalent.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.edges_dict">
<code class="sig-name descname">edges_dict</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.edges_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Edges iterator with access by bracket notation.</p>
<p>This iterator <strong>always</strong> returns unnormalised edges.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict or dict-like iterator</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.filter">
<code class="sig-name descname">filter</code><span class="sig-paren">(</span><em class="sig-param">edge_filter</em>, <em class="sig-param">queue=False</em>, <em class="sig-param">log_progress=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter edges in this object by using a
<code class="xref py py-class docutils literal notranslate"><span class="pre">MaskFilter</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>edge_filter</strong> – Class implementing <code class="xref py py-class docutils literal notranslate"><span class="pre">MaskFilter</span></code>.</p></li>
<li><p><strong>queue</strong> – If True, filter will be queued and can be executed
along with other queued filters using
<a class="reference internal" href="#fanc.matrix.RegionPairsTable.run_queued_filters" title="fanc.matrix.RegionPairsTable.run_queued_filters"><code class="xref py py-func docutils literal notranslate"><span class="pre">run_queued_filters()</span></code></a></p></li>
<li><p><strong>log_progress</strong> – If true, process iterating through all edges
will be continuously reported.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.find_region">
<code class="sig-name descname">find_region</code><span class="sig-paren">(</span><em class="sig-param">query_regions</em>, <em class="sig-param">_regions_dict=None</em>, <em class="sig-param">_region_ends=None</em>, <em class="sig-param">_chromosomes=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.find_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.flush">
<code class="sig-name descname">flush</code><span class="sig-paren">(</span><em class="sig-param">silent=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.flush" title="Permalink to this definition">¶</a></dt>
<dd><p>Write data to file and flush buffers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>silent</strong> – do not print flush progress</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.get_mask">
<code class="sig-name descname">get_mask</code><span class="sig-paren">(</span><em class="sig-param">key</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.get_mask" title="Permalink to this definition">¶</a></dt>
<dd><p>Search _mask table for key and return Mask.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> (<em>int</em>) – search by mask name</p></li>
<li><p><strong>key</strong> – search by mask ID</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>Mask</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.get_masks">
<code class="sig-name descname">get_masks</code><span class="sig-paren">(</span><em class="sig-param">ix</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.get_masks" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract mask IDs encoded in parameter and return masks.</p>
<p>IDs are powers of 2, so a single int field in the table can hold
multiple masks by simply adding up the IDs. Similar principle to
UNIX chmod (although that uses base 8)</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>ix</strong> (<em>int</em>) – integer that is the sum of powers of 2. Note that this value
is not necessarily itself a power of 2.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of Masks extracted from ix</p>
</dd>
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p>list (Mask)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.intervals">
<code class="sig-name descname">intervals</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.mappable">
<code class="sig-name descname">mappable</code><span class="sig-paren">(</span><em class="sig-param">region=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.mappable" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the mappability of regions in this object.</p>
<p>A “mappable” region has at least one contact to another region
in the genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">array</span></code> where True means mappable
and False unmappable</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.merge">
<em class="property">classmethod </em><code class="sig-name descname">merge</code><span class="sig-paren">(</span><em class="sig-param">pairs</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.merge" title="Permalink to this definition">¶</a></dt>
<dd><p>Merge two or more <a class="reference internal" href="#fanc.matrix.RegionPairsTable" title="fanc.matrix.RegionPairsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsTable</span></code></a> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pairs</strong> – list of <a class="reference internal" href="#fanc.matrix.RegionPairsTable" title="fanc.matrix.RegionPairsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsTable</span></code></a></p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>merged <a class="reference internal" href="#fanc.matrix.RegionPairsTable" title="fanc.matrix.RegionPairsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionPairsTable</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.region_bins">
<code class="sig-name descname">region_bins</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.region_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.region_data">
<code class="sig-name descname">region_data</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">value=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.region_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.region_intervals">
<code class="sig-name descname">region_intervals</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">score_field='score'</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.region_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.region_subset">
<code class="sig-name descname">region_subset</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.region_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.regions">
<em class="property">property </em><code class="sig-name descname">regions</code><a class="headerlink" href="#fanc.matrix.RegionPairsTable.regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.regions_and_edges">
<code class="sig-name descname">regions_and_edges</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.regions_and_edges" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenient access to regions and edges selected by key.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Edge selector, see <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>args</strong> – Positional arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to <a class="reference internal" href="#fanc.matrix.RegionPairsContainer.edges" title="fanc.matrix.RegionPairsContainer.edges"><code class="xref py py-func docutils literal notranslate"><span class="pre">edges()</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of row regions, list of col regions, iterator over edges</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.regions_dict">
<em class="property">property </em><code class="sig-name descname">regions_dict</code><a class="headerlink" href="#fanc.matrix.RegionPairsTable.regions_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.regions_identical">
<em class="property">static </em><code class="sig-name descname">regions_identical</code><span class="sig-paren">(</span><em class="sig-param">pairs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.regions_identical" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the regions in all objects in the list are identical.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>pairs</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">list</span></code> of <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code>
objects</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>True if chromosome, start, and end are identical between
all regions in the same list positions.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.run_queued_filters">
<code class="sig-name descname">run_queued_filters</code><span class="sig-paren">(</span><em class="sig-param">log_progress=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.run_queued_filters" title="Permalink to this definition">¶</a></dt>
<dd><p>Run queued filters.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>log_progress</strong> – If true, process iterating through all edges
will be continuously reported.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.subset">
<code class="sig-name descname">subset</code><span class="sig-paren">(</span><em class="sig-param">*regions</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Subset a Hic object by specifying one or more subset regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>regions</strong> – string or GenomicRegion object(s)</p></li>
<li><p><strong>kwargs</strong> – Supports
file_name: destination file name of subset Hic object;
tmpdir: if True works in tmp until object is closed</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.to_bed">
<code class="sig-name descname">to_bed</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.to_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.to_bigwig">
<code class="sig-name descname">to_bigwig</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.to_bigwig" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.matrix.RegionPairsTable.to_gff">
<code class="sig-name descname">to_gff</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.RegionPairsTable.to_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="fanc.matrix.as_edge">
<code class="sig-prename descclassname">fanc.matrix.</code><code class="sig-name descname">as_edge</code><span class="sig-paren">(</span><em class="sig-param">edge</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.matrix.as_edge" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert input to <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>edge</strong> – Can be <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a>,
tuple or list of the form (source, sink, weight),
tuple of the form (<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>),
dict, or <a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a> equivalent</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.matrix.Edge" title="fanc.matrix.Edge"><code class="xref py py-class docutils literal notranslate"><span class="pre">Edge</span></code></a></p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="hic.html" class="btn btn-neutral float-right" title="Hic module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="pairs.html" class="btn btn-neutral float-left" title="Pairs module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Kai Kruse, Vaquerizas lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>