

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Regions module &mdash; FAN-C 0.9.6-beta documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Pairs module" href="pairs.html" />
    <link rel="prev" title="Map module" href="map.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> FAN-C
          

          
          </a>

          
            
            
              <div class="version">
                0.9.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting started with FAN-C</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fanc.html">fanc: command line tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fancplot.html">fancplot: plotting from the command line</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">Python API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../quickstart.html">Quickstart</a></li>
<li class="toctree-l2"><a class="reference internal" href="../compatibility.html">Working with Juicer and Cooler files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interfaces.html">Interfaces</a></li>
<li class="toctree-l2"><a class="reference internal" href="../generate.html">Generate Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../analyse.html">Analyse Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../plot.html">Plotting (API)</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../modules.html">Reference</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="init.html">init</a></li>
<li class="toctree-l3"><a class="reference internal" href="map.html">map</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">regions</a></li>
<li class="toctree-l3"><a class="reference internal" href="pairs.html">pairs</a></li>
<li class="toctree-l3"><a class="reference internal" href="matrix.html">matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="hic.html">hic</a></li>
<li class="toctree-l3"><a class="reference internal" href="peaks.html">peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="compatibility.html">compatibility</a></li>
<li class="toctree-l3"><a class="reference internal" href="architecture.html">architecture</a></li>
<li class="toctree-l3"><a class="reference internal" href="plotting.html">plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="tools.html">tools</a></li>
</ul>
</li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">FAN-C</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../api.html">fanc Python module</a> &raquo;</li>
        
          <li><a href="../modules.html">FAN-C Python modules</a> &raquo;</li>
        
      <li>Regions module</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/api/modules/regions.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="module-fanc.regions">
<span id="regions-module"></span><h1>Regions module<a class="headerlink" href="#module-fanc.regions" title="Permalink to this headline">¶</a></h1>
<p>Module for handling genomic regions such as chromosomes, bins, and restriction fragments.</p>
<p>The class <a class="reference internal" href="#fanc.regions.RegionsTable" title="fanc.regions.RegionsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionsTable</span></code></a> is an implementation of the <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code>
interface from the <code class="xref py py-mod docutils literal notranslate"><span class="pre">genomic_regions</span></code> package. More details on how to use the
<code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code> interface can be found in the <code class="xref py py-mod docutils literal notranslate"><span class="pre">genomic_regions</span></code>
documentation, but here is an example to get you started:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>
<span class="n">rt</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">RegionsTable</span><span class="p">()</span>

<span class="c1"># demo how to add regions</span>
<span class="n">regions</span> <span class="o">=</span> <span class="p">[]</span>
<span class="k">for</span> <span class="n">chromosome</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="s1">&#39;chr2&#39;</span><span class="p">]:</span>
    <span class="k">for</span> <span class="n">start</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">10000</span><span class="p">,</span> <span class="mi">1000</span><span class="p">):</span>
        <span class="n">r</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">GenomicRegion</span><span class="p">(</span><span class="n">chromosome</span><span class="o">=</span><span class="n">chromosome</span><span class="p">,</span> <span class="n">start</span><span class="o">=</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="o">=</span><span class="n">start</span><span class="o">+</span><span class="mi">999</span><span class="p">)</span>
        <span class="n">regions</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>
<span class="n">rt</span><span class="o">.</span><span class="n">add_regions</span><span class="p">(</span><span class="n">regions</span><span class="p">)</span>

<span class="c1"># query regions on chromosome 1</span>
<span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">rt</span><span class="o">.</span><span class="n">regions</span><span class="p">(</span><span class="s1">&#39;chr1&#39;</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>  <span class="c1"># chr1:1-1000, chr1:1001-2000, ..., chr1:9001-10000</span>
</pre></div>
</div>
<p>The class <a class="reference internal" href="#fanc.regions.Chromosome" title="fanc.regions.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a> holds chromosome information, specifically the
chromosome name in the reference genome, its length in base pairs, and its DNA
sequence. Multiple <a class="reference internal" href="#fanc.regions.Chromosome" title="fanc.regions.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a> objects can be grouped into a <a class="reference internal" href="#fanc.regions.Genome" title="fanc.regions.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a>,
which provides convenient access to its chromosomes and has useful functions for
in silico digestion and genome binning.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>

<span class="c1"># create chromosomes</span>
<span class="n">chromosome1</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">Chromosome</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;chr1&#39;</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="s1">&#39;AAGTCCGTGCTGTCGATCATAGCTAGCTAGCTA&#39;</span><span class="p">)</span>
<span class="n">chromosome2</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">Chromosome</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s1">&#39;chr2&#39;</span><span class="p">,</span> <span class="n">sequence</span><span class="o">=</span><span class="s1">&#39;GTGTCGATCAAATCGAAA&#39;</span><span class="p">)</span>
<span class="nb">len</span><span class="p">(</span><span class="n">chromosome1</span><span class="p">)</span>  <span class="c1"># 33</span>

<span class="c1"># create genome</span>
<span class="n">genome</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">Genome</span><span class="p">()</span>
<span class="n">genome</span><span class="o">.</span><span class="n">add_chromosome</span><span class="p">(</span><span class="n">chromosome1</span><span class="p">)</span>
<span class="n">genome</span><span class="o">.</span><span class="n">add_chromosome</span><span class="p">(</span><span class="n">chromosome2</span><span class="p">)</span>

<span class="c1"># in silico digestion</span>
<span class="n">restriction_fragments</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">get_regions</span><span class="p">(</span><span class="s1">&#39;MboI&#39;</span><span class="p">)</span>  <span class="c1"># cuts &#39;GATC&#39;</span>
<span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">restriction_fragments</span><span class="o">.</span><span class="n">regions</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>  <span class="c1"># chr1:1-15, chr1:16-33, chr2:1-6, chr2:7-18</span>

<span class="c1"># binning</span>
<span class="n">bins</span> <span class="o">=</span> <span class="n">genome</span><span class="o">.</span><span class="n">get_regions</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>  <span class="c1"># cuts &#39;GATC&#39;</span>
<span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">bins</span><span class="o">.</span><span class="n">regions</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">r</span><span class="p">)</span>  <span class="c1"># chr1:1-10, chr1:11-20, chr1:21-30, chr1:31-33, chr2:1-10, chr2:11-18</span>

<span class="c1"># load genome from file</span>
<span class="n">genome_from_file</span> <span class="o">=</span> <span class="n">Genome</span><span class="o">.</span><span class="n">from_string</span><span class="p">(</span><span class="s2">&quot;hg19_chr18_19.fa&quot;</span><span class="p">)</span>
<span class="n">genome_from_file</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">()</span> <span class="c1"># [&#39;chr18&#39;, &#39;chr19&#39;]</span>
</pre></div>
</div>
<dl class="class">
<dt id="fanc.regions.Chromosome">
<em class="property">class </em><code class="sig-prename descclassname">fanc.regions.</code><code class="sig-name descname">Chromosome</code><span class="sig-paren">(</span><em class="sig-param">name=None</em>, <em class="sig-param">length=None</em>, <em class="sig-param">sequence=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Chromosome data type.</p>
<dl class="attribute">
<dt id="fanc.regions.Chromosome.name">
<code class="sig-name descname">name</code><a class="headerlink" href="#fanc.regions.Chromosome.name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of the chromosome</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.regions.Chromosome.length">
<code class="sig-name descname">length</code><a class="headerlink" href="#fanc.regions.Chromosome.length" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of the chromosome in base-pairs</p>
</dd></dl>

<dl class="attribute">
<dt id="fanc.regions.Chromosome.sequence">
<code class="sig-name descname">sequence</code><a class="headerlink" href="#fanc.regions.Chromosome.sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Base-pair sequence of DNA in the chromosome</p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.Chromosome.from_fasta">
<em class="property">classmethod </em><code class="sig-name descname">from_fasta</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">name=None</em>, <em class="sig-param">include_sequence=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Chromosome.from_fasta" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a <a class="reference internal" href="#fanc.regions.Chromosome" title="fanc.regions.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a> from a FASTA file.</p>
<p>This class method will load a FASTA file and convert it into
a <a class="reference internal" href="#fanc.regions.Chromosome" title="fanc.regions.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a> object. If the FASTA file contains multiple
sequences, the output will be a list of <a class="reference internal" href="#fanc.regions.Chromosome" title="fanc.regions.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a> objects.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to the FASTA file</p></li>
<li><p><strong>name</strong> – Chromosome name. If None (default), will be read
from the FASTA file header</p></li>
<li><p><strong>include_sequence</strong> – If True (default), stores the chromosome
sequence in memory. Else, the sequence
attribute will be set to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.regions.Chromosome" title="fanc.regions.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a> if there is only a single FASTA
sequence in the file, list(<a class="reference internal" href="#fanc.regions.Chromosome" title="fanc.regions.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a>) if
there are multiple sequences.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.Chromosome.get_restriction_sites">
<code class="sig-name descname">get_restriction_sites</code><span class="sig-paren">(</span><em class="sig-param">restriction_enzyme</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Chromosome.get_restriction_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the restriction sites of a provided enzyme in this chromosome.</p>
<p>Internally uses Biopython to find RE sites.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>restriction_enzyme</strong> – The name of the restriction enzyme
(e.g. HindIII)</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>List of RE sites in base-pairs (1-based)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.regions.Genome">
<em class="property">class </em><code class="sig-prename descclassname">fanc.regions.</code><code class="sig-name descname">Genome</code><span class="sig-paren">(</span><em class="sig-param">file_name=None</em>, <em class="sig-param">chromosomes=None</em>, <em class="sig-param">mode='a'</em>, <em class="sig-param">tmpdir=None</em>, <em class="sig-param">_table_name_chromosomes='chromosomes'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Genome" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.general.FileGroup</span></code></p>
<p>Class representing a collection of chromosomes.</p>
<p>Provides some convenience batch
methods that call <a class="reference internal" href="#fanc.regions.Chromosome" title="fanc.regions.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a> methods for every
chromosome in this object.</p>
<dl class="class">
<dt id="fanc.regions.Genome.ChromosomeDefinition">
<em class="property">class </em><code class="sig-name descname">ChromosomeDefinition</code><a class="headerlink" href="#fanc.regions.Genome.ChromosomeDefinition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.Genome.add_chromosome">
<code class="sig-name descname">add_chromosome</code><span class="sig-paren">(</span><em class="sig-param">chromosome</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Genome.add_chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a <a class="reference internal" href="#fanc.regions.Chromosome" title="fanc.regions.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a> to this object.</p>
<p>Will choose suitable defaults for missing attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>chromosome</strong> – <a class="reference internal" href="#fanc.regions.Chromosome" title="fanc.regions.Chromosome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Chromosome</span></code></a> object or similar
object (e.g. dict) with the same fields</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.Genome.chromosomes">
<code class="sig-name descname">chromosomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Genome.chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>Get list of chromosomes in this object</p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.Genome.close">
<code class="sig-name descname">close</code><span class="sig-paren">(</span><em class="sig-param">copy_tmp=True</em>, <em class="sig-param">remove_tmp=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Genome.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.Genome.from_folder">
<em class="property">classmethod </em><code class="sig-name descname">from_folder</code><span class="sig-paren">(</span><em class="sig-param">folder_name</em>, <em class="sig-param">file_name=None</em>, <em class="sig-param">exclude=None</em>, <em class="sig-param">include_sequence=True</em>, <em class="sig-param">tmpdir=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Genome.from_folder" title="Permalink to this definition">¶</a></dt>
<dd><p>Load every FASTA file from a folder as a chromosome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>folder_name</strong> – Path to the folder to load</p></li>
<li><p><strong>file_name</strong> – File to save Genome object to</p></li>
<li><p><strong>exclude</strong> – List or set of chromosome names that
should NOT be loaded</p></li>
<li><p><strong>include_sequence</strong> – If True, will save the
chromosome sequences in the
Genome object</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.Genome.from_string">
<em class="property">classmethod </em><code class="sig-name descname">from_string</code><span class="sig-paren">(</span><em class="sig-param">genome_string</em>, <em class="sig-param">file_name=None</em>, <em class="sig-param">tmpdir=None</em>, <em class="sig-param">mode='a'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Genome.from_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Convenience function to load a <a class="reference internal" href="#fanc.regions.Genome" title="fanc.regions.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a> from a string.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome_string</strong> – Path to FASTA file, path to folder with
FASTA files, comma-separated list of
paths to FASTA files, path to HDF5 file</p></li>
<li><p><strong>file_name</strong> – Path to save file</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>A <a class="reference internal" href="#fanc.regions.Genome" title="fanc.regions.Genome"><code class="xref py py-class docutils literal notranslate"><span class="pre">Genome</span></code></a> object</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.Genome.get_regions">
<code class="sig-name descname">get_regions</code><span class="sig-paren">(</span><em class="sig-param">split</em>, <em class="sig-param">file_name=None</em>, <em class="sig-param">chromosomes=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Genome.get_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract genomic regions from genome.</p>
<p>Provides two options:</p>
<ul class="simple">
<li><p>Splits chromosomes at restriction sites if the split
parameter is the name of a restriction enzyme.</p></li>
<li><p>Splits chromosomes at equi-distant points if split
is an integer</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>split</strong> – Name of a restriction enzyme or positive
integer</p></li>
<li><p><strong>file_name</strong> – Name of a file if the result of this
method should be saved to file</p></li>
<li><p><strong>chromosomes</strong> – List of chromosome names to include. Default: all</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegions</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.Genome.sub_sequence">
<code class="sig-name descname">sub_sequence</code><span class="sig-paren">(</span><em class="sig-param">chromosome</em>, <em class="sig-param">start=None</em>, <em class="sig-param">end=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.Genome.sub_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Extract the chromosome DNA sequence between start and end.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>chromosome</strong> – Name of chromosome</p></li>
<li><p><strong>start</strong> – start position in bp (1-based, inclusive)</p></li>
<li><p><strong>end</strong> – end position in bp (1-based, inclusive)</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.regions.RegionsTable">
<em class="property">class </em><code class="sig-prename descclassname">fanc.regions.</code><code class="sig-name descname">RegionsTable</code><span class="sig-paren">(</span><em class="sig-param">file_name=None</em>, <em class="sig-param">mode='a'</em>, <em class="sig-param">tmpdir=None</em>, <em class="sig-param">additional_fields=None</em>, <em class="sig-param">_table_name_regions='regions'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.regions.RegionBasedWithBins</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">fanc.general.FileGroup</span></code></p>
<p>PyTables Table wrapper for storing genomic regions.</p>
<p>This class is inherited by objects working with lists of genomic
regions, such as equidistant bins along chromosomes in a genome
(<a class="reference internal" href="hic.html#fanc.hic.Hic" title="fanc.hic.Hic"><code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code></a>) or restriction fragments of genomic DNA
(<a class="reference internal" href="pairs.html#fanc.pairs.ReadPairs" title="fanc.pairs.ReadPairs"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairs</span></code></a>)</p>
<p>Internally, each genomic region is encoded in a PyTables Table and
the following region attributes are represented as table columns:
ix, chromosome, start, end, and strand. To add additional region
attributes, such as a score, use the “additional_fields” parameter
of the __init__ method. This must be a dict where the keys are str
and values are PyTables column descriptors, such as
<code class="xref py py-class docutils literal notranslate"><span class="pre">StringCol</span></code>. Example for adding a score field:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">fanc</span>
<span class="kn">import</span> <span class="nn">tables</span>
<span class="n">rt</span> <span class="o">=</span> <span class="n">fanc</span><span class="o">.</span><span class="n">RegionsTable</span><span class="p">(</span>
        <span class="n">additional_fields</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;score&#39;</span><span class="p">:</span> <span class="n">tables</span><span class="o">.</span><span class="n">Float32Col</span><span class="p">()}</span>
     <span class="p">)</span>
</pre></div>
</div>
<dl class="class">
<dt id="fanc.regions.RegionsTable.ChromosomeDescription">
<em class="property">class </em><code class="sig-name descname">ChromosomeDescription</code><a class="headerlink" href="#fanc.regions.RegionsTable.ChromosomeDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="class">
<dt id="fanc.regions.RegionsTable.RegionDescription">
<em class="property">class </em><code class="sig-name descname">RegionDescription</code><a class="headerlink" href="#fanc.regions.RegionsTable.RegionDescription" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">tables.description.IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.add_region">
<code class="sig-name descname">add_region</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.add_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.add_regions">
<code class="sig-name descname">add_regions</code><span class="sig-paren">(</span><em class="sig-param">regions</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.add_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<a class="reference internal" href="#fanc.regions.RegionsTable.add_region" title="fanc.regions.RegionsTable.add_region"><code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code></a> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.bin_intervals">
<em class="property">static </em><code class="sig-name descname">bin_intervals</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bins</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.bin_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.bin_intervals_equidistant">
<em class="property">static </em><code class="sig-name descname">bin_intervals_equidistant</code><span class="sig-paren">(</span><em class="sig-param">intervals</em>, <em class="sig-param">bin_size</em>, <em class="sig-param">interval_range=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.bin_intervals_equidistant" title="Permalink to this definition">¶</a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.bin_size">
<em class="property">property </em><code class="sig-name descname">bin_size</code><a class="headerlink" href="#fanc.regions.RegionsTable.bin_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.binned_regions">
<code class="sig-name descname">binned_regions</code><span class="sig-paren">(</span><em class="sig-param">region=None</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.binned_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.bins_to_distance">
<code class="sig-name descname">bins_to_distance</code><span class="sig-paren">(</span><em class="sig-param">bins</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.bins_to_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.chromosome_bins">
<em class="property">property </em><code class="sig-name descname">chromosome_bins</code><a class="headerlink" href="#fanc.regions.RegionsTable.chromosome_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.chromosome_lengths">
<em class="property">property </em><code class="sig-name descname">chromosome_lengths</code><a class="headerlink" href="#fanc.regions.RegionsTable.chromosome_lengths" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.chromosomes">
<code class="sig-name descname">chromosomes</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.chromosomes" title="Permalink to this definition">¶</a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.close">
<code class="sig-name descname">close</code><span class="sig-paren">(</span><em class="sig-param">copy_tmp=True</em>, <em class="sig-param">remove_tmp=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.close" title="Permalink to this definition">¶</a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.distance_to_bins">
<code class="sig-name descname">distance_to_bins</code><span class="sig-paren">(</span><em class="sig-param">distance</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.distance_to_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.find_region">
<code class="sig-name descname">find_region</code><span class="sig-paren">(</span><em class="sig-param">query_regions</em>, <em class="sig-param">_regions_dict=None</em>, <em class="sig-param">_region_ends=None</em>, <em class="sig-param">_chromosomes=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.find_region" title="Permalink to this definition">¶</a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.flush">
<code class="sig-name descname">flush</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.flush" title="Permalink to this definition">¶</a></dt>
<dd><p>Write buffered data to file.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.intervals">
<code class="sig-name descname">intervals</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.region_bins">
<code class="sig-name descname">region_bins</code><span class="sig-paren">(</span><em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.region_bins" title="Permalink to this definition">¶</a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.region_data">
<code class="sig-name descname">region_data</code><span class="sig-paren">(</span><em class="sig-param">key</em>, <em class="sig-param">value=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.region_data" title="Permalink to this definition">¶</a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.region_intervals">
<code class="sig-name descname">region_intervals</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">bins=None</em>, <em class="sig-param">bin_size=None</em>, <em class="sig-param">smoothing_window=None</em>, <em class="sig-param">nan_replacement=None</em>, <em class="sig-param">zero_to_nan=False</em>, <em class="sig-param">score_field='score'</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.region_intervals" title="Permalink to this definition">¶</a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.region_subset">
<code class="sig-name descname">region_subset</code><span class="sig-paren">(</span><em class="sig-param">region</em>, <em class="sig-param">*args</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.region_subset" title="Permalink to this definition">¶</a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.regions">
<em class="property">property </em><code class="sig-name descname">regions</code><a class="headerlink" href="#fanc.regions.RegionsTable.regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.regions_dict">
<em class="property">property </em><code class="sig-name descname">regions_dict</code><a class="headerlink" href="#fanc.regions.RegionsTable.regions_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.to_bed">
<code class="sig-name descname">to_bed</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.to_bed" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.to_bigwig">
<code class="sig-name descname">to_bigwig</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.to_bigwig" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.RegionsTable.to_gff">
<code class="sig-name descname">to_gff</code><span class="sig-paren">(</span><em class="sig-param">file_name</em>, <em class="sig-param">subset=None</em>, <em class="sig-param">**kwargs</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.RegionsTable.to_gff" title="Permalink to this definition">¶</a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="fanc.regions.LazyGenomicRegion">
<em class="property">class </em><code class="sig-prename descclassname">fanc.regions.</code><code class="sig-name descname">LazyGenomicRegion</code><span class="sig-paren">(</span><em class="sig-param">row</em>, <em class="sig-param">ix=None</em>, <em class="sig-param">auto_update=True</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">genomic_regions.regions.GenomicRegion</span></code></p>
<p>A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object with lazy attribute loading.</p>
<p>This class is central to an efficient retrieval of regions from
objects subclassing <a class="reference internal" href="#fanc.regions.RegionsTable" title="fanc.regions.RegionsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionsTable</span></code></a>. Its handling should
be mostly identical to <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, but attributes
will only be loaded on demand. Changes to attributes will change
the underlying row in the HDF5 regions table if the auto_update
parameter is set to True (default). Else you can manually call
<code class="xref py py-func docutils literal notranslate"><span class="pre">update()</span></code>.</p>
<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.as_bed_line">
<code class="sig-name descname">as_bed_line</code><span class="sig-paren">(</span><em class="sig-param">score_field='score'</em>, <em class="sig-param">name_field='name'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.as_bed_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of this object as line in a BED file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>score_field</strong> – name of the attribute to be used in the
‘score’ field of the BED line</p></li>
<li><p><strong>name_field</strong> – name of the attribute to be used in the
‘name’ field of the BED line</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.as_gff_line">
<code class="sig-name descname">as_gff_line</code><span class="sig-paren">(</span><em class="sig-param">source_field='source'</em>, <em class="sig-param">feature_field='feature'</em>, <em class="sig-param">score_field='score'</em>, <em class="sig-param">frame_field='frame'</em>, <em class="sig-param">float_format='.2e'</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.as_gff_line" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a representation of this object as line in a GFF file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>source_field</strong> – name of the attribute to be used in the
‘source’ field of the GFF line</p></li>
<li><p><strong>feature_field</strong> – name of the attribute to be used in the
‘feature’ field of the GFF line</p></li>
<li><p><strong>score_field</strong> – name of the attribute to be used in the
‘score’ field of the GFF line</p></li>
<li><p><strong>frame_field</strong> – name of the attribute to be used in the
‘frame’ field of the GFF line</p></li>
<li><p><strong>float_format</strong> – Formatting string for the float fields</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.attributes">
<em class="property">property </em><code class="sig-name descname">attributes</code><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.attributes" title="Permalink to this definition">¶</a></dt>
<dd><p>List of all attributes in this region.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.center">
<em class="property">property </em><code class="sig-name descname">center</code><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.center" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the center coordinate of the <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>float</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.contains">
<code class="sig-name descname">contains</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.contains" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the specified region is completely contained in this region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object or string</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.copy">
<code class="sig-name descname">copy</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.copy" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a (shallow) copy of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.expand">
<code class="sig-name descname">expand</code><span class="sig-paren">(</span><em class="sig-param">absolute=None</em>, <em class="sig-param">relative=None</em>, <em class="sig-param">absolute_left=0</em>, <em class="sig-param">absolute_right=0</em>, <em class="sig-param">relative_left=0.0</em>, <em class="sig-param">relative_right=0.0</em>, <em class="sig-param">copy=True</em>, <em class="sig-param">from_center=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.expand" title="Permalink to this definition">¶</a></dt>
<dd><p>Expand this region by a relative or an absolute amount.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>absolute</strong> – Absolute amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on both
sides. New region start will be
&lt;old start - absolute&gt;, new region end
will be &lt;old end + absolute&gt;</p></li>
<li><p><strong>relative</strong> – Relative amount as fraction of region by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on both
sides. New region start will be
&lt;old start - relative*len(self)&gt;,
new region end will be
&lt;old end + relative*(len(self)&gt;</p></li>
<li><p><strong>absolute_left</strong> – Absolute amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
left side</p></li>
<li><p><strong>absolute_right</strong> – Absolute amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
right side</p></li>
<li><p><strong>relative_left</strong> – Relative amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
left side</p></li>
<li><p><strong>relative_right</strong> – Relative amount in base pairs by which to
expand the region represented by this
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object on the
right side</p></li>
<li><p><strong>copy</strong> – If True, return a copy of the original region,
if False will modify the existing region in place</p></li>
<li><p><strong>from_center</strong> – If True measures distance from center rather
than start and end of the old region</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.five_prime">
<em class="property">property </em><code class="sig-name descname">five_prime</code><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.five_prime" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the position of the 5’ end of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
on the reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.fix_chromosome">
<code class="sig-name descname">fix_chromosome</code><span class="sig-paren">(</span><em class="sig-param">copy=False</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.fix_chromosome" title="Permalink to this definition">¶</a></dt>
<dd><p>Change chromosome representation from chr&lt;NN&gt; to &lt;NN&gt; or vice versa.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>copy</strong> – If True, make copy of region, otherwise
will modify existing region in place.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.from_string">
<em class="property">classmethod </em><code class="sig-name descname">from_string</code><span class="sig-paren">(</span><em class="sig-param">region_string</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.from_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a string into a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>.</p>
<p>This is a very useful convenience function to quickly
define a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object from a descriptor
string. Numbers can be abbreviated as ‘12k’, ‘1.5M’, etc.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region_string</strong> – A string of the form
&lt;chromosome&gt;[:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;]]
(with square brackets indicating optional
parts of the string). If any optional
part of the string is omitted, intuitive
defaults will be chosen.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.is_forward">
<code class="sig-name descname">is_forward</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.is_forward" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if this region is on the forward strand
of the reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if on ‘+’ strand, False otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.is_reverse">
<code class="sig-name descname">is_reverse</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.is_reverse" title="Permalink to this definition">¶</a></dt>
<dd><p>Return True if this region is on the reverse strand
of the reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>True if on ‘-‘ strand, False otherwise.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.ix">
<em class="property">property </em><code class="sig-name descname">ix</code><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.ix" title="Permalink to this definition">¶</a></dt>
<dd><p>Region index.</p>
<p>Location in underlying list of regions.</p>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.overlap">
<code class="sig-name descname">overlap</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.overlap" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the overlap in base pairs between this
region and another region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> to find overlap for</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>overlap as int in base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.overlaps">
<code class="sig-name descname">overlaps</code><span class="sig-paren">(</span><em class="sig-param">region</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.overlaps" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if this region overlaps with the specified region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object or string</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.set_attribute">
<code class="sig-name descname">set_attribute</code><span class="sig-paren">(</span><em class="sig-param">attribute</em>, <em class="sig-param">value</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.set_attribute" title="Permalink to this definition">¶</a></dt>
<dd><p>Safely set an attribute on the <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object.</p>
<p>This automatically decodes bytes objects into UTF-8 strings.
If you do not care about this, you can also use
region.&lt;attribute&gt; = &lt;value&gt; directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>attribute</strong> – Name of the attribute to be set</p></li>
<li><p><strong>value</strong> – Value of the attribute to be set</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.strand">
<em class="property">property </em><code class="sig-name descname">strand</code><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.strand" title="Permalink to this definition">¶</a></dt>
<dd><p>Strand this region is located on as int.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>1: forward strand, -1 reverse strand, 0 or
None: unknown.</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.strand_string">
<em class="property">property </em><code class="sig-name descname">strand_string</code><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.strand_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the ‘strand’ attribute as string.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>strand as str (‘+’, ‘-‘, or ‘.’)</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.three_prime">
<em class="property">property </em><code class="sig-name descname">three_prime</code><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.three_prime" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the position of the 3’ end of this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
on the reference.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="method">
<dt id="fanc.regions.LazyGenomicRegion.to_string">
<code class="sig-name descname">to_string</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.LazyGenomicRegion.to_string" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert this <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> to its string representation.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="function">
<dt id="fanc.regions.genome_regions">
<code class="sig-prename descclassname">fanc.regions.</code><code class="sig-name descname">genome_regions</code><span class="sig-paren">(</span><em class="sig-param">re_or_file</em>, <em class="sig-param">restriction_enzyme=None</em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.regions.genome_regions" title="Permalink to this definition">¶</a></dt>
<dd><p>Obtain RE fragments or bins of equal size from a reference genome.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>re_or_file</strong> – Path to or <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code> file with regions corresponding
to restriction fragments, or path to FASTA file with chromosomes</p></li>
<li><p><strong>restriction_enzyme</strong> – If the first argument is a FASTA file, you must
provide the name of a restriction enzyme here
for in silico genome digestion. You can also provide
an integer, in which case the genome will be binned into
regions of this size in bp.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.regions.RegionsTable" title="fanc.regions.RegionsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionsTable</span></code></a> of restriction fragments</p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="pairs.html" class="btn btn-neutral float-right" title="Pairs module" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="map.html" class="btn btn-neutral float-left" title="Map module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Kai Kruse, Vaquerizas lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>