<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Domains module &mdash; FAN-C 0.9.26b1-beta documentation</title>
      <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../../../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
        <script src="../../../_static/jquery.js"></script>
        <script src="../../../_static/underscore.js"></script>
        <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
        <script src="../../../_static/doctools.js"></script>
        <script src="../../../_static/sphinx_highlight.js"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Plotting API" href="../plotting.html" />
    <link rel="prev" title="Compartments module" href="compartments.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../../../index.html" class="icon icon-home"> FAN-C
          </a>
              <div class="version">
                0.9.26b1
              </div>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting started with FAN-C</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../fanc.html">fanc: command line tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../fancplot.html">fancplot: plotting from the command line</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../../api.html">Python API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../../quickstart.html">Quickstart</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../compatibility.html">Working with Juicer and Cooler files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../interfaces.html">Interfaces</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../generate.html">Generate Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../analyse.html">Analyse Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../plot.html">Plotting (API)</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../../modules.html">Reference</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="../init.html">init</a></li>
<li class="toctree-l3"><a class="reference internal" href="../map.html">map</a></li>
<li class="toctree-l3"><a class="reference internal" href="../regions.html">regions</a></li>
<li class="toctree-l3"><a class="reference internal" href="../pairs.html">pairs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../matrix.html">matrix</a></li>
<li class="toctree-l3"><a class="reference internal" href="../hic.html">hic</a></li>
<li class="toctree-l3"><a class="reference internal" href="../peaks.html">peaks</a></li>
<li class="toctree-l3"><a class="reference internal" href="../compatibility.html">compatibility</a></li>
<li class="toctree-l3 current"><a class="reference internal" href="../architecture.html">architecture</a><ul class="current">
<li class="toctree-l4"><a class="reference internal" href="aggregate.html">aggregate</a></li>
<li class="toctree-l4"><a class="reference internal" href="comparisons.html">comparisons</a></li>
<li class="toctree-l4"><a class="reference internal" href="compartments.html">compartments</a></li>
<li class="toctree-l4 current"><a class="current reference internal" href="#">domains</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../plotting.html">plotting</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tools.html">tools</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">FAN-C</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home"></a> &raquo;</li>
          <li><a href="../../api.html">fanc Python module</a> &raquo;</li>
          <li><a href="../../modules.html">FAN-C Python modules</a> &raquo;</li>
          <li><a href="../architecture.html">FAN-C Python architecture modules</a> &raquo;</li>
      <li>Domains module</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../../../_sources/api/modules/architecture/domains.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="module-fanc.architecture.domains">
<span id="domains-module"></span><h1>Domains module<a class="headerlink" href="#module-fanc.architecture.domains" title="Permalink to this heading"></a></h1>
<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fanc.architecture.domains.</span></span><span class="sig-name descname"><span class="pre">Boundaries</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.architecture.domains.RegionScoreTable" title="fanc.architecture.domains.RegionScoreTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionScoreTable</span></code></a></p>
<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.ChromosomeDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ChromosomeDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.ChromosomeDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.RegionDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">RegionDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.RegionDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.add_region">
<span class="sig-name descname"><span class="pre">add_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.add_region" title="Permalink to this definition"></a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.add_regions">
<span class="sig-name descname"><span class="pre">add_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.add_regions" title="Permalink to this definition"></a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.bin_intervals">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.bin_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.bin_intervals_equidistant">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals_equidistant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.bin_intervals_equidistant" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.bin_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_size</span></span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.bin_size" title="Permalink to this definition"></a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.binned_regions">
<span class="sig-name descname"><span class="pre">binned_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.binned_regions" title="Permalink to this definition"></a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.bins_to_distance">
<span class="sig-name descname"><span class="pre">bins_to_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.bins_to_distance" title="Permalink to this definition"></a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.chromosome_bins">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_bins</span></span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.chromosome_bins" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.chromosome_lengths">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_lengths</span></span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.chromosome_lengths" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.chromosomes">
<span class="sig-name descname"><span class="pre">chromosomes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.chromosomes" title="Permalink to this definition"></a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">copy_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.close" title="Permalink to this definition"></a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.distance_to_bins">
<span class="sig-name descname"><span class="pre">distance_to_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distance</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.distance_to_bins" title="Permalink to this definition"></a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.find_region">
<span class="sig-name descname"><span class="pre">find_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_regions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_regions_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_region_ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_chromosomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.find_region" title="Permalink to this definition"></a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.flush" title="Permalink to this definition"></a></dt>
<dd><p>Write buffered data to file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.from_insulation_score">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_insulation_score</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">insulation_score</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">delta_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sub_bin_precision</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">call_maxima</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'score'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.from_insulation_score" title="Permalink to this definition"></a></dt>
<dd><p>Call insulation boundaries based on minima in an insulation vector of this object.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>insulation_score</strong> – <a class="reference internal" href="#fanc.architecture.domains.InsulationScores" title="fanc.architecture.domains.InsulationScores"><code class="xref py py-class docutils literal notranslate"><span class="pre">InsulationScores</span></code></a> or
<a class="reference internal" href="#fanc.architecture.domains.InsulationScore" title="fanc.architecture.domains.InsulationScore"><code class="xref py py-class docutils literal notranslate"><span class="pre">InsulationScore</span></code></a> object</p></li>
<li><p><strong>window_size</strong> – Window size in base pairs. Only necessary for
<a class="reference internal" href="#fanc.architecture.domains.InsulationScores" title="fanc.architecture.domains.InsulationScores"><code class="xref py py-class docutils literal notranslate"><span class="pre">InsulationScores</span></code></a> objects</p></li>
<li><p><strong>min_score</strong> – Minimum difference between minimum and the closest maximum
in the insulation vector for a region to be considered a
boundary</p></li>
<li><p><strong>delta_window</strong> – Window size in bins to control smoothing of the delta function
used to calculate the derivative of the insulation index.
Calculation takes into account d bins upstream and d
bins downstream for a total window size of 2*d + 1 bins.</p></li>
<li><p><strong>log</strong> – Log2-transform insulation index before boundary calls</p></li>
<li><p><strong>sub_bin_precision</strong> – Call boundaries with sub bin precision, by taking
into account the precise zero transition of the delta vector.</p></li>
<li><p><strong>call_maxima</strong> – Call maxima instead of minima as boundaries</p></li>
<li><p><strong>score_field</strong> – </p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.intervals">
<span class="sig-name descname"><span class="pre">intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.intervals" title="Permalink to this definition"></a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.region_bins">
<span class="sig-name descname"><span class="pre">region_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.region_bins" title="Permalink to this definition"></a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.region_data">
<span class="sig-name descname"><span class="pre">region_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.region_data" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.region_intervals">
<span class="sig-name descname"><span class="pre">region_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'score'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.region_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.region_subset">
<span class="sig-name descname"><span class="pre">region_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.region_subset" title="Permalink to this definition"></a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.regions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions</span></span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.regions" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.regions_dict">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions_dict</span></span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.regions_dict" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.scores">
<span class="sig-name descname"><span class="pre">scores</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.scores" title="Permalink to this definition"></a></dt>
<dd><p>Return scores as list.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.to_bigwig">
<span class="sig-name descname"><span class="pre">to_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.to_bigwig" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.Boundaries.to_gff">
<span class="sig-name descname"><span class="pre">to_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.Boundaries.to_gff" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fanc.architecture.domains.</span></span><span class="sig-name descname"><span class="pre">DirectionalityIndex</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.architecture.domains.RegionScoreTable" title="fanc.architecture.domains.RegionScoreTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionScoreTable</span></code></a></p>
<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.ChromosomeDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ChromosomeDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.ChromosomeDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.RegionDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">RegionDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.RegionDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.add_region">
<span class="sig-name descname"><span class="pre">add_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.add_region" title="Permalink to this definition"></a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.add_regions">
<span class="sig-name descname"><span class="pre">add_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.add_regions" title="Permalink to this definition"></a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.bin_intervals">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.bin_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.bin_intervals_equidistant">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals_equidistant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.bin_intervals_equidistant" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.bin_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_size</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.bin_size" title="Permalink to this definition"></a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.binned_regions">
<span class="sig-name descname"><span class="pre">binned_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.binned_regions" title="Permalink to this definition"></a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.bins_to_distance">
<span class="sig-name descname"><span class="pre">bins_to_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.bins_to_distance" title="Permalink to this definition"></a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.chromosome_bins">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_bins</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.chromosome_bins" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.chromosome_lengths">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_lengths</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.chromosome_lengths" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.chromosomes">
<span class="sig-name descname"><span class="pre">chromosomes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.chromosomes" title="Permalink to this definition"></a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">copy_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.close" title="Permalink to this definition"></a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.distance_to_bins">
<span class="sig-name descname"><span class="pre">distance_to_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distance</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.distance_to_bins" title="Permalink to this definition"></a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.find_region">
<span class="sig-name descname"><span class="pre">find_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_regions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_regions_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_region_ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_chromosomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.find_region" title="Permalink to this definition"></a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.flush" title="Permalink to this definition"></a></dt>
<dd><p>Write buffered data to file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.intervals">
<span class="sig-name descname"><span class="pre">intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.intervals" title="Permalink to this definition"></a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.region_bins">
<span class="sig-name descname"><span class="pre">region_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.region_bins" title="Permalink to this definition"></a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.region_data">
<span class="sig-name descname"><span class="pre">region_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.region_data" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.region_intervals">
<span class="sig-name descname"><span class="pre">region_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'score'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.region_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.region_subset">
<span class="sig-name descname"><span class="pre">region_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.region_subset" title="Permalink to this definition"></a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.regions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.regions" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.regions_dict">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions_dict</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.regions_dict" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.scores">
<span class="sig-name descname"><span class="pre">scores</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.scores" title="Permalink to this definition"></a></dt>
<dd><p>Return scores as list.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.to_bigwig">
<span class="sig-name descname"><span class="pre">to_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.to_bigwig" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndex.to_gff">
<span class="sig-name descname"><span class="pre">to_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndex.to_gff" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fanc.architecture.domains.</span></span><span class="sig-name descname"><span class="pre">DirectionalityIndexes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.architecture.domains.RegionScoreParameterTable" title="fanc.architecture.domains.RegionScoreParameterTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionScoreParameterTable</span></code></a></p>
<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.ChromosomeDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ChromosomeDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.ChromosomeDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.RegionDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">RegionDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.RegionDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.add_region">
<span class="sig-name descname"><span class="pre">add_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.add_region" title="Permalink to this definition"></a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.add_regions">
<span class="sig-name descname"><span class="pre">add_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.add_regions" title="Permalink to this definition"></a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.bin_intervals">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.bin_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.bin_intervals_equidistant">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals_equidistant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.bin_intervals_equidistant" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.bin_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_size</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.bin_size" title="Permalink to this definition"></a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.binned_regions">
<span class="sig-name descname"><span class="pre">binned_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.binned_regions" title="Permalink to this definition"></a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.bins_to_distance">
<span class="sig-name descname"><span class="pre">bins_to_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.bins_to_distance" title="Permalink to this definition"></a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.chromosome_bins">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_bins</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.chromosome_bins" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.chromosome_lengths">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_lengths</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.chromosome_lengths" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.chromosomes">
<span class="sig-name descname"><span class="pre">chromosomes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.chromosomes" title="Permalink to this definition"></a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">copy_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.close" title="Permalink to this definition"></a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.distance_to_bins">
<span class="sig-name descname"><span class="pre">distance_to_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distance</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.distance_to_bins" title="Permalink to this definition"></a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.find_region">
<span class="sig-name descname"><span class="pre">find_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_regions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_regions_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_region_ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_chromosomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.find_region" title="Permalink to this definition"></a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.flush" title="Permalink to this definition"></a></dt>
<dd><p>Write buffered data to file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.from_hic">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_hic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hic</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_sizes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weight_field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tmpdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.from_hic" title="Permalink to this definition"></a></dt>
<dd><p>Compute the directionality index for multiple window sizes.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hic</strong> – A compatible Hi-C object</p></li>
<li><p><strong>window_sizes</strong> – A list of window sizes</p></li>
<li><p><strong>weight_field</strong> – Internal. Key of the weight attribute for an edge in this object.</p></li>
<li><p><strong>file_name</strong> – Path to output file. If not provided, will work in memory.</p></li>
<li><p><strong>tmpdir</strong> – Optional. If <code class="docutils literal notranslate"><span class="pre">True</span></code>, will work in temporary directory until
file is closed.</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed on to <a class="reference internal" href="../matrix.html#fanc.matrix.RegionMatrixTable.edges" title="fanc.matrix.RegionMatrixTable.edges"><code class="xref py py-class docutils literal notranslate"><span class="pre">edges</span></code></a></p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.architecture.domains.DirectionalityIndexes" title="fanc.architecture.domains.DirectionalityIndexes"><code class="xref py py-class docutils literal notranslate"><span class="pre">DirectionalityIndexes</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.intervals">
<span class="sig-name descname"><span class="pre">intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.intervals" title="Permalink to this definition"></a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.region_bins">
<span class="sig-name descname"><span class="pre">region_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.region_bins" title="Permalink to this definition"></a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.region_data">
<span class="sig-name descname"><span class="pre">region_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.region_data" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.region_intervals">
<span class="sig-name descname"><span class="pre">region_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'score'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.region_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.region_subset">
<span class="sig-name descname"><span class="pre">region_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.region_subset" title="Permalink to this definition"></a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.regions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.regions" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.regions_dict">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions_dict</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.regions_dict" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.score_regions">
<span class="sig-name descname"><span class="pre">score_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.score_regions" title="Permalink to this definition"></a></dt>
<dd><p>Construct a new object with regions that have a <code class="docutils literal notranslate"><span class="pre">score</span></code> attribute
which corresponds to scores calculated with this parameter.
:param parameter: Use scores calculated with this parameter (e.g. window size)
:param kwargs: Keyword arguments passed to <a class="reference internal" href="../regions.html#fanc.regions.RegionsTable" title="fanc.regions.RegionsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionsTable</span></code></a>
:return: <a class="reference internal" href="#fanc.architecture.domains.RegionScoreTable" title="fanc.architecture.domains.RegionScoreTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionScoreTable</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.scores">
<span class="sig-name descname"><span class="pre">scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.scores" title="Permalink to this definition"></a></dt>
<dd><p>Return scores for a specific parameter size as list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parameter</strong> – Parameter scores were calculated for, such as window size</p></li>
<li><p><strong>scores</strong> – If provided, set scores for this parameter to the ones in this list.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of scores</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Write scores to BED file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to output file</p></li>
<li><p><strong>parameter</strong> – Parameter the scores were calculated for, such as window size</p></li>
<li><p><strong>subset</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or region string specifying a
region range to be written to file, e.g. “chr19:1-1mb”</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.to_bigwig">
<span class="sig-name descname"><span class="pre">to_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.to_bigwig" title="Permalink to this definition"></a></dt>
<dd><p>Write scores to BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to output file</p></li>
<li><p><strong>parameter</strong> – Parameter the scores were calculated for, such as window size</p></li>
<li><p><strong>subset</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or region string specifying a
region range to be written to file, e.g. “chr19:1-1mb”</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.to_gff">
<span class="sig-name descname"><span class="pre">to_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.to_gff" title="Permalink to this definition"></a></dt>
<dd><p>Write scores to GFF file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to output file</p></li>
<li><p><strong>parameter</strong> – Parameter the scores were calculated for, such as window size</p></li>
<li><p><strong>subset</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or region string specifying a
region range to be written to file, e.g. “chr19:1-1mb”</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.DirectionalityIndexes.window_sizes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_sizes</span></span><a class="headerlink" href="#fanc.architecture.domains.DirectionalityIndexes.window_sizes" title="Permalink to this definition"></a></dt>
<dd><p>Get a list of window sizes in this object.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of window sizes (int)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fanc.architecture.domains.</span></span><span class="sig-name descname"><span class="pre">InsulationScore</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.architecture.domains.RegionScoreTable" title="fanc.architecture.domains.RegionScoreTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionScoreTable</span></code></a></p>
<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.ChromosomeDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ChromosomeDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.ChromosomeDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.RegionDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">RegionDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.RegionDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.add_region">
<span class="sig-name descname"><span class="pre">add_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.add_region" title="Permalink to this definition"></a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.add_regions">
<span class="sig-name descname"><span class="pre">add_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.add_regions" title="Permalink to this definition"></a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.bin_intervals">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.bin_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.bin_intervals_equidistant">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals_equidistant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.bin_intervals_equidistant" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.bin_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_size</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.bin_size" title="Permalink to this definition"></a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.binned_regions">
<span class="sig-name descname"><span class="pre">binned_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.binned_regions" title="Permalink to this definition"></a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.bins_to_distance">
<span class="sig-name descname"><span class="pre">bins_to_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.bins_to_distance" title="Permalink to this definition"></a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.chromosome_bins">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_bins</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.chromosome_bins" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.chromosome_lengths">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_lengths</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.chromosome_lengths" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.chromosomes">
<span class="sig-name descname"><span class="pre">chromosomes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.chromosomes" title="Permalink to this definition"></a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">copy_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.close" title="Permalink to this definition"></a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.distance_to_bins">
<span class="sig-name descname"><span class="pre">distance_to_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distance</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.distance_to_bins" title="Permalink to this definition"></a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.find_region">
<span class="sig-name descname"><span class="pre">find_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_regions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_regions_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_region_ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_chromosomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.find_region" title="Permalink to this definition"></a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.flush" title="Permalink to this definition"></a></dt>
<dd><p>Write buffered data to file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.intervals">
<span class="sig-name descname"><span class="pre">intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.intervals" title="Permalink to this definition"></a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.region_bins">
<span class="sig-name descname"><span class="pre">region_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.region_bins" title="Permalink to this definition"></a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.region_data">
<span class="sig-name descname"><span class="pre">region_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.region_data" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.region_intervals">
<span class="sig-name descname"><span class="pre">region_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'score'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.region_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.region_subset">
<span class="sig-name descname"><span class="pre">region_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.region_subset" title="Permalink to this definition"></a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.regions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.regions" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.regions_dict">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions_dict</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.regions_dict" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.scores">
<span class="sig-name descname"><span class="pre">scores</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.scores" title="Permalink to this definition"></a></dt>
<dd><p>Return scores as list.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.to_bigwig">
<span class="sig-name descname"><span class="pre">to_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.to_bigwig" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScore.to_gff">
<span class="sig-name descname"><span class="pre">to_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScore.to_gff" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fanc.architecture.domains.</span></span><span class="sig-name descname"><span class="pre">InsulationScores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.architecture.domains.RegionScoreParameterTable" title="fanc.architecture.domains.RegionScoreParameterTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionScoreParameterTable</span></code></a></p>
<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.ChromosomeDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ChromosomeDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.ChromosomeDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.RegionDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">RegionDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.RegionDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.add_region">
<span class="sig-name descname"><span class="pre">add_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.add_region" title="Permalink to this definition"></a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.add_regions">
<span class="sig-name descname"><span class="pre">add_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.add_regions" title="Permalink to this definition"></a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.bin_intervals">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.bin_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.bin_intervals_equidistant">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals_equidistant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.bin_intervals_equidistant" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.bin_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_size</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.bin_size" title="Permalink to this definition"></a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.binned_regions">
<span class="sig-name descname"><span class="pre">binned_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.binned_regions" title="Permalink to this definition"></a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.bins_to_distance">
<span class="sig-name descname"><span class="pre">bins_to_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.bins_to_distance" title="Permalink to this definition"></a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.chromosome_bins">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_bins</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.chromosome_bins" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.chromosome_lengths">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_lengths</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.chromosome_lengths" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.chromosomes">
<span class="sig-name descname"><span class="pre">chromosomes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.chromosomes" title="Permalink to this definition"></a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">copy_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.close" title="Permalink to this definition"></a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.distance_to_bins">
<span class="sig-name descname"><span class="pre">distance_to_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distance</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.distance_to_bins" title="Permalink to this definition"></a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.find_region">
<span class="sig-name descname"><span class="pre">find_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_regions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_regions_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_region_ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_chromosomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.find_region" title="Permalink to this definition"></a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.flush" title="Permalink to this definition"></a></dt>
<dd><p>Write buffered data to file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.from_hic">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_hic</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">hic</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_sizes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">window_offset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tmpdir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">impute_missing</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">na_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalise</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">normalisation_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trim_mean_proportion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">geometric_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subtract_mean</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">log</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.from_hic" title="Permalink to this definition"></a></dt>
<dd><p>Calculate insulation scores with multiple window sizes.</p>
<p>Insulation scores provide a great way to quantify the level of
interactions the cross each genomic region. It is calculated by summing up
(normalised) contacts in a square next to the diagonal for each genomic
region. Therefore, low scores correspond to highly insulated regions with
few interactions spanning them.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>hic</strong> – A Hi-C object</p></li>
<li><p><strong>window_sizes</strong> – A window size or list of window sizes used for the
sliding window</p></li>
<li><p><strong>window_offset</strong> – An offset of the sliding window in bins from the
diagonal</p></li>
<li><p><strong>file_name</strong> – Path to file where insulation scores are saved</p></li>
<li><p><strong>tmpdir</strong> – Optional. If <code class="docutils literal notranslate"><span class="pre">True</span></code>, will work with file in temporary
directory until it is closed</p></li>
<li><p><strong>impute_missing</strong> – Will replace missing / masked values in matrix
with their expected value prior to insulation
score calculation</p></li>
<li><p><strong>na_threshold</strong> – Fraction of missing values that is tolerated
in a sliding window before the score is set to NaN</p></li>
<li><p><strong>normalise</strong> – Normalise insulation score by dividing by chromosome mean
or mean of a sliding window if <code class="docutils literal notranslate"><span class="pre">normalisation_window</span></code>
is set.</p></li>
<li><p><strong>normalisation_window</strong> – If <code class="docutils literal notranslate"><span class="pre">None</span></code> (default), normalisation is performed
by dividing insulation scores by the chromosome mean.
You can set this to a number of bins to perform a more
local normalisation using average values in a window of
that size</p></li>
<li><p><strong>trim_mean_proportion</strong> – If &gt; 0 will use a trimmed mean for normalisation
trimming this fraction of scores before calculating the
mean. Use this if you expect outliers in insulation
scores</p></li>
<li><p><strong>geometric_mean</strong> – Use a geometric mean instead of arithmetic. If using
log-transformed, and if you intend to subtract scores
from different samples for comparison, this is
recommended</p></li>
<li><p><strong>subtract_mean</strong> – For normalisation, subtract mean instead of dividing by it.</p></li>
<li><p><strong>log</strong> – Log2-transform insulation scores after calculation. In the default
parameters, this makes scores roughly symmetrical around 0.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p><a class="reference internal" href="#fanc.architecture.domains.InsulationScores" title="fanc.architecture.domains.InsulationScores"><code class="xref py py-class docutils literal notranslate"><span class="pre">InsulationScores</span></code></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.intervals">
<span class="sig-name descname"><span class="pre">intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.intervals" title="Permalink to this definition"></a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.region_bins">
<span class="sig-name descname"><span class="pre">region_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.region_bins" title="Permalink to this definition"></a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.region_data">
<span class="sig-name descname"><span class="pre">region_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.region_data" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.region_intervals">
<span class="sig-name descname"><span class="pre">region_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'score'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.region_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.region_subset">
<span class="sig-name descname"><span class="pre">region_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.region_subset" title="Permalink to this definition"></a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.regions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.regions" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.regions_dict">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions_dict</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.regions_dict" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.score_regions">
<span class="sig-name descname"><span class="pre">score_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.score_regions" title="Permalink to this definition"></a></dt>
<dd><p>Construct a new object with regions that have a <code class="docutils literal notranslate"><span class="pre">score</span></code> attribute
which corresponds to scores calculated with this parameter.
:param parameter: Use scores calculated with this parameter (e.g. window size)
:param kwargs: Keyword arguments passed to <a class="reference internal" href="../regions.html#fanc.regions.RegionsTable" title="fanc.regions.RegionsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionsTable</span></code></a>
:return: <a class="reference internal" href="#fanc.architecture.domains.RegionScoreTable" title="fanc.architecture.domains.RegionScoreTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionScoreTable</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.scores">
<span class="sig-name descname"><span class="pre">scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.scores" title="Permalink to this definition"></a></dt>
<dd><p>Return scores for a specific parameter size as list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parameter</strong> – Parameter scores were calculated for, such as window size</p></li>
<li><p><strong>scores</strong> – If provided, set scores for this parameter to the ones in this list.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of scores</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Write scores to BED file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to output file</p></li>
<li><p><strong>parameter</strong> – Parameter the scores were calculated for, such as window size</p></li>
<li><p><strong>subset</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or region string specifying a
region range to be written to file, e.g. “chr19:1-1mb”</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.to_bigwig">
<span class="sig-name descname"><span class="pre">to_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.to_bigwig" title="Permalink to this definition"></a></dt>
<dd><p>Write scores to BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to output file</p></li>
<li><p><strong>parameter</strong> – Parameter the scores were calculated for, such as window size</p></li>
<li><p><strong>subset</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or region string specifying a
region range to be written to file, e.g. “chr19:1-1mb”</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.to_gff">
<span class="sig-name descname"><span class="pre">to_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.to_gff" title="Permalink to this definition"></a></dt>
<dd><p>Write scores to GFF file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to output file</p></li>
<li><p><strong>parameter</strong> – Parameter the scores were calculated for, such as window size</p></li>
<li><p><strong>subset</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or region string specifying a
region range to be written to file, e.g. “chr19:1-1mb”</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.InsulationScores.window_sizes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">window_sizes</span></span><a class="headerlink" href="#fanc.architecture.domains.InsulationScores.window_sizes" title="Permalink to this definition"></a></dt>
<dd><p>Get a list of window sizes in this object.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of window sizes (int)</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fanc.architecture.domains.</span></span><span class="sig-name descname"><span class="pre">RegionMultiScoreTable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">score_fields</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="../regions.html#fanc.regions.RegionsTable" title="fanc.regions.RegionsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionsTable</span></code></a></p>
<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.ChromosomeDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ChromosomeDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.ChromosomeDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.RegionDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">RegionDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.RegionDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.add_region">
<span class="sig-name descname"><span class="pre">add_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.add_region" title="Permalink to this definition"></a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.add_regions">
<span class="sig-name descname"><span class="pre">add_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.add_regions" title="Permalink to this definition"></a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.bin_intervals">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.bin_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.bin_intervals_equidistant">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals_equidistant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.bin_intervals_equidistant" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.bin_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_size</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.bin_size" title="Permalink to this definition"></a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.binned_regions">
<span class="sig-name descname"><span class="pre">binned_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.binned_regions" title="Permalink to this definition"></a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.bins_to_distance">
<span class="sig-name descname"><span class="pre">bins_to_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.bins_to_distance" title="Permalink to this definition"></a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.chromosome_bins">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_bins</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.chromosome_bins" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.chromosome_lengths">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_lengths</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.chromosome_lengths" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.chromosomes">
<span class="sig-name descname"><span class="pre">chromosomes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.chromosomes" title="Permalink to this definition"></a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">copy_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.close" title="Permalink to this definition"></a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.distance_to_bins">
<span class="sig-name descname"><span class="pre">distance_to_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distance</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.distance_to_bins" title="Permalink to this definition"></a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.find_region">
<span class="sig-name descname"><span class="pre">find_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_regions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_regions_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_region_ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_chromosomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.find_region" title="Permalink to this definition"></a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.flush" title="Permalink to this definition"></a></dt>
<dd><p>Write buffered data to file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.intervals">
<span class="sig-name descname"><span class="pre">intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.intervals" title="Permalink to this definition"></a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.region_bins">
<span class="sig-name descname"><span class="pre">region_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.region_bins" title="Permalink to this definition"></a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.region_data">
<span class="sig-name descname"><span class="pre">region_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.region_data" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.region_intervals">
<span class="sig-name descname"><span class="pre">region_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'score'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.region_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.region_subset">
<span class="sig-name descname"><span class="pre">region_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.region_subset" title="Permalink to this definition"></a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.regions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.regions" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.regions_dict">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions_dict</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.regions_dict" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.to_bigwig">
<span class="sig-name descname"><span class="pre">to_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.to_bigwig" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionMultiScoreTable.to_gff">
<span class="sig-name descname"><span class="pre">to_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionMultiScoreTable.to_gff" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fanc.architecture.domains.</span></span><span class="sig-name descname"><span class="pre">RegionScoreParameterTable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameter_values</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'score_'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#fanc.architecture.domains.RegionMultiScoreTable" title="fanc.architecture.domains.RegionMultiScoreTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionMultiScoreTable</span></code></a></p>
<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.ChromosomeDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ChromosomeDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.ChromosomeDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.RegionDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">RegionDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.RegionDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.add_region">
<span class="sig-name descname"><span class="pre">add_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.add_region" title="Permalink to this definition"></a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.add_regions">
<span class="sig-name descname"><span class="pre">add_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.add_regions" title="Permalink to this definition"></a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.bin_intervals">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.bin_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.bin_intervals_equidistant">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals_equidistant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.bin_intervals_equidistant" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.bin_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_size</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.bin_size" title="Permalink to this definition"></a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.binned_regions">
<span class="sig-name descname"><span class="pre">binned_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.binned_regions" title="Permalink to this definition"></a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.bins_to_distance">
<span class="sig-name descname"><span class="pre">bins_to_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.bins_to_distance" title="Permalink to this definition"></a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.chromosome_bins">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_bins</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.chromosome_bins" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.chromosome_lengths">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_lengths</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.chromosome_lengths" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.chromosomes">
<span class="sig-name descname"><span class="pre">chromosomes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.chromosomes" title="Permalink to this definition"></a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">copy_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.close" title="Permalink to this definition"></a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.distance_to_bins">
<span class="sig-name descname"><span class="pre">distance_to_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distance</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.distance_to_bins" title="Permalink to this definition"></a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.find_region">
<span class="sig-name descname"><span class="pre">find_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_regions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_regions_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_region_ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_chromosomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.find_region" title="Permalink to this definition"></a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.flush" title="Permalink to this definition"></a></dt>
<dd><p>Write buffered data to file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.intervals">
<span class="sig-name descname"><span class="pre">intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.intervals" title="Permalink to this definition"></a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.region_bins">
<span class="sig-name descname"><span class="pre">region_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.region_bins" title="Permalink to this definition"></a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.region_data">
<span class="sig-name descname"><span class="pre">region_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.region_data" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.region_intervals">
<span class="sig-name descname"><span class="pre">region_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'score'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.region_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.region_subset">
<span class="sig-name descname"><span class="pre">region_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.region_subset" title="Permalink to this definition"></a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.regions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.regions" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.regions_dict">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions_dict</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.regions_dict" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.score_regions">
<span class="sig-name descname"><span class="pre">score_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.score_regions" title="Permalink to this definition"></a></dt>
<dd><p>Construct a new object with regions that have a <code class="docutils literal notranslate"><span class="pre">score</span></code> attribute
which corresponds to scores calculated with this parameter.
:param parameter: Use scores calculated with this parameter (e.g. window size)
:param kwargs: Keyword arguments passed to <a class="reference internal" href="../regions.html#fanc.regions.RegionsTable" title="fanc.regions.RegionsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionsTable</span></code></a>
:return: <a class="reference internal" href="#fanc.architecture.domains.RegionScoreTable" title="fanc.architecture.domains.RegionScoreTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionScoreTable</span></code></a></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.scores">
<span class="sig-name descname"><span class="pre">scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scores</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.scores" title="Permalink to this definition"></a></dt>
<dd><p>Return scores for a specific parameter size as list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>parameter</strong> – Parameter scores were calculated for, such as window size</p></li>
<li><p><strong>scores</strong> – If provided, set scores for this parameter to the ones in this list.</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>list of scores</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Write scores to BED file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to output file</p></li>
<li><p><strong>parameter</strong> – Parameter the scores were calculated for, such as window size</p></li>
<li><p><strong>subset</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or region string specifying a
region range to be written to file, e.g. “chr19:1-1mb”</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.to_bigwig">
<span class="sig-name descname"><span class="pre">to_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.to_bigwig" title="Permalink to this definition"></a></dt>
<dd><p>Write scores to BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to output file</p></li>
<li><p><strong>parameter</strong> – Parameter the scores were calculated for, such as window size</p></li>
<li><p><strong>subset</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or region string specifying a
region range to be written to file, e.g. “chr19:1-1mb”</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreParameterTable.to_gff">
<span class="sig-name descname"><span class="pre">to_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">parameter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreParameterTable.to_gff" title="Permalink to this definition"></a></dt>
<dd><p>Write scores to GFF file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path to output file</p></li>
<li><p><strong>parameter</strong> – Parameter the scores were calculated for, such as window size</p></li>
<li><p><strong>subset</strong> – A <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or region string specifying a
region range to be written to file, e.g. “chr19:1-1mb”</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">fanc.architecture.domains.</span></span><span class="sig-name descname"><span class="pre">RegionScoreTable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="../regions.html#fanc.regions.RegionsTable" title="fanc.regions.RegionsTable"><code class="xref py py-class docutils literal notranslate"><span class="pre">RegionsTable</span></code></a></p>
<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.ChromosomeDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ChromosomeDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.ChromosomeDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of the chromosomes in this object.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.RegionDescription">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">RegionDescription</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.RegionDescription" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">IsDescription</span></code></p>
<p>Description of a genomic region for PyTables Table</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.add_region">
<span class="sig-name descname"><span class="pre">add_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.add_region" title="Permalink to this definition"></a></dt>
<dd><p>Add a genomic region to this object.</p>
<p>This method offers some flexibility in the types of objects
that can be loaded. See parameters for details.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – Can be a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, a str in the form
‘&lt;chromosome&gt;:&lt;start&gt;-&lt;end&gt;[:&lt;strand&gt;], a dict with
at least the fields ‘chromosome’, ‘start’, and
‘end’, optionally ‘ix’, or a list of length 3
(chromosome, start, end) or 4 (ix, chromosome,
start, end).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.add_regions">
<span class="sig-name descname"><span class="pre">add_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">regions</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.add_regions" title="Permalink to this definition"></a></dt>
<dd><p>Bulk insert multiple genomic regions.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>regions</strong> – List (or any iterator) with objects that
describe a genomic region. See
<code class="xref py py-class docutils literal notranslate"><span class="pre">add_region</span></code> for options.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.bin_intervals">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.bin_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into a fixed number of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.bin_intervals_equidistant">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_intervals_equidistant</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">intervals</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interval_range</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.bin_intervals_equidistant" title="Permalink to this definition"></a></dt>
<dd><p>Bin a given set of intervals into bins with a fixed size.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>intervals</strong> – iterator of tuples (start, end, score)</p></li>
<li><p><strong>bin_size</strong> – Size of each bin in base pairs</p></li>
<li><p><strong>interval_range</strong> – Optional. Tuple (start, end) in base pairs
of range of interval to be binned. Useful if
intervals argument does not cover to exact
genomic range to be binned.</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.bin_size">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">bin_size</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.bin_size" title="Permalink to this definition"></a></dt>
<dd><p>Return the length of the first region in the dataset.</p>
<p>Assumes all bins have equal size.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.binned_regions">
<span class="sig-name descname"><span class="pre">binned_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.binned_regions" title="Permalink to this definition"></a></dt>
<dd><p>Same as region_intervals, but returns <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>
objects instead of tuples.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> objects</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.bins_to_distance">
<span class="sig-name descname"><span class="pre">bins_to_distance</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bins</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.bins_to_distance" title="Permalink to this definition"></a></dt>
<dd><p>Convert fraction of bins to base pairs</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>bins</strong> – float, fraction of bins</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>int, base pairs</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.chromosome_bins">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_bins</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.chromosome_bins" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and the start
and end index of the bins they cover.</p>
<p>Returned list is range-compatible, i.e. chromosome
bins [0,5] cover chromosomes 1, 2, 3, and 4, not 5.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.chromosome_lengths">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">chromosome_lengths</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.chromosome_lengths" title="Permalink to this definition"></a></dt>
<dd><p>Returns a dictionary of chromosomes and their length
in bp.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.chromosomes">
<span class="sig-name descname"><span class="pre">chromosomes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.chromosomes" title="Permalink to this definition"></a></dt>
<dd><p>List all chromosomes in this regions table.
:return: list of chromosome names.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.close">
<span class="sig-name descname"><span class="pre">close</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">copy_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_tmp</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.close" title="Permalink to this definition"></a></dt>
<dd><p>Close this HDF5 file and run exit operations.</p>
<p>If file was opened with tmpdir in read-only mode:
close file and delete temporary copy.</p>
<p>If file was opened with tmpdir in write or append mode:
Replace original file with copy and delete copy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>copy_tmp</strong> – If False, does not overwrite original with modified file.</p></li>
<li><p><strong>remove_tmp</strong> – If False, does not delete temporary copy of file.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.distance_to_bins">
<span class="sig-name descname"><span class="pre">distance_to_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">distance</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.distance_to_bins" title="Permalink to this definition"></a></dt>
<dd><p>Convert base pairs to fraction of bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>distance</strong> – distance in base pairs</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>float, distance as fraction of bin size</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.find_region">
<span class="sig-name descname"><span class="pre">find_region</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">query_regions</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_regions_dict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_region_ends</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_chromosomes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.find_region" title="Permalink to this definition"></a></dt>
<dd><p>Find the region that is at the center of a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>query_regions</strong> – Region selector string, :class:~GenomicRegion, or
list of the former</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>index (or list of indexes) of the region at the center of the
query region</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.flush" title="Permalink to this definition"></a></dt>
<dd><p>Write buffered data to file.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.intervals">
<span class="sig-name descname"><span class="pre">intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.intervals" title="Permalink to this definition"></a></dt>
<dd><p>Alias for region_intervals.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.region_bins">
<span class="sig-name descname"><span class="pre">region_bins</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.region_bins" title="Permalink to this definition"></a></dt>
<dd><p>Return slice of start and end indices spanned by a region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>args</strong> – provide a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> here to get
the slice of start and end bins of onlythis region.
To get the slice over all regions leave this blank.</p>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.region_data">
<span class="sig-name descname"><span class="pre">region_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">value</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.region_data" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve or add vector-data to this object. If there is existing data in this
object with the same name, it will be replaced</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>key</strong> – Name of the data column</p></li>
<li><p><strong>value</strong> – vector with region-based data (one entry per region)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.region_intervals">
<span class="sig-name descname"><span class="pre">region_intervals</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bins</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">smoothing_window</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nan_replacement</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_to_nan</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_field</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'score'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.region_intervals" title="Permalink to this definition"></a></dt>
<dd><p>Return equally-sized genomic intervals and associated scores.</p>
<p>Use either bins or bin_size argument to control binning.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object denoting the region to be binned</p></li>
<li><p><strong>bins</strong> – Number of bins to divide the region into</p></li>
<li><p><strong>bin_size</strong> – Size of each bin (alternative to bins argument)</p></li>
<li><p><strong>smoothing_window</strong> – Size of window (in bins) to smooth scores
over</p></li>
<li><p><strong>nan_replacement</strong> – NaN values in the scores will be replaced
with this value</p></li>
<li><p><strong>zero_to_nan</strong> – If True, will convert bins with score 0 to NaN</p></li>
<li><p><strong>args</strong> – Arguments passed to _region_intervals</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments passed to _region_intervals</p></li>
</ul>
</dd>
<dt class="field-even">Returns</dt>
<dd class="field-even"><p>iterator of tuples: (start, end, score)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.region_subset">
<span class="sig-name descname"><span class="pre">region_subset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">region</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.region_subset" title="Permalink to this definition"></a></dt>
<dd><p>Takes a class:<cite>~GenomicRegion</cite> and returns all regions that
overlap with the supplied region.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><p><strong>region</strong> – String or class:<cite>~GenomicRegion</cite>
object for which covered bins will
be returned.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.regions">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.regions" title="Permalink to this definition"></a></dt>
<dd><p>Iterate over genomic regions in this object.</p>
<p>Will return a <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> object in every iteration.
Can also be used to get the number of regions by calling
len() on the object returned by this method.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>RegionIter</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.regions_dict">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">regions_dict</span></span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.regions_dict" title="Permalink to this definition"></a></dt>
<dd><p>Return a dictionary with region index as keys
and regions as values.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>dict {region.ix: region, …}</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.scores">
<span class="sig-name descname"><span class="pre">scores</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.scores" title="Permalink to this definition"></a></dt>
<dd><p>Return scores as list.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns</dt>
<dd class="field-odd"><p>list of float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.to_bed">
<span class="sig-name descname"><span class="pre">to_bed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.to_bed" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as BED file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bed()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.to_bigwig">
<span class="sig-name descname"><span class="pre">to_bigwig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.to_bigwig" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as BigWig file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_bigwig()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="fanc.architecture.domains.RegionScoreTable.to_gff">
<span class="sig-name descname"><span class="pre">to_gff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subset</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#fanc.architecture.domains.RegionScoreTable.to_gff" title="Permalink to this definition"></a></dt>
<dd><p>Export regions as GFF file</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>file_name</strong> – Path of file to write regions to</p></li>
<li><p><strong>subset</strong> – optional <code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code> or str to
write only regions overlapping this region</p></li>
<li><p><strong>kwargs</strong> – Passed to <code class="xref py py-func docutils literal notranslate"><span class="pre">write_gff()</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="compartments.html" class="btn btn-neutral float-left" title="Compartments module" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="../plotting.html" class="btn btn-neutral float-right" title="Plotting API" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2020, Kai Kruse, Vaquerizas lab.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>