

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Generating and filtering read pairs &mdash; FAN-C 0.9.6-beta documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="../../_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="../../" src="../../_static/documentation_options.js"></script>
        <script type="text/javascript" src="../../_static/jquery.js"></script>
        <script type="text/javascript" src="../../_static/underscore.js"></script>
        <script type="text/javascript" src="../../_static/doctools.js"></script>
        <script type="text/javascript" src="../../_static/language_data.js"></script>
    
    <script type="text/javascript" src="../../_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Generate, bin, filter, and correct Hic objects" href="hic.html" />
    <link rel="prev" title="Mapping FASTQ files" href="map.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="../../index.html" class="icon icon-home"> FAN-C
          

          
          </a>

          
            
            
              <div class="version">
                0.9.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting started with FAN-C</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fanc.html">fanc: command line tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../fancplot.html">fancplot: plotting from the command line</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="../api.html">Python API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../quickstart.html">Quickstart</a></li>
<li class="toctree-l2"><a class="reference internal" href="../compatibility.html">Working with Juicer and Cooler files</a></li>
<li class="toctree-l2"><a class="reference internal" href="../interfaces.html">Interfaces</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="../generate.html">Generate Hi-C matrices (API)</a><ul class="current">
<li class="toctree-l3"><a class="reference internal" href="map.html">Mapping FASTQ files</a></li>
<li class="toctree-l3 current"><a class="current reference internal" href="#">Generating and filtering ReadPairs</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#obtaining-fragments">Obtaining fragments</a></li>
<li class="toctree-l4"><a class="reference internal" href="#setting-up-filters-at-the-read-level">Setting up filters at the read level</a></li>
<li class="toctree-l4"><a class="reference internal" href="#importing-and-accessing-read-pairs">Importing and accessing read pairs</a></li>
<li class="toctree-l4"><a class="reference internal" href="#read-fragment-pair-filters">Read/fragment pair filters</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="hic.html">Generating, binning, and filtering Hic objects</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../analyse.html">Analyse Hi-C matrices (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../plot.html">Plotting (API)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../modules.html">Reference</a></li>
</ul>
</li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../index.html">FAN-C</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="../../index.html">Docs</a> &raquo;</li>
        
          <li><a href="../api.html">fanc Python module</a> &raquo;</li>
        
          <li><a href="../generate.html">Generate Hi-C matrices (API)</a> &raquo;</li>
        
      <li>Generating and filtering read pairs</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="../../_sources/api/generate/pairs.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="generating-and-filtering-read-pairs">
<span id="api-pairs"></span><h1>Generating and filtering read pairs<a class="headerlink" href="#generating-and-filtering-read-pairs" title="Permalink to this headline">¶</a></h1>
<p>Once we have mapped and sorted BAM files (see <a class="reference internal" href="map.html#api-map"><span class="std std-ref">Mapping FASTQ files</span></a>) we can load the paired reads
into a <a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReadPairs" title="fanc.pairs.ReadPairs"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairs</span></code></a> object.</p>
<div class="section" id="obtaining-fragments">
<h2>Obtaining fragments<a class="headerlink" href="#obtaining-fragments" title="Permalink to this headline">¶</a></h2>
<p><a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReadPairs" title="fanc.pairs.ReadPairs"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairs</span></code></a> map reads to restriction fragments, so first of all we are
going to need a list of those. The <a class="reference internal" href="../modules/regions.html#fanc.regions.genome_regions" title="fanc.regions.genome_regions"><code class="xref py py-func docutils literal notranslate"><span class="pre">fanc.regions.genome_regions()</span></code></a> function has been
made for that purpose. If you provide it with a FASTA file and a restriction enzyme name,
it will build the list of fragments for you:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">fanc.regions</span> <span class="kn">import</span> <span class="n">genome_regions</span>
<span class="n">genome_file</span> <span class="o">=</span> <span class="s1">&#39;hg19_chr18_19.fa&#39;</span>
<span class="n">fragments</span> <span class="o">=</span> <span class="n">genome_regions</span><span class="p">(</span><span class="n">genome_file</span><span class="p">,</span> <span class="n">restriction_enzyme</span><span class="o">=</span><span class="s1">&#39;HindIII&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can also give it a <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code> compatible file with fragments
directly (BED, GFF and similar work). Note that the fragments should cover the entire genome,
so don’t just use a list of restriction sites!</p>
<p>The returned object is of class <code class="xref py py-class docutils literal notranslate"><span class="pre">RegionBased</span></code>. See
<a class="reference internal" href="../interfaces/regions_interface.html#genomic-regions"><span class="std std-ref">the main article</span></a>
if you want to learn how to interact with objects of that class.</p>
</div>
<div class="section" id="setting-up-filters-at-the-read-level">
<h2>Setting up filters at the read level<a class="headerlink" href="#setting-up-filters-at-the-read-level" title="Permalink to this headline">¶</a></h2>
<p>Filters can be used to remove problematic reads, such as those with poor quality
(<a class="reference internal" href="../modules/pairs.html#fanc.pairs.QualityFilter" title="fanc.pairs.QualityFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">QualityFilter</span></code></a>) or with multiple mapping locations throughout
the genome (<a class="reference internal" href="../modules/pairs.html#fanc.pairs.UniquenessFilter" title="fanc.pairs.UniquenessFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">UniquenessFilter</span></code></a>). These filters (of class
<a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReadFilter" title="fanc.pairs.ReadFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadFilter</span></code></a>) have to be added already at the read pair import step,
as the necessary information is otherwise lost in later processing stages:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">fanc.pairs</span> <span class="kn">import</span> <span class="n">QualityFilter</span><span class="p">,</span> <span class="n">UniquenessFilter</span>
<span class="n">quality_filter</span> <span class="o">=</span> <span class="n">QualityFilter</span><span class="p">(</span><span class="mi">30</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="s1">&#39;MAPQ&#39;</span><span class="p">)</span>
<span class="n">uniqueness_filter</span> <span class="o">=</span> <span class="n">UniquenessFilter</span><span class="p">(</span><span class="n">strict</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="s1">&#39;unique&#39;</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Both filters are specific to reads mapped with <code class="docutils literal notranslate"><span class="pre">bowtie2</span></code>, and there are versions of
these filters specific for BWA-mapped reads (<a class="reference internal" href="../modules/pairs.html#fanc.pairs.BwaMemQualityFilter" title="fanc.pairs.BwaMemQualityFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">BwaMemQualityFilter</span></code></a>
and <a class="reference internal" href="../modules/pairs.html#fanc.pairs.BwaMemUniquenessFilter" title="fanc.pairs.BwaMemUniquenessFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">BwaMemUniquenessFilter</span></code></a>). Make sure you use the appropriate
filter for your chosen mapper, as the quality and uniqueness definitions used by each
application differ substantially!</p>
</div>
<p>The <a class="reference internal" href="../modules/pairs.html#fanc.pairs.QualityFilter" title="fanc.pairs.QualityFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">QualityFilter</span></code></a> removes all read pairs where one or both reads have
a MAPQ value lower than <code class="docutils literal notranslate"><span class="pre">30</span></code> (in this case). The <a class="reference internal" href="../modules/pairs.html#fanc.pairs.UniquenessFilter" title="fanc.pairs.UniquenessFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">UniquenessFilter</span></code></a>
has two settings. If <code class="docutils literal notranslate"><span class="pre">strict=True</span></code> it will remove any read pair where either of the two
reads has the <code class="docutils literal notranslate"><span class="pre">XS</span></code> tag, which indicates the presence of a secondary alignment. If
<code class="docutils literal notranslate"><span class="pre">strict=False</span></code>, it further requires that the secondary alignment score is higher or equal
to the primary alignment score (<code class="docutils literal notranslate"><span class="pre">AS</span></code>). The latter setting is therefore more permissive.</p>
<p>Now we have set up the necessary filters, we can start importing read pairs and match them
to restriction fragments.</p>
</div>
<div class="section" id="importing-and-accessing-read-pairs">
<h2>Importing and accessing read pairs<a class="headerlink" href="#importing-and-accessing-read-pairs" title="Permalink to this headline">¶</a></h2>
<p>For importing read pairs from SAM files we can use <a class="reference internal" href="../modules/pairs.html#fanc.pairs.generate_pairs_split" title="fanc.pairs.generate_pairs_split"><code class="xref py py-func docutils literal notranslate"><span class="pre">generate_pairs_split()</span></code></a>.
It requires the two paired-end SAM (or BAM) files and fragment definitions. In addition, we
provide it with the read filters we created above, so that only reads meeting our
requirements are imported into the object. If we do not pass a file name with <code class="docutils literal notranslate"><span class="pre">output_file</span></code>,
the resulting object is created in memory, which is useful for testing, but highly inadvisable
for actual Hi-C libraries. Here, we request four threads to be used for the loading in parallel:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">fanc.pairs</span> <span class="kn">import</span> <span class="n">generate_pairs_split</span> <span class="k">as</span> <span class="n">generate_pairs</span>
<span class="n">pairs_folder</span> <span class="o">=</span> <span class="n">mkdir</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_folder</span><span class="p">,</span> <span class="s1">&#39;pairs&#39;</span><span class="p">))</span>
<span class="n">pairs</span> <span class="o">=</span> <span class="n">generate_pairs</span><span class="p">(</span><span class="n">sam_1_file</span><span class="p">,</span> <span class="n">sam_2_file</span><span class="p">,</span> <span class="n">fragments</span><span class="p">,</span>
                       <span class="n">read_filters</span><span class="o">=</span><span class="p">(</span><span class="n">quality_filter</span><span class="p">,</span> <span class="n">uniqueness_filter</span><span class="p">),</span>
                       <span class="n">output_file</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pairs_folder</span><span class="p">,</span> <span class="s1">&#39;example.pairs&#39;</span><span class="p">),</span>
                       <span class="n">check_sorted</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">threads</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>
</pre></div>
</div>
<p>The resulting object is of the type <a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReadPairs" title="fanc.pairs.ReadPairs"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairs</span></code></a>, which implements
<a class="reference internal" href="../interfaces/edges_interface.html#edges-interface"><span class="std std-ref">RegionPairsContainer</span></a>. It contains all valid read pairs matched to their respective
restriction fragments.</p>
<p>Let’s look at some basic information:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">chromosomes</span> <span class="o">=</span> <span class="n">pairs</span><span class="o">.</span><span class="n">chromosomes</span><span class="p">()</span>
<span class="k">print</span><span class="p">(</span><span class="n">chromosomes</span><span class="p">)</span>
<span class="c1"># [&#39;chr18&#39;, &#39;chr19&#39;]</span>

<span class="n">pair</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span>
<span class="c1"># chr18: 3187827-(3191583[1])-3192106 -- chr18: 3187827-(3192073[-1])-3192106</span>
<span class="nb">type</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span>
<span class="c1"># fanc.pairs.FragmentReadPair</span>

<span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="o">.</span><span class="n">left</span><span class="p">)</span>
<span class="c1"># chr18: 3187827-(3191583[1])-3192106</span>

<span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="o">.</span><span class="n">right</span><span class="p">)</span>
<span class="c1"># chr18: 3187827-(3192073[-1])-3192106</span>
<span class="nb">type</span><span class="p">(</span><span class="n">pair</span><span class="o">.</span><span class="n">right</span><span class="p">)</span>
<span class="c1"># fanc.pairs.FragmentRead</span>
<span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="o">.</span><span class="n">right</span><span class="o">.</span><span class="n">fragment</span><span class="p">)</span>
<span class="c1"># chr18:3187827-3192106</span>
<span class="nb">type</span><span class="p">(</span><span class="n">pair</span><span class="o">.</span><span class="n">right</span><span class="o">.</span><span class="n">fragment</span><span class="p">)</span>
<span class="c1"># genomic_regions.regions.GenomicRegion</span>
<span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="o">.</span><span class="n">right</span><span class="o">.</span><span class="n">position</span><span class="p">)</span>
<span class="c1"># 3192073</span>
<span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="o">.</span><span class="n">right</span><span class="o">.</span><span class="n">re_distance</span><span class="p">())</span>
<span class="c1"># 33</span>
<span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="o">.</span><span class="n">right</span><span class="o">.</span><span class="n">strand</span><span class="p">)</span>
<span class="c1"># -1</span>
</pre></div>
</div>
<p>As you can see, <a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReadPairs" title="fanc.pairs.ReadPairs"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReadPairs</span></code></a> is a container for
<a class="reference internal" href="../modules/pairs.html#fanc.pairs.FragmentReadPair" title="fanc.pairs.FragmentReadPair"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentReadPair</span></code></a> objects. Each object contains information about its
two associated reads in the <code class="xref py py-attr docutils literal notranslate"><span class="pre">left</span></code> and
<code class="xref py py-attr docutils literal notranslate"><span class="pre">right</span></code> properties, respectively, which return a
<a class="reference internal" href="../modules/pairs.html#fanc.pairs.FragmentRead" title="fanc.pairs.FragmentRead"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentRead</span></code></a> object.</p>
<p><a class="reference internal" href="../modules/pairs.html#fanc.pairs.FragmentRead" title="fanc.pairs.FragmentRead"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentRead</span></code></a> objects contain information about the read’s mapping
location (<code class="docutils literal notranslate"><span class="pre">pair.right.position</span></code> and <code class="docutils literal notranslate"><span class="pre">pair.right.strand</span></code>), and the fragment it maps to
(<code class="docutils literal notranslate"><span class="pre">pair.right.fragment</span></code>). For convenience, you can also directly calculate the distance of
the read to the end of the fragment (which is the nearest restriction site) using
<code class="docutils literal notranslate"><span class="pre">pair.right.re_distance()</span></code>. The <code class="docutils literal notranslate"><span class="pre">pair.right.fragment</span></code> attribute is of type
<code class="xref py py-class docutils literal notranslate"><span class="pre">GenomicRegion</span></code>, which has lots of useful properties which
you can find in the <a class="reference external" href="https://github.com/vaquerizaslab/genomic_regions">genomic_regions documentation</a>.
Incidentally, the above pair is a self-ligated” fragment, which will be filtered out
in post-processing.</p>
<p>Typically you would not access each pair individually, but iterate over all pairs, or pair
subsets with the <a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReadPairs.pairs" title="fanc.pairs.ReadPairs.pairs"><code class="xref py py-func docutils literal notranslate"><span class="pre">fanc.pairs.ReadPairs.pairs()</span></code></a> function:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># all pairs</span>
<span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">pairs</span><span class="o">.</span><span class="n">pairs</span><span class="p">:</span>
    <span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span>
<span class="c1"># pair subset (only pairs with both fragments on chromosome 18)</span>
<span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">pairs</span><span class="o">.</span><span class="n">pairs</span><span class="p">((</span><span class="s1">&#39;chr18&#39;</span><span class="p">,</span> <span class="s1">&#39;chr18&#39;</span><span class="p">)):</span>
    <span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span>
</pre></div>
</div>
<p>The <a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReadPairs.pairs" title="fanc.pairs.ReadPairs.pairs"><code class="xref py py-func docutils literal notranslate"><span class="pre">fanc.pairs.ReadPairs.pairs()</span></code></a> function supports <a class="reference internal" href="../interfaces/edges_interface.html#lazy-evaluation"><span class="std std-ref">lazy evaluation</span></a>
with the <code class="docutils literal notranslate"><span class="pre">lazy=True</span></code> argument, but make sure to read the caveats in the main article!</p>
</div>
<div class="section" id="read-fragment-pair-filters">
<h2>Read/fragment pair filters<a class="headerlink" href="#read-fragment-pair-filters" title="Permalink to this headline">¶</a></h2>
<p>In addition to the quality and uniqueness filters, which had to be used during read pair import,
FAN-C offers a number of <a class="reference internal" href="../modules/pairs.html#fanc.pairs.FragmentReadPairFilter" title="fanc.pairs.FragmentReadPairFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">FragmentReadPairFilter</span></code></a> to remove problematic pairs.
First of all, however, we’ll introduce one little function that you can use to reset all filters
(and restore the original pairs) in case you found your filtering t be too harsh or if you want
a fresh start, which is called <code class="xref py py-func docutils literal notranslate"><span class="pre">reset_filters()</span></code>:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pairs</span><span class="o">.</span><span class="n">reset_filters</span><span class="p">()</span>
</pre></div>
</div>
<p>We will illustrate how filters are used with the <a class="reference internal" href="../modules/pairs.html#fanc.pairs.SelfLigationFilter" title="fanc.pairs.SelfLigationFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">SelfLigationFilter</span></code></a>.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">fanc.pairs</span> <span class="kn">import</span> <span class="n">SelfLigationFilter</span>
<span class="n">sl_filter</span> <span class="o">=</span> <span class="n">SelfLigationFilter</span><span class="p">(</span><span class="n">mask</span><span class="o">=</span><span class="s1">&#39;self-ligation&#39;</span><span class="p">)</span>
<span class="c1"># alternative:</span>
<span class="c1"># from fanc.general import Mask</span>
<span class="c1"># sl_filter = SelfLigationFilter(mask=Mask(name=&#39;self-ligation&#39;, description=&quot;Filter for self-ligated fragments&quot;)</span>
<span class="n">pairs</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">sl_filter</span><span class="p">)</span>
</pre></div>
</div>
<p>In general, we first create a filter instance. The <code class="docutils literal notranslate"><span class="pre">mask</span></code> parameter allows us to specify a
filter name. We can also pass it a more elaborate <code class="xref py py-class docutils literal notranslate"><span class="pre">Mask</span></code> object with
properties for name and description (see commented code). Name and description of the mask will
be stored in the <code class="docutils literal notranslate"><span class="pre">pairs</span></code> object upon filtering. The filter is run simply by calling
<a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReadPairs.filter" title="fanc.pairs.ReadPairs.filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">filter()</span></code></a> with the corresponding filter instance.</p>
<p>If you want to run multiple filters, it is more efficient to “queue” filters and then execute
them all in one go:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">fanc.pairs</span> <span class="kn">import</span> <span class="n">SelfLigationFilter</span><span class="p">,</span> <span class="n">ReDistanceFilter</span>
<span class="n">sl_filter</span> <span class="o">=</span> <span class="n">SelfLigationFilter</span><span class="p">(</span><span class="n">mask</span><span class="o">=</span><span class="s1">&#39;self-ligation&#39;</span><span class="p">)</span>
<span class="n">rd_filter</span> <span class="o">=</span> <span class="n">ReDistanceFilter</span><span class="p">(</span><span class="mi">500</span><span class="p">,</span> <span class="n">mask</span><span class="o">=</span><span class="s1">&#39;re-site-distance&#39;</span><span class="p">)</span>
<span class="n">pairs</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">sl_filter</span><span class="p">,</span> <span class="n">queue</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">pairs</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">rd_filter</span><span class="p">,</span> <span class="n">queue</span><span class="o">=</span><span class="bp">True</span><span class="p">)</span>
<span class="n">pairs</span><span class="o">.</span><span class="n">run_queued_filters</span><span class="p">()</span>
</pre></div>
</div>
<p>Once the <a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReadPairs.filter" title="fanc.pairs.ReadPairs.filter"><code class="xref py py-func docutils literal notranslate"><span class="pre">filter()</span></code></a> command has completed, the filtered pairs will
appear to have been deleted from the object.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">pair</span> <span class="o">=</span> <span class="n">pairs</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span>
<span class="c1"># chr18: 899140-(899308[-1])-899476 -- chr18: 1509911-(1510021[1])-1510076</span>
</pre></div>
</div>
<p>In reality, filtering only masks (“hides”) edges, so we have the opportunity to reset filters
or selectively disable them when iterating:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">pair</span> <span class="ow">in</span> <span class="n">pairs</span><span class="o">.</span><span class="n">pairs</span><span class="p">(</span><span class="n">excluded_filters</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;self-ligation&#39;</span><span class="p">]):</span>
    <span class="k">print</span><span class="p">(</span><span class="n">pair</span><span class="p">)</span>
</pre></div>
</div>
<p>To obtain a dictionary with the filter statistics, use <a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReadPairs.filter_statistics" title="fanc.pairs.ReadPairs.filter_statistics"><code class="xref py py-func docutils literal notranslate"><span class="pre">filter_statistics()</span></code></a></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="n">statistics</span> <span class="o">=</span> <span class="n">pairs</span><span class="o">.</span><span class="n">filter_statistics</span><span class="p">()</span>
</pre></div>
</div>
<div class="section" id="filter-types">
<h3>Filter types<a class="headerlink" href="#filter-types" title="Permalink to this headline">¶</a></h3>
<p>Here are the filters available in FAN-C for read/fragment pairs:</p>
<ul class="simple">
<li><p><a class="reference internal" href="../modules/pairs.html#fanc.pairs.SelfLigationFilter" title="fanc.pairs.SelfLigationFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">SelfLigationFilter</span></code></a> removes all pairs where both reads map to the same fragment</p></li>
<li><p><a class="reference internal" href="../modules/pairs.html#fanc.pairs.ReDistanceFilter" title="fanc.pairs.ReDistanceFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">ReDistanceFilter</span></code></a> can filter for an expected DNA (not restriction) fragment
size in Hi-C libraries that result from fragmentation (typically sonication). It sums up the
distance of both reads to their nearest restriction site. If that sum exceeds the cutoff set at
filter creation, it will be marked as invalid</p></li>
<li><p><a class="reference internal" href="../modules/pairs.html#fanc.pairs.PCRDuplicateFilter" title="fanc.pairs.PCRDuplicateFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">PCRDuplicateFilter</span></code></a> will find pairs mapping to the exact same genomic locations
(both reads), and will only retain one copy of the exact duplicates.</p></li>
<li><p><a class="reference internal" href="../modules/pairs.html#fanc.pairs.InwardPairsFilter" title="fanc.pairs.InwardPairsFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">InwardPairsFilter</span></code></a> and <a class="reference internal" href="../modules/pairs.html#fanc.pairs.OutwardPairsFilter" title="fanc.pairs.OutwardPairsFilter"><code class="xref py py-class docutils literal notranslate"><span class="pre">OutwardPairsFilter</span></code></a> are removing
pairs where both reads map within a specific distance on the same chromosome and are oriented
towards or away from each other (in terms of strand), respectively. This is designed to remove
ligation products that have likely arisen from uncut restriction sites or that are unligated.</p></li>
</ul>
<p>Next, we will convert the filtered pairs to a <a class="reference internal" href="../modules/hic.html#fanc.hic.Hic" title="fanc.hic.Hic"><code class="xref py py-class docutils literal notranslate"><span class="pre">Hic</span></code></a> object.</p>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="hic.html" class="btn btn-neutral float-right" title="Generate, bin, filter, and correct Hic objects" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="map.html" class="btn btn-neutral float-left" title="Mapping FASTQ files" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Kai Kruse, Vaquerizas lab

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>